{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Metagene plots\n",
    "Now, let's draw **metagene plots**: (using seprate conda env and ipython kernel)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using my `guitar` envirnment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reload_ext rpy2.ipython"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "suppressMessages(suppressWarnings(library (Guitar)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "R[write to console]: Import genomic features from the file as a GRanges object ... \n",
      "R[write to console]: OK\n",
      "\n",
      "R[write to console]: Prepare the 'metadata' data frame ... \n",
      "R[write to console]: OK\n",
      "\n",
      "R[write to console]: Make the TxDb object ... \n",
      "R[write to console]: OK\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%R \n",
    "txdb <- makeTxDbFromGFF('/rumi/shams/abe/genomes/hg38/gencode.v34/gencode.v34.annotation.gtf',organism='Homo sapiens')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"20210323213038\"\n",
      "[1] \"There are 228048 transcripts of 60669 genes in the genome.\"\n",
      "[1] \"total 228048 transcripts extracted ...\"\n",
      "[1] \"total 97941 transcripts left after ambiguity filter ...\"\n",
      "[1] \"total 97941 transcripts left after check chromosome validity ...\"\n",
      "[1] \"total 10939 mRNAs left after component length filter ...\"\n",
      "[1] \"total 60598 ncRNAs left after ncRNA length filter ...\"\n",
      "[1] \"generate components for all tx\"\n",
      "[1] \"generate components for mRNA\"\n",
      "[1] \"generate components for lncRNA\"\n",
      "[1] \"generate chiped transcriptome\"\n",
      "[1] \"generate coverage checking ranges for tx\"\n",
      "[1] \"generate coverage checking ranges for mrna\"\n",
      "[1] \"generate coverage checking ranges for ncrna\"\n",
      "[1] \"20210323213529\"\n",
      "[1] \"import BED file radar/result.sig.bed\"\n",
      "[1] \"sample 10 points for Group1\"\n",
      "[1] \"start figure plotting for tx ...\"\n",
      "[1] \"start figure plotting for mrna ...\"\n",
      "[1] \"start figure plotting for ncrna ...\"\n"
     ]
    }
   ],
   "source": [
    "%%R \n",
    "GuitarPlot(txTxdb = txdb,stBedFiles = list(\"radar/result.sig.bed\"),miscOutFilePrefix = \"Guitar\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renamed 'Guitar_mrna_test.pdf' -> 'plots/Guitar_mrna_test.pdf'\r\n",
      "renamed 'Guitar_ncrna_test.pdf' -> 'plots/Guitar_ncrna_test.pdf'\r\n",
      "renamed 'Guitar_tx_test.pdf' -> 'plots/Guitar_tx_test.pdf'\r\n"
     ]
    }
   ],
   "source": [
    "mv -v Guitar_* plots/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "plots/Guitar_mrna_test.pdf > plots/Guitar_mrna_test.png\r\n",
      "/rumi/shams/abe/Workflows/my_scripts/pdf2png.sh: line 5: magick: command not found\r\n",
      "done!\r\n",
      "plots/Guitar_ncrna_test.pdf > plots/Guitar_ncrna_test.png\r\n",
      "/rumi/shams/abe/Workflows/my_scripts/pdf2png.sh: line 5: magick: command not found\r\n",
      "done!\r\n",
      "plots/Guitar_tx_test.pdf > plots/Guitar_tx_test.png\r\n",
      "/rumi/shams/abe/Workflows/my_scripts/pdf2png.sh: line 5: magick: command not found\r\n",
      "done!\r\n"
     ]
    }
   ],
   "source": [
    "!for f in plots/Guitar*.pdf; do bash ~/Workflows/my_scripts/pdf2png.sh $f; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FIRE \n",
    "https://github.com/goodarzilab/FIRE\n",
    "### Motif analysis "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 1: Extract mRNA exon sequences for significant hits from RADAR results "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "# step 1: extract mRNA exon sequences for significant hits from RADAR \n",
    "bedtools getfasta -name -s -fi /rumi/shams/genomes/hg38/hg38.fa \\\n",
    "-bed radar/result.all.bed \\\n",
    "-split -fo radar/result.all.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 2: Prepare inputs for FIRE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using my `bioseq` conda environment . "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# %%bash\n",
    "# # step 2: prepare inputs for FIRE\n",
    "# perl ${TEISERDIR}prep_seqs_for_teiser_run.pl radar/result.sig.fa radar/result.sig\n",
    "\n",
    "!python $TEISERDIR/prep_fasta_for_fire_run.py radar/result.all.fa"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 3: Run FIRE for known m6A motifs (non-discovery mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[AG]GAC\r\n",
      "[AGT][AG]AC[ACT]\r\n"
     ]
    }
   ],
   "source": [
    "cat radar/known_motifs.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Mar 23 23:00:41 UTC 2021\n",
      "RNA, Processing result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt\n",
      "RNA, Step 2: skip seed optimization, use -motiffile_rna option instead.\n",
      "RNA, Step 3: evaluation of motif significances.\n",
      "Processing motif [AG]GAC.\n",
      "MI=0.00076809\n",
      "Shuffle rank=17\n",
      "Z-score=6.380\n",
      "Sigificance category= NOT-SIGNIF\n",
      "Robustness (jn_f=3)=0/10\n",
      "\n",
      "Processing motif [AGT][AG]AC[ACT].\n",
      "MI=0.00008237\n",
      "Shuffle rank=3228\n",
      "Z-score=0.039\n",
      "Sigificance category= NOT-SIGNIF\n",
      "Robustness (jn_f=3)=0/10\n",
      "\n",
      "RNA, Step 3.5: creating binary expression profiles.\n",
      "RNA, Step 4: discovery of distance constraints.\n",
      "Read 2 motifs.\n",
      "Read 9252 (genes) x 2 (conditions) matrix.\n",
      "Read 9252 (genes) x 1 (conditions) matrix.\n",
      "Reading sequences and building motif profiles ...Done.\n",
      "Average sequence length is 50.\n",
      "Quantized E vector into 2 bins\n",
      "Motif [AG]GAC, present in 3240 genes.\n",
      "       Average distance: MI=0.0014, rank=1711/10000, Z=0.8359, mbins=5\n",
      "  Orientation bias {5'): MI=0.0008, rank=16/10000, Z=6.2295\n",
      "  Orientation bias (3'): MI=0.0002, rank=886/10000, Z=1.4430\n",
      "       Copy number (1+): 3240 genes, MI=0.0008, rank=22/10000, Z=6.2899\n",
      "       Copy number (2+): 669 genes, MI=0.0001, rank=3788/10000, Z=-0.1049\n",
      "       Copy number (3+): 85 genes, MI=0.0000, rank=6662/10000, Z=-0.5004\n",
      "       Copy number (4+): 8 genes, MI=0.0000, rank=7212/10000, Z=-0.3673\n",
      "       Copy number (5+): 2 genes, MI=0.0000, rank=10000/10000, Z=-0.9944\n",
      "       Copy number (6+): 2 genes, MI=0.0000, rank=10000/10000, Z=-1.0082\n",
      "       Copy number (7+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "       Copy number (8+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif [AGT][AG]AC[ACT], present in 5732 genes.\n",
      "       Average distance: MI=0.0000, rank=10000/10000, Z=0.0000, mbins=2\n",
      "  Orientation bias {5'): MI=0.0001, rank=3133/10000, Z=0.0415\n",
      "  Orientation bias (3'): MI=0.0003, rank=610/10000, Z=1.8317\n",
      "       Copy number (1+): 5732 genes, MI=0.0001, rank=3121/10000, Z=0.0500\n",
      "       Copy number (2+): 2457 genes, MI=0.0000, rank=7416/10000, Z=-0.6148\n",
      "       Copy number (3+): 698 genes, MI=0.0000, rank=7844/10000, Z=-0.6456\n",
      "       Copy number (4+): 121 genes, MI=0.0001, rank=1991/10000, Z=0.6360\n",
      "       Copy number (5+): 24 genes, MI=0.0001, rank=5385/10000, Z=-0.2422\n",
      "       Copy number (6+): 8 genes, MI=0.0000, rank=7290/10000, Z=-0.3741\n",
      "       Copy number (7+): 3 genes, MI=0.0000, rank=10000/10000, Z=-0.5792\n",
      "       Copy number (8+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "\n",
      "RNA, Step 5.5: conservation.\n",
      "Setting -rna to 1.\n",
      "[AG]GAC\t98.12\n",
      "[AGT][AG]AC[ACT]\t21.42\n",
      "RNA, Step 5: summarize information.\n",
      "Reading RNA motifs mi_dist output ...Done.\n",
      "Parsing RNA motif file ... Done.\n",
      "Reading mi_signif output file for RNA motifs ...!!!! 1004024!!!! 900512Done.\n",
      "Got 0 motifs\n",
      "Writing good columns to result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.signifcolumns ...Done.\n",
      "Outputing summary ... Done.\n",
      "HI!HELLO!result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary.binary created.\n",
      "RNA, Step 5.3: generate profiles for motifs in summary file.\n",
      "Setting -rna to 1.\n",
      "RNA, Step 5.3 bis: Generate profiles for /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa.\n",
      "Setting -rna to 1.\n",
      "RNA, Determining target gene sets.\n",
      "RNA, Drawing position histograms.\n",
      "RNA, Step 6: cluster motifs.\n",
      "Using positive correlation.\n",
      "No motifs in summary file.\n",
      "RNA, Step 5.6: generate genome-wide GO enrichments for motifs in summary file.\n",
      "RNA, draw motif maps for co-localizing pairs\n",
      "RNA, Step 6.8: identifying motifs that match micrornas.\n",
      "RNA, Step 7: draw matrix figure.\n",
      "result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "RNA, Step 7b: draw density figure.\n",
      "result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "Parsing distance report file ...  Done.\n",
      "RNA, Step 10: motif report.\n",
      "Tue Mar 23 23:01:23 UTC 2021\n",
      "renamed 'radar/result.sig_fire.txt_FIRE/' -> 'fire/known_motifs_FIRE'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No motifs.\n",
      "Can't use 'defined(@array)' (Maybe you should just omit the defined()?) at /flash/bin/FIRE-1.1/SCRIPTS/draw_position_histogram.pl line 134.\n",
      "Table.pm: cannot open file \"result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.clusters\" ..\n",
      "Table.pm: cannot open file \"result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.mimatrix\" ..\n",
      "No motifs in @MOTIFS ...\n",
      "No motifs in @MOTIFS ...\n",
      "No motifs in @MOTIFS ...\n",
      "No motifs in @MOTIFS ...\n",
      "result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.fullmimatrix does not exists. at /flash/bin/FIRE-1.1/SCRIPTS/Sets.pm line 165.\n",
      "result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.fullmimatrix does not exists. at /flash/bin/FIRE-1.1/SCRIPTS/Sets.pm line 165.\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd radar/\n",
    "# step 3: run FIRE for known m6A motifs (non-discovery mode)\n",
    "perl $FIREDIR/fire.pl \\\n",
    "    --expfile=result.sig_fire.txt \\\n",
    "    --exptype=discrete \\\n",
    "    --fastafile_rna=result.sig_fire.fa \\\n",
    "    --nodups=1 --dodna=0 --dodnarna=0 \\\n",
    "    --species=human \\\n",
    "    --doskipdiscovery=1 \\\n",
    "    --motiffile_rna=known_motifs.txt\\\n",
    "    --oribiasonly=0 #--jn_f=0\n",
    "\n",
    "cd ../\n",
    "\n",
    "mkdir -p fire \n",
    "rm -r fire/known_motifs_FIRE\n",
    "mv -v radar/result.sig_fire.txt_FIRE/ fire/known_motifs_FIRE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Mar 23 23:01:24 UTC 2021\n",
      "RNA, Processing result.all_fire.txt_FIRE/RNA/result.all_fire.txt\n",
      "RNA, Step 2: skip seed optimization, use -motiffile_rna option instead.\n",
      "RNA, Step 3: evaluation of motif significances.\n",
      "Processing motif [AG]GAC.\n",
      "MI=0.00017159\n",
      "Shuffle rank=0\n",
      "Z-score=14.473\n",
      "Sigificance category= OK-NO-SEED\n",
      "Robustness (jn_f=3)=3/10\n",
      "\n",
      "Processing motif [AGT][AG]AC[ACT].\n",
      "MI=0.00000411\n",
      "Shuffle rank=4747\n",
      "Z-score=-0.336\n",
      "Sigificance category= NOT-SIGNIF\n",
      "Robustness (jn_f=3)=0/10\n",
      "\n",
      "RNA, Step 3.5: creating binary expression profiles.\n",
      "RNA, Step 4: discovery of distance constraints.\n",
      "Read 2 motifs.\n",
      "Read 91132 (genes) x 2 (conditions) matrix.\n",
      "Read 91132 (genes) x 1 (conditions) matrix.\n",
      "Reading sequences and building motif profiles ...Done.\n",
      "Average sequence length is 118.\n",
      "Quantized E vector into 2 bins\n",
      "Motif [AG]GAC, present in 43454 genes.\n",
      "       Average distance: MI=0.0003, rank=13/10000, Z=4.8676, mbins=5\n",
      "  Orientation bias {5'): MI=0.0002, rank=0/10000, Z=14.5830\n",
      "  Orientation bias (3'): MI=0.0005, rank=0/10000, Z=39.1810\n",
      "       Copy number (1+): 43454 genes, MI=0.0002, rank=0/10000, Z=14.3535\n",
      "       Copy number (2+): 20260 genes, MI=0.0003, rank=0/10000, Z=24.5077\n",
      "       Copy number (3+): 10943 genes, MI=0.0003, rank=0/10000, Z=26.2810\n",
      "       Copy number (4+): 6862 genes, MI=0.0003, rank=0/10000, Z=24.8346\n",
      "       Copy number (5+): 4531 genes, MI=0.0003, rank=0/10000, Z=25.2067\n",
      "       Copy number (6+): 3034 genes, MI=0.0002, rank=0/10000, Z=19.3466\n",
      "       Copy number (7+): 2035 genes, MI=0.0001, rank=0/10000, Z=12.7806\n",
      "       Copy number (8+): 1377 genes, MI=0.0001, rank=1/10000, Z=9.4928\n",
      "       Copy number (9+): 998 genes, MI=0.0001, rank=15/10000, Z=6.0411\n",
      "       Copy number (10+): 693 genes, MI=0.0001, rank=3/10000, Z=8.9028\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif [AGT][AG]AC[ACT], present in 64459 genes.\n",
      "       Average distance: MI=0.0000, rank=10000/10000, Z=0.0000, mbins=2\n",
      "  Orientation bias {5'): MI=0.0000, rank=4698/10000, Z=-0.3399\n",
      "  Orientation bias (3'): MI=0.0002, rank=0/10000, Z=20.9175\n",
      "       Copy number (1+): 64459 genes, MI=0.0000, rank=4816/10000, Z=-0.3418\n",
      "       Copy number (2+): 40847 genes, MI=0.0000, rank=3465/10000, Z=-0.0593\n",
      "       Copy number (3+): 26004 genes, MI=0.0000, rank=9373/10000, Z=-0.7070\n",
      "       Copy number (4+): 18097 genes, MI=0.0000, rank=8611/10000, Z=-0.6960\n",
      "       Copy number (5+): 13678 genes, MI=0.0000, rank=5508/10000, Z=-0.4520\n",
      "       Copy number (6+): 10726 genes, MI=0.0000, rank=8099/10000, Z=-0.6612\n",
      "       Copy number (7+): 8487 genes, MI=0.0000, rank=8931/10000, Z=-0.6895\n",
      "       Copy number (8+): 6890 genes, MI=0.0000, rank=6207/10000, Z=-0.5194\n",
      "       Copy number (9+): 5632 genes, MI=0.0000, rank=9883/10000, Z=-0.7162\n",
      "       Copy number (10+): 4630 genes, MI=0.0000, rank=10000/10000, Z=-0.7146\n",
      "\n",
      "RNA, Step 5.5: conservation.\n",
      "Setting -rna to 1.\n",
      "[AG]GAC\t98.12\n",
      "[AGT][AG]AC[ACT]\t21.42\n",
      "RNA, Step 5: summarize information.\n",
      "Reading RNA motifs mi_dist output ...Done.\n",
      "Parsing RNA motif file ... Done.\n",
      "Reading mi_signif output file for RNA motifs ...!!!! 1132328!!!! 900512Done.\n",
      "Got 0 motifs\n",
      "Writing good columns to result.all_fire.txt_FIRE/RNA/result.all_fire.txt.signifcolumns ...Done.\n",
      "Outputing summary ... Done.\n",
      "HI!HELLO!result.all_fire.txt_FIRE/RNA/result.all_fire.txt.summary.binary created.\n",
      "RNA, Step 5.3: generate profiles for motifs in summary file.\n",
      "Setting -rna to 1.\n",
      "RNA, Step 5.3 bis: Generate profiles for /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa.\n",
      "Setting -rna to 1.\n",
      "RNA, Determining target gene sets.\n",
      "RNA, Drawing position histograms.\n",
      "RNA, Step 6: cluster motifs.\n",
      "Using positive correlation.\n",
      "No motifs in summary file.\n",
      "RNA, Step 5.6: generate genome-wide GO enrichments for motifs in summary file.\n",
      "RNA, draw motif maps for co-localizing pairs\n",
      "RNA, Step 6.8: identifying motifs that match micrornas.\n",
      "RNA, Step 7: draw matrix figure.\n",
      "result.all_fire.txt_FIRE/RNA/result.all_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "result.all_fire.txt_FIRE/RNA/result.all_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "RNA, Step 7b: draw density figure.\n",
      "result.all_fire.txt_FIRE/RNA/result.all_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "result.all_fire.txt_FIRE/RNA/result.all_fire.txt.clusters does not exist, ignoring.\n",
      "Now doing the graphical display.\n",
      "Parsing distance report file ...  Done.\n",
      "RNA, Step 10: motif report.\n",
      "Tue Mar 23 23:11:06 UTC 2021\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No motifs.\n",
      "Can't use 'defined(@array)' (Maybe you should just omit the defined()?) at /flash/bin/FIRE-1.1/SCRIPTS/draw_position_histogram.pl line 134.\n",
      "Table.pm: cannot open file \"result.all_fire.txt_FIRE/RNA/result.all_fire.txt.clusters\" ..\n",
      "Table.pm: cannot open file \"result.all_fire.txt_FIRE/RNA/result.all_fire.txt.mimatrix\" ..\n",
      "No motifs in @MOTIFS ...\n",
      "No motifs in @MOTIFS ...\n",
      "No motifs in @MOTIFS ...\n",
      "No motifs in @MOTIFS ...\n",
      "result.all_fire.txt_FIRE/RNA/result.all_fire.txt.fullmimatrix does not exists. at /flash/bin/FIRE-1.1/SCRIPTS/Sets.pm line 165.\n",
      "result.all_fire.txt_FIRE/RNA/result.all_fire.txt.fullmimatrix does not exists. at /flash/bin/FIRE-1.1/SCRIPTS/Sets.pm line 165.\n",
      "mv: cannot stat 'radar/result.sig_fire.txt_FIRE/': No such file or directory\n"
     ]
    },
    {
     "ename": "CalledProcessError",
     "evalue": "Command 'b'cd radar/\\n# step 3: run FIRE for known m6A motifs (non-discovery mode)\\nperl $FIREDIR/fire.pl \\\\\\n    --expfile=result.all_fire.txt \\\\\\n    --exptype=discrete \\\\\\n    --fastafile_rna=result.all_fire.fa \\\\\\n    --nodups=1 --dodna=0 --dodnarna=0 \\\\\\n    --species=human \\\\\\n    --doskipdiscovery=1 \\\\\\n    --motiffile_rna=known_motifs.txt\\\\\\n    --oribiasonly=0 #--jn_f=0\\n\\ncd ../\\n\\nmkdir -p fire \\nrm -r fire/known_motifs_FIRE\\nmv -v radar/result.sig_fire.txt_FIRE/ fire/known_motifs_all_res_FIRE\\n'' returned non-zero exit status 1.",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mCalledProcessError\u001b[0m                        Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-4-11bcd4f29c3c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mget_ipython\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrun_cell_magic\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'bash'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m''\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'cd radar/\\n# step 3: run FIRE for known m6A motifs (non-discovery mode)\\nperl $FIREDIR/fire.pl \\\\\\n    --expfile=result.all_fire.txt \\\\\\n    --exptype=discrete \\\\\\n    --fastafile_rna=result.all_fire.fa \\\\\\n    --nodups=1 --dodna=0 --dodnarna=0 \\\\\\n    --species=human \\\\\\n    --doskipdiscovery=1 \\\\\\n    --motiffile_rna=known_motifs.txt\\\\\\n    --oribiasonly=0 #--jn_f=0\\n\\ncd ../\\n\\nmkdir -p fire \\nrm -r fire/known_motifs_FIRE\\nmv -v radar/result.sig_fire.txt_FIRE/ fire/known_motifs_all_res_FIRE\\n'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/envs/perl/lib/python3.8/site-packages/IPython/core/interactiveshell.py\u001b[0m in \u001b[0;36mrun_cell_magic\u001b[0;34m(self, magic_name, line, cell)\u001b[0m\n\u001b[1;32m   2379\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mbuiltin_trap\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2380\u001b[0m                 \u001b[0margs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mmagic_arg_s\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2381\u001b[0;31m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2382\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mresult\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2383\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/perl/lib/python3.8/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mnamed_script_magic\u001b[0;34m(line, cell)\u001b[0m\n\u001b[1;32m    140\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    141\u001b[0m                 \u001b[0mline\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mscript\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 142\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshebang\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mline\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    143\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    144\u001b[0m         \u001b[0;31m# write a basic docstring:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<decorator-gen-103>\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/perl/lib/python3.8/site-packages/IPython/core/magic.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(f, *a, **k)\u001b[0m\n\u001b[1;32m    185\u001b[0m     \u001b[0;31m# but it's overkill for just that one bit of state.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    186\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mmagic_deco\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 187\u001b[0;31m         \u001b[0mcall\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    188\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    189\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mcallable\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/anaconda3/envs/perl/lib/python3.8/site-packages/IPython/core/magics/script.py\u001b[0m in \u001b[0;36mshebang\u001b[0;34m(self, line, cell)\u001b[0m\n\u001b[1;32m    243\u001b[0m             \u001b[0msys\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstderr\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflush\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    244\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mraise_error\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m!=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 245\u001b[0;31m             \u001b[0;32mraise\u001b[0m \u001b[0mCalledProcessError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreturncode\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0moutput\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mstderr\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0merr\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_run_script\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mp\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcell\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mto_close\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mCalledProcessError\u001b[0m: Command 'b'cd radar/\\n# step 3: run FIRE for known m6A motifs (non-discovery mode)\\nperl $FIREDIR/fire.pl \\\\\\n    --expfile=result.all_fire.txt \\\\\\n    --exptype=discrete \\\\\\n    --fastafile_rna=result.all_fire.fa \\\\\\n    --nodups=1 --dodna=0 --dodnarna=0 \\\\\\n    --species=human \\\\\\n    --doskipdiscovery=1 \\\\\\n    --motiffile_rna=known_motifs.txt\\\\\\n    --oribiasonly=0 #--jn_f=0\\n\\ncd ../\\n\\nmkdir -p fire \\nrm -r fire/known_motifs_FIRE\\nmv -v radar/result.sig_fire.txt_FIRE/ fire/known_motifs_all_res_FIRE\\n'' returned non-zero exit status 1."
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd radar/\n",
    "# step 3: run FIRE for known m6A motifs (non-discovery mode)\n",
    "perl $FIREDIR/fire.pl \\\n",
    "    --expfile=result.all_fire.txt \\\n",
    "    --exptype=discrete \\\n",
    "    --fastafile_rna=result.all_fire.fa \\\n",
    "    --nodups=1 --dodna=0 --dodnarna=0 \\\n",
    "    --species=human \\\n",
    "    --doskipdiscovery=1 \\\n",
    "    --motiffile_rna=known_motifs.txt\\\n",
    "    --oribiasonly=0 #--jn_f=0\n",
    "\n",
    "cd ../\n",
    "\n",
    "mkdir -p fire \n",
    "rm -r fire/known_motifs_FIRE\n",
    "mv -v radar/result.sig_fire.txt_FIRE/ fire/known_motifs_all_res_FIRE"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Step 4: Run FIRE for all possible motifs (discovery mode)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Tue Mar 23 22:07:32 UTC 2021\n",
      "RNA, Processing result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt\n",
      "RNA, Step 1: seed discovery.\n",
      "Setting -shuffle to 16384.\n",
      "Expression file is a 9252 by 1 matrix.\n",
      "Min expression value read is 0, max is 1\n",
      "Reading sequences ... Done\n",
      "Number of ORFs = 9252\n",
      "Calculating Mutual Information for all seeds ... Done\n",
      "top 10 seeds...\n",
      "AGAAACC\n",
      "Determining threshold ...\n",
      "Evaluating seed #0, AGAAACC ... passed.\n",
      "Evaluating seed #100, CCCTCGA ... not passed.\n",
      "Decreasing intervals phase.\n",
      "Evaluating seed #50, AGAGAAA ... passed.\n",
      "Evaluating seed #75, AGTCGAG ... not passed.\n",
      "Evaluating seed #62, GTAATGA ... not passed.\n",
      "Evaluating seed #56, CTTGACA ... not passed.\n",
      "Searches for 5 consecutive 'not passed'.\n",
      "Threshold set to seed #50 (included).\n",
      "AGAAACC\t0.0050\n",
      "GAGAAAC\t0.0046\n",
      "GAATGTG\t0.0046\n",
      "AATGTGG\t0.0039\n",
      "CATACTG\t0.0037\n",
      "ATACTGG\t0.0035\n",
      "ACTGGAG\t0.0030\n",
      "CTGGAGA\t0.0027\n",
      "TGTGGCA\t0.0026\n",
      "TGAATGT\t0.0026\n",
      "TTCATAC\t0.0024\n",
      "GTGGGAA\t0.0024\n",
      "TCATACT\t0.0024\n",
      "CCCCCCT\t0.0020\n",
      "GTGGCAA\t0.0018\n",
      "TACAAAT\t0.0018\n",
      "GAAACCT\t0.0018\n",
      "ATGAATG\t0.0017\n",
      "AACCTTA\t0.0017\n",
      "ATGTGGC\t0.0017\n",
      "GTCAGGT\t0.0017\n",
      "TCCCCCC\t0.0017\n",
      "ACTGATT\t0.0017\n",
      "AAACCCT\t0.0016\n",
      "TACTGGA\t0.0016\n",
      "ACCTTAC\t0.0016\n",
      "CCCCCCC\t0.0016\n",
      "ATTCATA\t0.0016\n",
      "CAAATGT\t0.0016\n",
      "TATATTT\t0.0015\n",
      "TGGAGAG\t0.0015\n",
      "AACCCTA\t0.0015\n",
      "AATGTGA\t0.0015\n",
      "TGGGAAA\t0.0015\n",
      "AAGCCAT\t0.0014\n",
      "CATCAGA\t0.0014\n",
      "ATGTGGA\t0.0014\n",
      "ATGTGAA\t0.0014\n",
      "GAGAGAA\t0.0014\n",
      "TGGAAAA\t0.0014\n",
      "AGAATTC\t0.0014\n",
      "CATCTTC\t0.0014\n",
      "CCTGGAC\t0.0014\n",
      "TGGCAAA\t0.0014\n",
      "AATTCAT\t0.0013\n",
      "GCGACCG\t0.0013\n",
      "AAGCCTT\t0.0013\n",
      "GAATTCA\t0.0013\n",
      "GCAAGCC\t0.0013\n",
      "AAAGCCT\t0.0013\n",
      "AGAGAAA\t0.0013\n",
      "RNA, Step 2: seed optimization.\n",
      "Using minr = 2.0000.\n",
      "Motif length will be 9.\n",
      "Creating prefix tree for storing patterns ... Done.\n",
      "Reading expression data ... Done.\n",
      "Creating hash table for gene names ... Done.\n",
      "Reading sequences and building the index.\n",
      "157378 k-mers in index\n",
      "Matrix occupancy = 360419 / 1456061256.000000\n",
      "Done.                                \n",
      "\n",
      "Evaluating seed AGAAACC ... \n",
      "optimizing.\n",
      "-1213315568\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: GA[AG]AA[AG]C[CG].\n",
      "Mutual information: 0.007572\n",
      "Optimization stability: 2\n",
      "Z-score: 69.687\n",
      "Starting seed: AGAAACC\n",
      "Seed MI: 0.005169\n",
      "\n",
      "Evaluating seed GAGAAAC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0156\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.016 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GAATGTG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.3651\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.365 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AATGTGG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.3129\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.313 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CATACTG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.2594\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.259 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ATACTGG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1431\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.143 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ACTGGAG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1317\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.132 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CTGGAGA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1050\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.105 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TGTGGCA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.4155\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.416 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TGAATGT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.3653\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.365 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TTCATAC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.2206\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.221 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GTGGGAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.9145\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.914 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TCATACT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1658\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.166 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CCCCCCT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=30562.3311\n",
      "optimizing.\n",
      "-1211723424\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: [AC][CT]CCCC[CG][CT].\n",
      "Mutual information: 0.003972\n",
      "Optimization stability: 1\n",
      "Z-score: 33.928\n",
      "Starting seed: CCCCCCT\n",
      "Seed MI: 0.001887\n",
      "\n",
      "Evaluating seed GTGGCAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.3500\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.350 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TACAAAT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0950\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.095 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GAAACCT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0254\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.025 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ATGAATG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1425\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.143 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AACCTTA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0697\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.070 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ATGTGGC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.3855\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.385 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GTCAGGT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=40.2978\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=13.3273\n",
      "optimizing.\n",
      "-1210575568\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: .GTCAGGT.\n",
      "Mutual information: 0.001623\n",
      "Optimization stability: 10\n",
      "Z-score: 13.308\n",
      "Starting seed: GTCAGGT\n",
      "Seed MI: 0.001623\n",
      "\n",
      "Evaluating seed TCCCCCC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=15.1288\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=0.0307\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.031 for [AC][CT]CCCC[CG][CT]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ACTGATT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=18.1945\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=21.8418\n",
      "Motif .GTCAGGT. : I(B;E|A)/I(A;B)=195.9997\n",
      "optimizing.\n",
      "-1209797952\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: ..[ACT].G[ACG]T[ACT][AG]\n",
      "Mutual information: 0.006954\n",
      "Optimization stability: 1\n",
      "Z-score: 61.005\n",
      "Starting seed: ACTGATT\n",
      "Seed MI: 0.001584\n",
      "\n",
      "Evaluating seed AAACCCT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0203\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.020 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TACTGGA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0719\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.072 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ACCTTAC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1033\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.103 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CCCCCCC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=13.8261\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=0.0077\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.008 for [AC][CT]CCCC[CG][CT]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ATTCATA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.2062\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.206 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CAAATGT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1301\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.130 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TATATTT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=18.8017\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=17.6658\n",
      "Motif .GTCAGGT. : I(B;E|A)/I(A;B)=299.4142\n",
      "Motif ..[ACT].G[ACG]T[ACT][AG] : I(B;E|A)/I(A;B)=1.3527\n",
      "optimizing.\n",
      "-1208650096\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: [ACG][CT][AGT][CGT][AC]T[AGT][CT].\n",
      "Mutual information: 0.006331\n",
      "Optimization stability: 4\n",
      "Z-score: 55.904\n",
      "Starting seed: TATATTT\n",
      "Seed MI: 0.001445\n",
      "\n",
      "Evaluating seed TGGAGAG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0896\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.090 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AACCCTA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0242\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.024 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AATGTGA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.2186\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.219 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TGGGAAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=1.5498\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 1.550 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AAGCCAT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=1.8072\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 1.807 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CATCAGA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=1.7754\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 1.775 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ATGTGGA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.5138\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.514 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed ATGTGAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.3683\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.368 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GAGAGAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0136\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.014 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed TGGAAAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=1.0498\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 1.050 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AGAATTC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.3084\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.308 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CATCTTC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=10.7278\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=11.4080\n",
      "Motif .GTCAGGT. : I(B;E|A)/I(A;B)=106.6125\n",
      "Motif ..[ACT].G[ACG]T[ACT][AG] : I(B;E|A)/I(A;B)=79.5600\n",
      "Motif [ACG][CT][AGT][CGT][AC]T[AGT][CT]. : I(B;E|A)/I(A;B)=0.1484\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.148 for [ACG][CT][AGT][CGT][AC]T[AGT][CT]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed CCTGGAC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=7.2022\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=92369601.4683\n",
      "Motif .GTCAGGT. : I(B;E|A)/I(A;B)=6.3025\n",
      "Motif ..[ACT].G[ACG]T[ACT][AG] : I(B;E|A)/I(A;B)=5.6828\n",
      "Motif [ACG][CT][AGT][CGT][AC]T[AGT][CT]. : I(B;E|A)/I(A;B)=23.5019\n",
      "optimizing.\n",
      "-1207057952\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: [ACT].[ACG][AT]GGA.[AGT]\n",
      "Mutual information: 0.008344\n",
      "Optimization stability: 1\n",
      "Z-score: 76.036\n",
      "Starting seed: CCTGGAC\n",
      "Seed MI: 0.001230\n",
      "\n",
      "Evaluating seed TGGCAAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.4428\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.443 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AATTCAT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.2297\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.230 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GCGACCG ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=40.2948\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=47.4165\n",
      "Motif .GTCAGGT. : I(B;E|A)/I(A;B)=428.3121\n",
      "Motif ..[ACT].G[ACG]T[ACT][AG] : I(B;E|A)/I(A;B)=436.9845\n",
      "Motif [ACG][CT][AGT][CGT][AC]T[AGT][CT]. : I(B;E|A)/I(A;B)=15.0292\n",
      "Motif [ACT].[ACG][AT]GGA.[AGT] : I(B;E|A)/I(A;B)=2.4092\n",
      "optimizing.\n",
      "-1206169232\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: .[CGT]C[AG]A[ACT][CT]..\n",
      "Mutual information: 0.005055\n",
      "Optimization stability: 1\n",
      "Z-score: 45.532\n",
      "Starting seed: GCGACCG\n",
      "Seed MI: 0.001298\n",
      "\n",
      "Evaluating seed AAGCCTT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.2374\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.237 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GAATTCA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.1992\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.199 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed GCAAGCC ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=18.1883\n",
      "Motif [AC][CT]CCCC[CG][CT]. : I(B;E|A)/I(A;B)=21.8342\n",
      "Motif .GTCAGGT. : I(B;E|A)/I(A;B)=195.9333\n",
      "Motif ..[ACT].G[ACG]T[ACT][AG] : I(B;E|A)/I(A;B)=11.0085\n",
      "Motif [ACG][CT][AGT][CGT][AC]T[AGT][CT]. : I(B;E|A)/I(A;B)=17.8158\n",
      "Motif [ACT].[ACG][AT]GGA.[AGT] : I(B;E|A)/I(A;B)=10.7179\n",
      "Motif .[CGT]C[AG]A[ACT][CT].. : I(B;E|A)/I(A;B)=6.6690\n",
      "optimizing.\n",
      "-1205317568\t-1221286016\t9252\t2\t2\n",
      "Best optimized motif: .GC[AG]A[GT]CC.\n",
      "Mutual information: 0.002124\n",
      "Optimization stability: 10\n",
      "Z-score: 19.314\n",
      "Starting seed: GCAAGCC\n",
      "Seed MI: 0.001187\n",
      "\n",
      "Evaluating seed AAAGCCT ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.2987\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.299 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "\n",
      "Evaluating seed AGAGAAA ... \n",
      "Motif GA[AG]AA[AG]C[CG]. : I(B;E|A)/I(A;B)=0.0155\n",
      "Break early (motif is too close to existing optimized motif).\n",
      "not optimized, minratio = 0.015 for GA[AG]AA[AG]C[CG]. (< minr=2.000).\n",
      "RNA, Step 3: evaluation of motif significances.\n",
      "Processing motif GA[AG]AA[AG]C[CG]..\n",
      "MI=0.00757221\n",
      "Shuffle rank=0\n",
      "Z-score=68.138\n",
      "Sigificance category= OK\n",
      "Robustness (jn_f=3)=10/10\n",
      "\n",
      "Processing motif [AC][CT]CCCC[CG][CT]..\n",
      "MI=0.00397160\n",
      "Shuffle rank=0\n",
      "Z-score=34.743\n",
      "Sigificance category= OK\n",
      "Robustness (jn_f=3)=10/10\n",
      "\n",
      "Processing motif .GTCAGGT..\n",
      "MI=0.00162317\n",
      "Shuffle rank=0\n",
      "Z-score=13.221\n",
      "Sigificance category= OK\n",
      "Robustness (jn_f=3)=0/10\n",
      "\n",
      "Processing motif ..[ACT].G[ACG]T[ACT][AG].\n",
      "MI=0.00701643\n",
      "Shuffle rank=0\n",
      "Z-score=62.366\n",
      "Sigificance category= OK\n",
      "Robustness (jn_f=3)=10/10\n",
      "\n",
      "Processing motif [ACG][CT][AGT][CGT][AC]T[AGT][CT]..\n",
      "MI=0.00635926\n",
      "Shuffle rank=0\n",
      "Z-score=55.622\n",
      "Sigificance category= OK\n",
      "Robustness (jn_f=3)=10/10\n",
      "\n",
      "Processing motif [ACT].[ACG][AT]GGA.[AGT].\n",
      "MI=0.00834410\n",
      "Shuffle rank=0\n",
      "Z-score=74.214\n",
      "Sigificance category= OK\n",
      "Robustness (jn_f=3)=10/10\n",
      "\n",
      "Processing motif .[CGT]C[AG]A[ACT][CT]...\n",
      "MI=0.00511263\n",
      "Shuffle rank=0\n",
      "Z-score=45.546\n",
      "Sigificance category= OK-NO-SEED\n",
      "Robustness (jn_f=3)=10/10\n",
      "\n",
      "Processing motif .GC[AG]A[GT]CC..\n",
      "MI=0.00220314\n",
      "Shuffle rank=0\n",
      "Z-score=18.888\n",
      "Sigificance category= OK-NO-SEED\n",
      "Robustness (jn_f=3)=7/10\n",
      "\n",
      "RNA, Step 3.5: creating binary expression profiles.\n",
      "RNA, Step 4: discovery of distance constraints.\n",
      "Read 8 motifs.\n",
      "Read 9252 (genes) x 8 (conditions) matrix.\n",
      "Read 9252 (genes) x 1 (conditions) matrix.\n",
      "Reading sequences and building motif profiles ...Done.\n",
      "Average sequence length is 50.\n",
      "Quantized E vector into 2 bins\n",
      "Motif GA[AG]AA[AG]C[CG]., present in 155 genes.\n",
      "       Average distance: MI=0.0274, rank=2355/10000, Z=0.5464, mbins=5\n",
      "  Orientation bias {5'): MI=0.0076, rank=0/10000, Z=72.0074\n",
      "  Orientation bias (3'): MI=0.0002, rank=1998/10000, Z=0.7628\n",
      "       Copy number (1+): 155 genes, MI=0.0076, rank=0/10000, Z=68.8650\n",
      "       Copy number (2+): 3 genes, MI=0.0003, rank=2519/10000, Z=1.7232\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif [AC][CT]CCCC[CG][CT]., present in 136 genes.\n",
      "       Average distance: MI=0.0226, rank=4058/10000, Z=0.0097, mbins=5\n",
      "  Orientation bias {5'): MI=0.0040, rank=0/10000, Z=35.0477\n",
      "  Orientation bias (3'): MI=0.0005, rank=104/10000, Z=4.2421\n",
      "       Copy number (1+): 136 genes, MI=0.0040, rank=0/10000, Z=34.3324\n",
      "       Copy number (2+): 15 genes, MI=0.0007, rank=72/10000, Z=5.3734\n",
      "       Copy number (3+): 6 genes, MI=0.0006, rank=294/10000, Z=4.4374\n",
      "       Copy number (4+): 3 genes, MI=0.0003, rank=2500/10000, Z=1.7320\n",
      "       Copy number (5+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif .GTCAGGT., present in 15 genes.\n",
      "       Average distance: MI=0.0000, rank=10000/10000, Z=0.0000, mbins=2\n",
      "  Orientation bias {5'): MI=0.0016, rank=1/10000, Z=13.2152\n",
      "  Orientation bias (3'): MI=0.0001, rank=2825/10000, Z=0.4507\n",
      "       Copy number (1+): 15 genes, MI=0.0016, rank=1/10000, Z=13.7047\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif ..[ACT].G[ACG]T[ACT][AG], present in 2549 genes.\n",
      "       Average distance: MI=0.0017, rank=1067/10000, Z=1.2825, mbins=4\n",
      "  Orientation bias {5'): MI=0.0070, rank=0/10000, Z=60.6223\n",
      "  Orientation bias (3'): MI=0.0024, rank=0/10000, Z=21.6808\n",
      "       Copy number (1+): 2549 genes, MI=0.0070, rank=0/10000, Z=61.6926\n",
      "       Copy number (2+): 426 genes, MI=0.0019, rank=0/10000, Z=16.4055\n",
      "       Copy number (3+): 42 genes, MI=0.0003, rank=917/10000, Z=1.6932\n",
      "       Copy number (4+): 9 genes, MI=0.0000, rank=10000/10000, Z=-0.6277\n",
      "       Copy number (5+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif [ACG][CT][AGT][CGT][AC]T[AGT][CT]., present in 2738 genes.\n",
      "       Average distance: MI=0.0002, rank=3561/10000, Z=-0.0607, mbins=2\n",
      "  Orientation bias {5'): MI=0.0064, rank=0/10000, Z=57.3609\n",
      "  Orientation bias (3'): MI=0.0009, rank=7/10000, Z=7.5517\n",
      "       Copy number (1+): 2738 genes, MI=0.0064, rank=0/10000, Z=56.9850\n",
      "       Copy number (2+): 573 genes, MI=0.0039, rank=0/10000, Z=35.2527\n",
      "       Copy number (3+): 98 genes, MI=0.0008, rank=19/10000, Z=6.9766\n",
      "       Copy number (4+): 12 genes, MI=0.0005, rank=363/10000, Z=3.3756\n",
      "       Copy number (5+): 2 genes, MI=0.0000, rank=10000/10000, Z=-1.0034\n",
      "       Copy number (6+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "       Copy number (7+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "       Copy number (8+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "       Copy number (9+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "       Copy number (10+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif [ACT].[ACG][AT]GGA.[AGT], present in 2147 genes.\n",
      "       Average distance: MI=0.0026, rank=1036/10000, Z=1.3088, mbins=5\n",
      "  Orientation bias {5'): MI=0.0083, rank=0/10000, Z=74.8522\n",
      "  Orientation bias (3'): MI=0.0000, rank=7929/10000, Z=-0.6536\n",
      "       Copy number (1+): 2147 genes, MI=0.0083, rank=0/10000, Z=74.6616\n",
      "       Copy number (2+): 322 genes, MI=0.0019, rank=0/10000, Z=16.7793\n",
      "       Copy number (3+): 48 genes, MI=0.0005, rank=112/10000, Z=4.2902\n",
      "       Copy number (4+): 7 genes, MI=0.0008, rank=140/10000, Z=5.5328\n",
      "       Copy number (5+): 1 genes, MI=0.0001, rank=10000/10000, Z=0.0000\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif .[CGT]C[AG]A[ACT][CT].., present in 1851 genes.\n",
      "       Average distance: MI=0.0029, rank=1097/10000, Z=1.2640, mbins=5\n",
      "  Orientation bias {5'): MI=0.0051, rank=0/10000, Z=46.4261\n",
      "  Orientation bias (3'): MI=0.0029, rank=0/10000, Z=24.7400\n",
      "       Copy number (1+): 1851 genes, MI=0.0051, rank=0/10000, Z=47.7117\n",
      "       Copy number (2+): 198 genes, MI=0.0021, rank=0/10000, Z=18.6015\n",
      "       Copy number (3+): 19 genes, MI=0.0000, rank=10000/10000, Z=-0.6697\n",
      "       Copy number (4+): 3 genes, MI=0.0003, rank=2526/10000, Z=1.7200\n",
      "\n",
      "Quantized E vector into 2 bins\n",
      "Motif .GC[AG]A[GT]CC., present in 59 genes.\n",
      "       Average distance: MI=0.0185, rank=5668/10000, Z=-0.2976, mbins=3\n",
      "  Orientation bias {5'): MI=0.0022, rank=0/10000, Z=19.6078\n",
      "  Orientation bias (3'): MI=0.0009, rank=8/10000, Z=7.5574\n",
      "       Copy number (1+): 59 genes, MI=0.0022, rank=0/10000, Z=18.5160\n",
      "       Copy number (2+): 2 genes, MI=0.0002, rank=4973/10000, Z=1.0054\n",
      "\n",
      "RNA, Step 5.5: conservation.\n",
      "Setting -rna to 1.\n",
      "GA[AG]AA[AG]C[CG].\t56.18\n",
      "[AC][CT]CCCC[CG][CT].\t85.94\n",
      ".GTCAGGT.\t24.59\n",
      "..[ACT].G[ACG]T[ACT][AG]\t77.98\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t53.13\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t93.93\n",
      ".[CGT]C[AG]A[ACT][CT]..\t86.58\n",
      ".GC[AG]A[GT]CC.\t90.48\n",
      "RNA, Step 5: summarize information.\n",
      "Reading RNA motifs mi_dist output ...Done.\n",
      "Parsing RNA motif file ... Done.\n",
      "Reading mi_signif output file for RNA motifs ...!!!! 257240!!!! -132400!!!! -238192!!!! 76136!!!! -2640!!!! -2712!!!! -2736!!!! -224216Done.\n",
      "Got 7 motifs\n",
      "Writing good columns to result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.signifcolumns ...Done.\n",
      "Outputing summary ... Done.\n",
      "HI!HELLO!result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary.binary created.\n",
      "RNA, Step 5.3: generate profiles for motifs in summary file.\n",
      "Setting -rna to 1.\n",
      "Generating profile for [ACT].[ACG][AT]GGA.[AGT] ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re [ACT].[ACG][AT]GGA.[AGT] -fastafile result.sig_fire.fa -flank 25  -rna 1  -reldist 0  > genregexp8091671670351Done.\n",
      "Generating profile for GA[AG]AA[AG]C[CG]. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re GA[AG]AA[AG]C[CG]. -fastafile result.sig_fire.fa -flank 25  -rna 1  -reldist 0  > genregexp6671627376051Done.\n",
      "Generating profile for ..[ACT].G[ACG]T[ACT][AG] ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re ..[ACT].G[ACG]T[ACT][AG] -fastafile result.sig_fire.fa -flank 25  -rna 1  -reldist 0  > genregexp3611061377652Done.\n",
      "Generating profile for [ACG][CT][AGT][CGT][AC]T[AGT][CT]. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re [ACG][CT][AGT][CGT][AC]T[AGT][CT]. -fastafile result.sig_fire.fa -flank 25  -rna 1  -reldist 0  > genregexp6932166077135Done.\n",
      "Generating profile for .[CGT]C[AG]A[ACT][CT].. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re .[CGT]C[AG]A[ACT][CT].. -fastafile result.sig_fire.fa -flank 25  -rna 1  -reldist 0  > genregexp8996372671501Done.\n",
      "Generating profile for [AC][CT]CCCC[CG][CT]. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re [AC][CT]CCCC[CG][CT]. -fastafile result.sig_fire.fa -flank 25  -rna 1  -reldist 0  > genregexp3423157613760Done.\n",
      "Generating profile for .GC[AG]A[GT]CC. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re .GC[AG]A[GT]CC. -fastafile result.sig_fire.fa -flank 25  -rna 1  -reldist 0  > genregexp9173107676531Done.\n",
      "RNA, Step 5.3 bis: Generate profiles for /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa.\n",
      "Setting -rna to 1.\n",
      "Generating profile for [ACT].[ACG][AT]GGA.[AGT] ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re [ACT].[ACG][AT]GGA.[AGT] -fastafile /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa -flank 25  -rna 1  -reldist 0  > genregexp2971036375176Done.\n",
      "Generating profile for GA[AG]AA[AG]C[CG]. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re GA[AG]AA[AG]C[CG]. -fastafile /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa -flank 25  -rna 1  -reldist 0  > genregexp8826313507617Done.\n",
      "Generating profile for ..[ACT].G[ACG]T[ACT][AG] ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re ..[ACT].G[ACG]T[ACT][AG] -fastafile /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa -flank 25  -rna 1  -reldist 0  > genregexp5801564173706Done.\n",
      "Generating profile for [ACG][CT][AGT][CGT][AC]T[AGT][CT]. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re [ACG][CT][AGT][CGT][AC]T[AGT][CT]. -fastafile /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa -flank 25  -rna 1  -reldist 0  > genregexp1806534061717Done.\n",
      "Generating profile for .[CGT]C[AG]A[ACT][CT].. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re .[CGT]C[AG]A[ACT][CT].. -fastafile /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa -flank 25  -rna 1  -reldist 0  > genregexp5343175656701Done.\n",
      "Generating profile for [AC][CT]CCCC[CG][CT]. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re [AC][CT]CCCC[CG][CT]. -fastafile /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa -flank 25  -rna 1  -reldist 0  > genregexp8026105173676Done.\n",
      "Generating profile for .GC[AG]A[GT]CC. ... /flash/bin/FIRE-1.1/PROGRAMS/genregexp -re .GC[AG]A[GT]CC. -fastafile /flash/bin/FIRE-1.1/FIRE_DATA/HUMAN/CONSERVATION/human_chicken_3_0_300_chicken_orthologs.fa -flank 25  -rna 1  -reldist 0  > genregexp151567013676Done.\n",
      "RNA, Determining target gene sets.\n",
      "RNA, Drawing position histograms.\n",
      "RNA, Step 6: cluster motifs.\n",
      "Using positive correlation.\n",
      "Doing all stats.\n",
      "Summary file has 0 DNA motifs and 7 RNA motifs.\n",
      "Read in RNA sequences and build motif profiles ...Done.\n",
      "Outputing MI matrix ... Done.\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t0\n",
      "[ACT].[ACG][AT]GGA.[AGT]\tGA[AG]AA[AG]C[CG].\t0.0084\t0\t73.9033\t0.1054\t1823\t1.3079\t1823\t1.3079\n",
      "GA[AG]AA[AG]C[CG].\t0\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t..[ACT].G[ACG]T[ACT][AG]\t0.0378\t0\t340.9816\t0.0000\tnan\tnan\tnan\tnan\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t0.0182\t0\t162.1391\t0.1361\t337\t2.2750\t337\t2.2750\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\tGA[AG]AA[AG]C[CG].\t0.0032\t0\t27.7061\t0.0499\t1386\t1.5114\t1386\t1.5114\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t0\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t.[CGT]C[AG]A[ACT][CT]..\t0.0269\t0\t243.7382\t0.0000\tnan\tnan\tnan\tnan\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t[AC][CT]CCCC[CG][CT].\t0.0025\t0\t21.9693\t0.0000\tnan\tnan\tnan\tnan\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t.GC[AG]A[GT]CC.\t0.0011\t7\t9.3500\t0.0000\tnan\tnan\tnan\tnan\n",
      "..[ACT].G[ACG]T[ACT][AG]\t1\n",
      "..[ACT].G[ACG]T[ACT][AG]\t.[CGT]C[AG]A[ACT][CT]..\t0.0188\t0\t173.9911\t0.1468\t751\t1.6194\t751\t1.6194\n",
      ".[CGT]C[AG]A[ACT][CT]..\t1\n",
      "..[ACT].G[ACG]T[ACT][AG]\t[AC][CT]CCCC[CG][CT].\t0.0005\t156\t3.3836\t0.0136\tnan\tnan\tnan\tnan\n",
      "..[ACT].G[ACG]T[ACT][AG]\t.GC[AG]A[GT]CC.\t0.0001\t4310\t-0.2336\t0.0066\tnan\tnan\tnan\tnan\n",
      "[AC][CT]CCCC[CG][CT].\t2\n",
      "[AC][CT]CCCC[CG][CT].\t.GC[AG]A[GT]CC.\t0.0000\t10000\t-0.8692\t0.0121\tnan\tnan\tnan\tnan\n",
      ".GC[AG]A[GT]CC.\t3\n",
      "Finish calculated rank, z-scores for remaining entries ... [ACT].[ACG][AT]GGA.[AGT]\t[ACT].[ACG][AT]GGA.[AGT]\t0.5808\t0\t5305.2316\t1.0000\t302\t2.4679\t302\t2.4679\n",
      "GA[AG]AA[AG]C[CG].\tGA[AG]AA[AG]C[CG].\t0.1099\t0\t927.8289\t1.0000\t10100\t-100.0000\t10100\t-100.0000\n",
      "GA[AG]AA[AG]C[CG].\t..[ACT].G[ACG]T[ACT][AG]\t0.0037\t0\t31.9945\t0.0000\tnan\tnan\tnan\tnan\n",
      "GA[AG]AA[AG]C[CG].\t.[CGT]C[AG]A[ACT][CT]..\t0.0026\t0\t22.4076\t0.0000\tnan\tnan\tnan\tnan\n",
      "GA[AG]AA[AG]C[CG].\t[AC][CT]CCCC[CG][CT].\t0.0002\t2174\t1.4738\t0.0000\tnan\tnan\tnan\tnan\n",
      "GA[AG]AA[AG]C[CG].\t.GC[AG]A[GT]CC.\t0.0001\t6470\t0.3737\t0.0000\tnan\tnan\tnan\tnan\n",
      "..[ACT].G[ACG]T[ACT][AG]\t..[ACT].G[ACG]T[ACT][AG]\t0.6337\t0\t5703.9682\t1.0000\t775\t1.5671\t775\t1.5671\n",
      "..[ACT].G[ACG]T[ACT][AG]\t[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t0.0470\t0\t416.6222\t0.0000\tnan\tnan\tnan\tnan\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t0.6563\t0\t5985.5855\t1.0000\t1017\t1.3270\t1017\t1.3270\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t.[CGT]C[AG]A[ACT][CT]..\t0.0335\t0\t307.6566\t0.0000\tnan\tnan\tnan\tnan\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t[AC][CT]CCCC[CG][CT].\t0.0031\t0\t27.4071\t0.0000\tnan\tnan\tnan\tnan\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t.GC[AG]A[GT]CC.\t0.0014\t5\t11.3727\t0.0000\tnan\tnan\tnan\tnan\n",
      ".[CGT]C[AG]A[ACT][CT]..\t.[CGT]C[AG]A[ACT][CT]..\t0.5202\t0\t4734.1410\t1.0000\t946\t1.3840\t946\t1.3840\n",
      ".[CGT]C[AG]A[ACT][CT]..\t[AC][CT]CCCC[CG][CT].\t0.0018\t0\t15.4837\t0.0352\t4067\t0.3170\t4067\t0.3170\n",
      ".[CGT]C[AG]A[ACT][CT]..\t.GC[AG]A[GT]CC.\t0.0019\t0\t15.2054\t0.0594\t1667\t1.0840\t1667\t1.0840\n",
      "[AC][CT]CCCC[CG][CT].\t[AC][CT]CCCC[CG][CT].\t0.0917\t0\t897.6088\t1.0000\t5179\t-0.2135\t5179\t-0.2135\n",
      ".GC[AG]A[GT]CC.\t.GC[AG]A[GT]CC.\t0.0477\t0\t766.8153\t1.0000\t10100\t-100.0000\t10100\t-100.0000\n",
      "Done.\n",
      "RNA, Step 5.6: generate genome-wide GO enrichments for motifs in summary file.\n",
      "RNA, draw motif maps for co-localizing pairs\n",
      "Draw motif map for [ACT].[ACG][AT]GGA.[AGT] .. Done.\n",
      "Draw motif map for GA[AG]AA[AG]C[CG]. .. Done.\n",
      "Draw motif map for ..[ACT].G[ACG]T[ACT][AG] .. Done.\n",
      "Draw motif map for [ACG][CT][AGT][CGT][AC]T[AGT][CT]. .. Done.\n",
      "Draw motif map for .[CGT]C[AG]A[ACT][CT].. .. Done.\n",
      "Draw motif map for [AC][CT]CCCC[CG][CT]. .. Done.\n",
      "Draw motif map for .GC[AG]A[GT]CC. .. Done.\n",
      "RNA, Step 6.8: identifying motifs that match micrornas.\n",
      "RNA, Step 7: draw matrix figure.\n",
      "Now doing the graphical display.\n",
      "xsize = 1200, ysize = 580, xbase = 35, ybase = 250\n",
      "Processing [ACT].[ACG][AT]GGA.[AGT] ... Outputing motif 0.eps ... Done.\n",
      "Processing GA[AG]AA[AG]C[CG]. ... Outputing motif 1.eps ... Done.\n",
      "Processing [ACG][CT][AGT][CGT][AC]T[AGT][CT]. ... Outputing motif 3.eps ... Done.\n",
      "Processing ..[ACT].G[ACG]T[ACT][AG] ... Outputing motif 2.eps ... Done.\n",
      "Processing .[CGT]C[AG]A[ACT][CT].. ... Outputing motif 4.eps ... Done.\n",
      "Processing [AC][CT]CCCC[CG][CT]. ... Outputing motif 5.eps ... Done.\n",
      "Processing .GC[AG]A[GT]CC. ... Outputing motif 6.eps ... Done.\n",
      "Plotting significance boxes.\n",
      "Creating result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary.eps ...Done.\n",
      "Creating PDF result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary.pdf ... Done.\n",
      "Now doing the graphical display.\n",
      "Processing [ACT].[ACG][AT]GGA.[AGT] ... Outputing motif 0.png ... \n",
      "Processing GA[AG]AA[AG]C[CG]. ... Outputing motif 1.png ... \n",
      "Processing [ACG][CT][AGT][CGT][AC]T[AGT][CT]. ... Outputing motif 3.png ... \n",
      "Processing ..[ACT].G[ACG]T[ACT][AG] ... Outputing motif 2.png ... \n",
      "Processing .[CGT]C[AG]A[ACT][CT].. ... Outputing motif 4.png ... \n",
      "Processing [AC][CT]CCCC[CG][CT]. ... Outputing motif 5.png ... \n",
      "Processing .GC[AG]A[GT]CC. ... Outputing motif 6.png ... \n",
      "RNA, Step 7b: draw density figure.\n",
      "Now doing the graphical display.\n",
      "xsize = 1200, ysize = 580, xbase = 35, ybase = 250\n",
      "Processing [ACT].[ACG][AT]GGA.[AGT] ... Outputing motif 0.eps ... Done.\n",
      "Processing GA[AG]AA[AG]C[CG]. ... Outputing motif 1.eps ... Done.\n",
      "Processing [ACG][CT][AGT][CGT][AC]T[AGT][CT]. ... Outputing motif 3.eps ... Done.\n",
      "Processing ..[ACT].G[ACG]T[ACT][AG] ... Outputing motif 2.eps ... Done.\n",
      "Processing .[CGT]C[AG]A[ACT][CT].. ... Outputing motif 4.eps ... Done.\n",
      "Processing [AC][CT]CCCC[CG][CT]. ... Outputing motif 5.eps ... Done.\n",
      "Processing .GC[AG]A[GT]CC. ... Outputing motif 6.eps ... Done.\n",
      "Plotting significance boxes.\n",
      "Creating result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.densities.eps ...Done.\n",
      "Creating PDF result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.densities.pdf ... Done.\n",
      "Now doing the graphical display.\n",
      "Processing [ACT].[ACG][AT]GGA.[AGT] ... Outputing motif 0.png ... \n",
      "Processing GA[AG]AA[AG]C[CG]. ... Outputing motif 1.png ... \n",
      "Processing [ACG][CT][AGT][CGT][AC]T[AGT][CT]. ... Outputing motif 3.png ... \n",
      "Processing ..[ACT].G[ACG]T[ACT][AG] ... Outputing motif 2.png ... \n",
      "Processing .[CGT]C[AG]A[ACT][CT].. ... Outputing motif 4.png ... \n",
      "Processing [AC][CT]CCCC[CG][CT]. ... Outputing motif 5.png ... \n",
      "Processing .GC[AG]A[GT]CC. ... Outputing motif 6.png ... \n",
      "Parsing distance report file ...  Done.\n",
      "..[ACT].G[ACG]T[ACT][AG] has no position or orientation bias.\n",
      ".[CGT]C[AG]A[ACT][CT].. has no position or orientation bias.\n",
      "Read in summary file ... Done.\n",
      "Now doing the graphical display.\n",
      "xsize = 380, ysize = 460, xbase = 35, ybase = 115\n",
      "[ACT].[ACG][AT]GGA.[AGT]\t0\n",
      "GA[AG]AA[AG]C[CG].\t1\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t3\n",
      "..[ACT].G[ACG]T[ACT][AG]\t2\n",
      ".[CGT]C[AG]A[ACT][CT]..\t4\n",
      "[AC][CT]CCCC[CG][CT].\t5\n",
      ".GC[AG]A[GT]CC.\t6\n",
      "Creating result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.fullmimatrix.eps ...Done.\n",
      "Creating PDF result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.fullmimatrix.pdf ... Done.\n",
      "Read in summary file ... Done.\n",
      "Now doing the graphical display.\n",
      "xsize = 350, ysize = 405, xbase = 75, ybase = 130\n",
      "result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.fullmimatrix.html[ACT].[ACG][AT]GGA.[AGT]\t0\n",
      "GA[AG]AA[AG]C[CG].\t1\n",
      "[ACG][CT][AGT][CGT][AC]T[AGT][CT].\t3\n",
      "..[ACT].G[ACG]T[ACT][AG]\t2\n",
      ".[CGT]C[AG]A[ACT][CT]..\t4\n",
      "[AC][CT]CCCC[CG][CT].\t5\n",
      ".GC[AG]A[GT]CC.\t6\n",
      "RNA, Step 10: motif report.\n",
      "Tue Mar 23 22:16:25 UTC 2021\n",
      "renamed 'radar/result.sig_fire.txt_FIRE/' -> 'fire/discovery_FIRE'\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "free(): invalid next size (fast)\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "Segmentation fault (core dumped)\n",
      "Segmentation fault (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "corrupted size vs. prev_size\n",
      "Aborted (core dumped)\n",
      "Can't use 'defined(@array)' (Maybe you should just omit the defined()?) at /flash/bin/FIRE-1.1/SCRIPTS/draw_position_histogram.pl line 134.\n",
      "Can't locate Hypergeom.pm in @INC (you may need to install the Hypergeom module) (@INC contains: /nvme/bins/FIRE-1.1/SCRIPTS/../modules/lib/perl5/site_perl/5.12.4/darwin-thread-multi-2level/nvme/bins/FIRE-1.1/SCRIPTS/../modules/lib/perl5/site_perl/5.10.0/darwin-thread-multi-2level/Hypergeom.pm /nvme/bins/FIRE-1.1/SCRIPTS/../modules/lib/perl5/site_perl/5.18.2/darwin-thread-multi-2level/Hypergeom.pm /nvme/bins/FIRE-1.1/SCRIPTS/../modules/lib/perl5/site_perl/5.22.0/x86_64-linux/Hypergeom.pm /flash/bin/FIRE-1.1/SCRIPTS /rumi/shams/abe/perl5/lib/perl5/5.26.2/x86_64-linux-thread-multi /rumi/shams/abe/perl5/lib/perl5/5.26.2 /rumi/shams/abe/perl5/lib/perl5/x86_64-linux-thread-multi /rumi/shams/abe/perl5/lib/perl5 /rumi/shams/abe/anaconda3/envs/perl/lib/site_perl/5.26.2/x86_64-linux-thread-multi /rumi/shams/abe/anaconda3/envs/perl/lib/site_perl/5.26.2 /rumi/shams/abe/anaconda3/envs/perl/lib/5.26.2/x86_64-linux-thread-multi /rumi/shams/abe/anaconda3/envs/perl/lib/5.26.2 .) at /flash/bin/FIRE-1.1/SCRIPTS/GroupEnrichment.pm line 24.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/over.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/under.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/C0.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/C1.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/motif.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/location.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/mi.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/z.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/robustness.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/ori.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/cons.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/seed.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/name.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/parray.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/over.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/under.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/C0.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/C1.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/motif.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/location.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/mi.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/z.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/robustness.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/ori.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/cons.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/seed.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/name.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/parray.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/3UTR_pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/3UTR_pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/3UTR_pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/3UTR_pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/3UTR_pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/3UTR_pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n",
      "convert: profile 'icc': 'RGB ': RGB color space not permitted on grayscale PNG `result.sig_fire.txt_FIRE/RNA/result.sig_fire.txt.summary_OUT/3UTR_pos.png' @ warning/png.c/MagickPNGWarningHandler/1748.\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd radar/\n",
    "# run FIRE --doskipdiscovery=0\n",
    "perl $FIREDIR/fire.pl \\\n",
    "    --expfile=result.sig_fire.txt \\\n",
    "    --exptype=discrete \\\n",
    "    --fastafile_rna=result.sig_fire.fa \\\n",
    "    --nodups=1 --dodna=0 --dodnarna=0 \\\n",
    "    --species=human --oribiasonly=0\n",
    "\n",
    "cd ../\n",
    "rm -r fire/discovery_FIRE\n",
    "\n",
    "mkdir -p fire \n",
    "mv -v radar/result.sig_fire.txt_FIRE/ fire/discovery_FIRE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "renamed 'fire/discovery_FIRE/RNA/result.sig_fire.txt.densities.pdf' -> 'fire/discovery_FIRE.densities.pdf'\n",
      "renamed 'fire/discovery_FIRE/RNA/result.sig_fire.txt.fullmimatrix.pdf' -> 'fire/discovery_FIRE.fullmimatrix.pdf'\n",
      "renamed 'fire/discovery_FIRE/RNA/result.sig_fire.txt.summary.pdf' -> 'fire/discovery_FIRE.summary.pdf'\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "for f in  fire/*/RNA/*.pdf; do \n",
    "    o=${f/\\/RNA\\/result.sig_fire.txt/}\n",
    "    mv -v $f $o; \n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Using my `perl` env."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "fire/discovery_FIRE.densities.pdf > fire/discovery_FIRE.densities.png\n",
      "done!\n",
      "fire/discovery_FIRE.fullmimatrix.pdf > fire/discovery_FIRE.fullmimatrix.png\n",
      "done!\n",
      "fire/discovery_FIRE.summary.pdf > fire/discovery_FIRE.summary.png\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "!for f in fire/*pdf; do bash /rumi/shams/abe/Workflows/my_scripts/pdf2png.sh $f; done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### DRACH & RGAC motifs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=fire/known_motifs_FIRE.png style=\"width:1100px\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Discover over/under represented motifs "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<img src=fire/discovery_FIRE.summary.png style=\"width:1100px\">"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# iPAGE\n",
    "https://github.com/goodarzilab/PAGE"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "ipage_cm='~/GitHub/Abe/my_scripts/ipage_human_ensembl.sh'\n",
    "nohup ls hl60_delta_mtyl.txt | parallel -j18 -k bash $ipage_cm {} &> ipage.out &"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:perl]",
   "language": "python",
   "name": "conda-env-perl-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
