**A multiomics approach reveals RNA dynamics promote cellular sensitivity to DNA hypomethylation**

This repository contains all the scripts necessary for producing the results of the manuscript, for which a preprint is available at:

[![DOI:10.1101/2022.12.14.518457](http://img.shields.io/badge/DOI-10.1101/2022.12.14.518457-B31B1B.svg)](https://www.biorxiv.org/content/early/2022/12/14/2022.12.14.518457)

Raw data is also available at:

[![GEO:GSE222886](https://img.shields.io/badge/GEO-GSE222886-green.svg)](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE222886)

___
List of reproducible pipelines and other repositories developed as part of this project:
- Differential gene expression analysis: The Salmon-tximport-DESeq2 pipeline
- Differential RNA stability analysis: The STAR-featureCounts-REMBRANDTS-limma pipeline
- Integrated methods for meRIP-seq data analysis – https://github.com/abearab/imRIP
- Pathway analysis using PAGE algorithm and downstream analysis – https://github.com/abearab/pager
- Scripts for mapping NGS reads to HERVs – https://github.com/abearab/HERVs

___
This repo has been developed in Gilbert at UCSF and Arc Institute by [Abolfazl Arab](https://github.com/abearab), in collaboration with Goodarzi lab at UCSF. This also incldue some original/revised codes from  Alex Y. Ge,  Hosseinali Asgharian, and Hani Goodarzi.
