{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "a2f2383c-494a-4400-8996-27c0cc71d1ef",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from functools import reduce\n",
    "import pathlib\n",
    "import pandas as pd \n",
    "import sys\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from adjustText import adjust_text # https://stackoverflow.com/questions/34693991/repel-annotations-in-matplotlib\n",
    "\n",
    "# sys.path.append('/rumi/shams/abe/CanDI/')\n",
    "sys.path.append('/data_gilbert2/backups/aarab/CanDI')\n",
    "\n",
    "from CanDI import candi as can"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "403a3533-0b9b-4195-810e-9356e14522ac",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "sys.path.append('..')\n",
    "\n",
    "from scripts.candi_functions import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "80bdda1a-f116-4439-bf7b-ede104cc2c1b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from matplotlib_inline.backend_inline import set_matplotlib_formats\n",
    "\n",
    "set_matplotlib_formats('svg')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "385db20f-d4b1-4131-acd3-4cfa24a666fe",
   "metadata": {
    "tags": []
   },
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "8b0af013-a86b-408c-a8ac-2fc1aeb5fec2",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from venn import venn "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "7117caaf-e2f9-472d-933f-dbaa0c667fdb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "wd = '/data_gilbert/home/aarab/AML/Decitabine-treatment'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "94fce0c0-d4a0-4d85-b741-0459370df9d6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from scripts.util import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "226e73ab",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "54\n"
     ]
    }
   ],
   "source": [
    "aml = can.Cancer(\"Leukemia\", subtype='AML')\n",
    "\n",
    "# Number of Leukemia lines\n",
    "print(len(aml.depmap_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1f0cbfaa-ca44-40e4-831a-9792b7955da6",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "our_aml_panel = can.data.cell_lines[can.data.cell_lines.cell_line_name.isin([\n",
    "    \"SKM-1\",\"MOLM-13\",'HL-60',\n",
    "    'MOLM-14','OCI-AML3','KG-1'\n",
    "])].cell_line_name.to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "531e436e-824f-4497-b3ca-fef00ba176f5",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Subtype</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cell_line_name</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HL-60</th>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M3 (Promyelo...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>OCI-AML3</th>\n",
       "      <td>Acute Myelogenous Leukemia (AML), M4 (Myelomon...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOLM-13</th>\n",
       "      <td>Acute Myelogenous Leukemia (AML)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SKM-1</th>\n",
       "      <td>Acute Myelogenous Leukemia (AML)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>KG-1</th>\n",
       "      <td>Acute Myelogenous Leukemia (AML)</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MOLM-14</th>\n",
       "      <td>Acute Myelogenous Leukemia (AML)</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                          Subtype\n",
       "cell_line_name                                                   \n",
       "HL-60           Acute Myelogenous Leukemia (AML), M3 (Promyelo...\n",
       "OCI-AML3        Acute Myelogenous Leukemia (AML), M4 (Myelomon...\n",
       "MOLM-13                          Acute Myelogenous Leukemia (AML)\n",
       "SKM-1                            Acute Myelogenous Leukemia (AML)\n",
       "KG-1                             Acute Myelogenous Leukemia (AML)\n",
       "MOLM-14                          Acute Myelogenous Leukemia (AML)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# lu.depmap_ids\n",
    "can.data.cell_lines.loc[\n",
    "    can.data.cell_lines.cell_line_name.isin(our_aml_panel.values()),:\n",
    "].set_index('cell_line_name')[['Subtype']]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "80423241-c747-413e-b582-cfa5292655bb",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6a24b7e0-2cad-4058-99b5-94c96f898021",
   "metadata": {},
   "outputs": [],
   "source": [
    "not_aml = can.CellLineCluster(list(set(can.data.cell_lines.index) - set(aml.depmap_ids)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "6ad7fd87-4a52-49f2-8827-486c83de538e",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "candi = can.CellLineCluster(can.data.cell_lines.index.to_list())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b21c3c-cd90-49e8-bd9a-d81811d7dd43",
   "metadata": {},
   "source": [
    "## expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "597194d1-d9fa-4eb2-a9fd-6bccf082cc04",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e4946a2b-01a1-4009-8f83-0d896cd5bb90",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "from pydeseq2.preprocessing import deseq2_norm, deseq2_norm_fit, deseq2_norm_transform"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "0054f960-136b-497b-8aa9-a82ed12420f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_panel = [\n",
    "    # 'TP53',\n",
    "    'FTO','ALKBH5',\n",
    "    'DCK', 'SLC29A1', 'BCL2', 'MCL1',\n",
    "    'DCPS', 'DCP2', 'DCP1A', 'SCAF8',\n",
    "    'METTL3', 'YTHDF2', 'YTHDC1', 'ZC3H13', 'VIRMA',#'KIAA1429', \n",
    "    'CBLL1',\n",
    "    'PTEN', 'SQLE', 'INTS5', 'DHODH', 'MYBBP1A',\n",
    "    'INO80D', 'PMPCA', 'ZNF777', 'SLC7A6', 'RNF126', 'RBM14-RBM4',\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "9a65823c-aa4e-4013-bf02-907f04b351b9",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "rnaseq_reads has not been loaded. Do you want to load, y/n?>  y\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Load Complete\n"
     ]
    }
   ],
   "source": [
    "aml_exp = aml.rnaseq_reads\n",
    "\n",
    "aml_exp_norm = deseq2_norm(np.log1p(aml_exp.T))[0].T"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "c748f842-d229-42f3-963d-8921890ddea7",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "12978000.0"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aml_exp.max().max()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "edc46d3e-7873-48f3-8b3b-df13c6abb138",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "15.358628608876638"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "aml_exp_norm.max().max()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "eb4b881b-9493-4efe-b4d7-27aab55817b3",
   "metadata": {},
   "source": [
    "### DHODH gene effect"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "16f5d313-11ad-4447-baf4-f0dc7df795ba",
   "metadata": {},
   "source": [
    "https://forum.depmap.org/t/depmap-genetic-dependencies-faq/131\n",
    "> The “gene effect” file contains the corrected CERES scores, which measure the effect size of knocking out a gene, normalized against the distributions of non-essential and pan-essential genes. The probabilities assess, given a gene score, how likely to be a member of the non-essential distribution or the common essential distribution in that cell line. The key difference between using a fixed threshold on CERES score and a threshold on the probabilities is that the probabilities take into account the screening quality, which varies from line to line."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "id": "6ca1f7aa-f239-4f41-85d5-682806226162",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "gene_name = 'DHODH'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "8fa760c1-4486-4fc3-a6be-ff38f6f58ba7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dep_df = pd.DataFrame({'dependency':candi.effect_of(gene_name)})\n",
    "dep_df['primary_disease'] = can.data.cell_lines['primary_disease'][dep_df.index.to_list()]\n",
    "dep_df=dep_df[~dep_df.primary_disease.eq('Non-Cancerous')]\n",
    "\n",
    "dep_df['primary_disease'] = pd.Categorical(\n",
    "    dep_df['primary_disease'], categories=dep_df.groupby('primary_disease')['dependency'].median().sort_values(ascending=True).index\n",
    ")\n",
    "\n",
    "dep_df['name'] = [our_aml_panel[cl] if cl in our_aml_panel.keys() else 'other' for cl in dep_df.index]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4afc321c-61f2-4847-afc3-375654635fff",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>dependency</th>\n",
       "      <th>primary_disease</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ACH-000336</th>\n",
       "      <td>-0.492481</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>OCI-AML3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-000362</th>\n",
       "      <td>-0.746526</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>MOLM-13</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ACH-001574</th>\n",
       "      <td>-0.867626</td>\n",
       "      <td>Leukemia</td>\n",
       "      <td>MOLM-14</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            dependency primary_disease      name\n",
       "ACH-000336   -0.492481        Leukemia  OCI-AML3\n",
       "ACH-000362   -0.746526        Leukemia   MOLM-13\n",
       "ACH-001574   -0.867626        Leukemia   MOLM-14"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dep_df.query('name != \"other\"')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "a8353820-299e-4fbc-bb39-a7044ea4dac4",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "dep_df['Leukemia'] = dep_df.primary_disease == 'Leukemia'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "e2c3ee82-370d-42aa-9399-a0084afbbe16",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data_gilbert/home/aarab/anaconda3/envs/mamba/envs/candi-kg/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/data_gilbert/home/aarab/anaconda3/envs/mamba/envs/candi-kg/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n",
      "/data_gilbert/home/aarab/anaconda3/envs/mamba/envs/candi-kg/lib/python3.9/site-packages/seaborn/_oldcore.py:1498: FutureWarning: is_categorical_dtype is deprecated and will be removed in a future version. Use isinstance(dtype, CategoricalDtype) instead\n",
      "  if pd.api.types.is_categorical_dtype(vector):\n"
     ]
    },
    {
     "data": {
      "image/svg+xml": [
       "<?xml version=\"1.0\" encoding=\"utf-8\" standalone=\"no\"?>\n",
       "<!DOCTYPE svg PUBLIC \"-//W3C//DTD SVG 1.1//EN\"\n",
       "  \"http://www.w3.org/Graphics/SVG/1.1/DTD/svg11.dtd\">\n",
       "<svg xmlns:xlink=\"http://www.w3.org/1999/xlink\" width=\"170.960938pt\" height=\"351.993622pt\" viewBox=\"0 0 170.960938 351.993622\" xmlns=\"http://www.w3.org/2000/svg\" version=\"1.1\">\n",
       " <metadata>\n",
       "  <rdf:RDF xmlns:dc=\"http://purl.org/dc/elements/1.1/\" xmlns:cc=\"http://creativecommons.org/ns#\" xmlns:rdf=\"http://www.w3.org/1999/02/22-rdf-syntax-ns#\">\n",
       "   <cc:Work>\n",
       "    <dc:type rdf:resource=\"http://purl.org/dc/dcmitype/StillImage\"/>\n",
       "    <dc:date>2024-02-20T00:30:12.984913</dc:date>\n",
       "    <dc:format>image/svg+xml</dc:format>\n",
       "    <dc:creator>\n",
       "     <cc:Agent>\n",
       "      <dc:title>Matplotlib v3.8.0, https://matplotlib.org/</dc:title>\n",
       "     </cc:Agent>\n",
       "    </dc:creator>\n",
       "   </cc:Work>\n",
       "  </rdf:RDF>\n",
       " </metadata>\n",
       " <defs>\n",
       "  <style type=\"text/css\">*{stroke-linejoin: round; stroke-linecap: butt}</style>\n",
       " </defs>\n",
       " <g id=\"figure_1\">\n",
       "  <g id=\"patch_1\">\n",
       "   <path d=\"M 0 351.993622 \n",
       "L 170.960938 351.993622 \n",
       "L 170.960938 0 \n",
       "L 0 0 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "  </g>\n",
       "  <g id=\"axes_1\">\n",
       "   <g id=\"patch_2\">\n",
       "    <path d=\"M 52.160938 299.518125 \n",
       "L 163.760938 299.518125 \n",
       "L 163.760938 22.318125 \n",
       "L 52.160938 22.318125 \n",
       "z\n",
       "\" style=\"fill: #ffffff\"/>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_1\">\n",
       "    <defs>\n",
       "     <path id=\"m8108afdd3a\" d=\"M 80.076077 -97.163035 \n",
       "L 80.045798 -97.163035 \n",
       "L 80.034442 -98.762195 \n",
       "L 80.017836 -100.361355 \n",
       "L 79.995633 -101.960515 \n",
       "L 79.96854 -103.559675 \n",
       "L 79.938445 -105.158835 \n",
       "L 79.908155 -106.757995 \n",
       "L 79.880794 -108.357156 \n",
       "L 79.859128 -109.956316 \n",
       "L 79.845058 -111.555476 \n",
       "L 79.839381 -113.154636 \n",
       "L 79.841683 -114.753796 \n",
       "L 79.850196 -116.352956 \n",
       "L 79.861633 -117.952116 \n",
       "L 79.871254 -119.551276 \n",
       "L 79.873433 -121.150436 \n",
       "L 79.862727 -122.749596 \n",
       "L 79.835038 -124.348756 \n",
       "L 79.788254 -125.947916 \n",
       "L 79.722056 -127.547076 \n",
       "L 79.637135 -129.146236 \n",
       "L 79.534535 -130.745396 \n",
       "L 79.415856 -132.344556 \n",
       "L 79.284326 -133.943716 \n",
       "L 79.145904 -135.542877 \n",
       "L 79.009172 -137.142037 \n",
       "L 78.883358 -138.741197 \n",
       "L 78.775116 -140.340357 \n",
       "L 78.685763 -141.939517 \n",
       "L 78.610641 -143.538677 \n",
       "L 78.54105 -145.137837 \n",
       "L 78.467598 -146.736997 \n",
       "L 78.382983 -148.336157 \n",
       "L 78.282767 -149.935317 \n",
       "L 78.164151 -151.534477 \n",
       "L 78.024114 -153.133637 \n",
       "L 77.858678 -154.732797 \n",
       "L 77.664254 -156.331957 \n",
       "L 77.440619 -157.931117 \n",
       "L 77.193857 -159.530277 \n",
       "L 76.937285 -161.129437 \n",
       "L 76.689349 -162.728598 \n",
       "L 76.469005 -164.327758 \n",
       "L 76.290466 -165.926918 \n",
       "L 76.159431 -167.526078 \n",
       "L 76.071947 -169.125238 \n",
       "L 76.015668 -170.724398 \n",
       "L 75.972289 -172.323558 \n",
       "L 75.920015 -173.922718 \n",
       "L 75.835652 -175.521878 \n",
       "L 75.69668 -177.121038 \n",
       "L 75.483809 -178.720198 \n",
       "L 75.183988 -180.319358 \n",
       "L 74.793065 -181.918518 \n",
       "L 74.316879 -183.517678 \n",
       "L 73.769808 -185.116838 \n",
       "L 73.170696 -186.715998 \n",
       "L 72.537124 -188.315159 \n",
       "L 71.879772 -189.914319 \n",
       "L 71.1988 -191.513479 \n",
       "L 70.48371 -193.112639 \n",
       "L 69.717114 -194.711799 \n",
       "L 68.881332 -196.310959 \n",
       "L 67.965403 -197.910119 \n",
       "L 66.969625 -199.509279 \n",
       "L 65.90589 -201.108439 \n",
       "L 64.794444 -202.707599 \n",
       "L 63.659742 -204.306759 \n",
       "L 62.52834 -205.905919 \n",
       "L 61.429721 -207.505079 \n",
       "L 60.398441 -209.104239 \n",
       "L 59.474947 -210.703399 \n",
       "L 58.703758 -212.302559 \n",
       "L 58.130023 -213.901719 \n",
       "L 57.796394 -215.50088 \n",
       "L 57.740938 -217.10004 \n",
       "L 57.994685 -218.6992 \n",
       "L 58.576843 -220.29836 \n",
       "L 59.487556 -221.89752 \n",
       "L 60.701088 -223.49668 \n",
       "L 62.163759 -225.09584 \n",
       "L 63.799486 -226.695 \n",
       "L 65.522077 -228.29416 \n",
       "L 67.250034 -229.89332 \n",
       "L 68.918583 -231.49248 \n",
       "L 70.485433 -233.09164 \n",
       "L 71.929808 -234.6908 \n",
       "L 73.246805 -236.28996 \n",
       "L 74.439972 -237.88912 \n",
       "L 75.51458 -239.48828 \n",
       "L 76.473121 -241.08744 \n",
       "L 77.313805 -242.686601 \n",
       "L 78.032005 -244.285761 \n",
       "L 78.62378 -245.884921 \n",
       "L 79.08979 -247.484081 \n",
       "L 79.437819 -249.083241 \n",
       "L 79.682826 -250.682401 \n",
       "L 79.844673 -252.281561 \n",
       "L 79.944663 -253.880721 \n",
       "L 80.002302 -255.479881 \n",
       "L 80.119573 -255.479881 \n",
       "L 80.119573 -255.479881 \n",
       "L 80.177212 -253.880721 \n",
       "L 80.277202 -252.281561 \n",
       "L 80.439049 -250.682401 \n",
       "L 80.684056 -249.083241 \n",
       "L 81.032085 -247.484081 \n",
       "L 81.498095 -245.884921 \n",
       "L 82.08987 -244.285761 \n",
       "L 82.80807 -242.686601 \n",
       "L 83.648754 -241.08744 \n",
       "L 84.607295 -239.48828 \n",
       "L 85.681903 -237.88912 \n",
       "L 86.87507 -236.28996 \n",
       "L 88.192067 -234.6908 \n",
       "L 89.636442 -233.09164 \n",
       "L 91.203292 -231.49248 \n",
       "L 92.871841 -229.89332 \n",
       "L 94.599798 -228.29416 \n",
       "L 96.322389 -226.695 \n",
       "L 97.958116 -225.09584 \n",
       "L 99.420787 -223.49668 \n",
       "L 100.634319 -221.89752 \n",
       "L 101.545032 -220.29836 \n",
       "L 102.12719 -218.6992 \n",
       "L 102.380938 -217.10004 \n",
       "L 102.325481 -215.50088 \n",
       "L 101.991852 -213.901719 \n",
       "L 101.418117 -212.302559 \n",
       "L 100.646928 -210.703399 \n",
       "L 99.723434 -209.104239 \n",
       "L 98.692154 -207.505079 \n",
       "L 97.593535 -205.905919 \n",
       "L 96.462133 -204.306759 \n",
       "L 95.327431 -202.707599 \n",
       "L 94.215985 -201.108439 \n",
       "L 93.15225 -199.509279 \n",
       "L 92.156472 -197.910119 \n",
       "L 91.240543 -196.310959 \n",
       "L 90.404761 -194.711799 \n",
       "L 89.638165 -193.112639 \n",
       "L 88.923075 -191.513479 \n",
       "L 88.242103 -189.914319 \n",
       "L 87.584751 -188.315159 \n",
       "L 86.951179 -186.715998 \n",
       "L 86.352067 -185.116838 \n",
       "L 85.804996 -183.517678 \n",
       "L 85.32881 -181.918518 \n",
       "L 84.937887 -180.319358 \n",
       "L 84.638066 -178.720198 \n",
       "L 84.425195 -177.121038 \n",
       "L 84.286223 -175.521878 \n",
       "L 84.20186 -173.922718 \n",
       "L 84.149586 -172.323558 \n",
       "L 84.106207 -170.724398 \n",
       "L 84.049928 -169.125238 \n",
       "L 83.962444 -167.526078 \n",
       "L 83.831409 -165.926918 \n",
       "L 83.65287 -164.327758 \n",
       "L 83.432526 -162.728598 \n",
       "L 83.18459 -161.129437 \n",
       "L 82.928018 -159.530277 \n",
       "L 82.681256 -157.931117 \n",
       "L 82.457621 -156.331957 \n",
       "L 82.263197 -154.732797 \n",
       "L 82.097761 -153.133637 \n",
       "L 81.957724 -151.534477 \n",
       "L 81.839108 -149.935317 \n",
       "L 81.738892 -148.336157 \n",
       "L 81.654277 -146.736997 \n",
       "L 81.580825 -145.137837 \n",
       "L 81.511234 -143.538677 \n",
       "L 81.436112 -141.939517 \n",
       "L 81.346759 -140.340357 \n",
       "L 81.238517 -138.741197 \n",
       "L 81.112703 -137.142037 \n",
       "L 80.975971 -135.542877 \n",
       "L 80.837549 -133.943716 \n",
       "L 80.706019 -132.344556 \n",
       "L 80.58734 -130.745396 \n",
       "L 80.48474 -129.146236 \n",
       "L 80.399819 -127.547076 \n",
       "L 80.333621 -125.947916 \n",
       "L 80.286837 -124.348756 \n",
       "L 80.259148 -122.749596 \n",
       "L 80.248442 -121.150436 \n",
       "L 80.250621 -119.551276 \n",
       "L 80.260242 -117.952116 \n",
       "L 80.271679 -116.352956 \n",
       "L 80.280192 -114.753796 \n",
       "L 80.282494 -113.154636 \n",
       "L 80.276817 -111.555476 \n",
       "L 80.262747 -109.956316 \n",
       "L 80.241081 -108.357156 \n",
       "L 80.21372 -106.757995 \n",
       "L 80.18343 -105.158835 \n",
       "L 80.153335 -103.559675 \n",
       "L 80.126242 -101.960515 \n",
       "L 80.104039 -100.361355 \n",
       "L 80.087433 -98.762195 \n",
       "L 80.076077 -97.163035 \n",
       "z\n",
       "\" style=\"stroke: #7a7a7a; stroke-width: 1.5\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#pe3887c7f71)\">\n",
       "     <use xlink:href=\"#m8108afdd3a\" x=\"0\" y=\"351.993622\" style=\"fill: #cccccc; stroke: #7a7a7a; stroke-width: 1.5\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"PolyCollection_2\">\n",
       "    <defs>\n",
       "     <path id=\"m66211c402e\" d=\"M 135.982679 -68.922262 \n",
       "L 135.739196 -68.922262 \n",
       "L 135.693642 -70.88346 \n",
       "L 135.635962 -72.844657 \n",
       "L 135.564762 -74.805854 \n",
       "L 135.479069 -76.767052 \n",
       "L 135.378477 -78.728249 \n",
       "L 135.263244 -80.689447 \n",
       "L 135.134308 -82.650644 \n",
       "L 134.993197 -84.611841 \n",
       "L 134.84183 -86.573039 \n",
       "L 134.682213 -88.534236 \n",
       "L 134.516079 -90.495434 \n",
       "L 134.344514 -92.456631 \n",
       "L 134.167637 -94.417829 \n",
       "L 133.984415 -96.379026 \n",
       "L 133.792646 -98.340223 \n",
       "L 133.589159 -100.301421 \n",
       "L 133.370214 -102.262618 \n",
       "L 133.132054 -104.223816 \n",
       "L 132.871543 -106.185013 \n",
       "L 132.586785 -108.146211 \n",
       "L 132.277613 -110.107408 \n",
       "L 131.945864 -112.068605 \n",
       "L 131.595389 -114.029803 \n",
       "L 131.231757 -115.991 \n",
       "L 130.861717 -117.952198 \n",
       "L 130.492465 -119.913395 \n",
       "L 130.130844 -121.874592 \n",
       "L 129.78257 -123.83579 \n",
       "L 129.45161 -125.796987 \n",
       "L 129.139774 -127.758185 \n",
       "L 128.846581 -129.719382 \n",
       "L 128.569386 -131.68058 \n",
       "L 128.30375 -133.641777 \n",
       "L 128.043975 -135.602974 \n",
       "L 127.783767 -137.564172 \n",
       "L 127.516911 -139.525369 \n",
       "L 127.237927 -141.486567 \n",
       "L 126.942632 -143.447764 \n",
       "L 126.628578 -145.408962 \n",
       "L 126.295335 -147.370159 \n",
       "L 125.944618 -149.331356 \n",
       "L 125.580261 -151.292554 \n",
       "L 125.208057 -153.253751 \n",
       "L 124.835505 -155.214949 \n",
       "L 124.471473 -157.176146 \n",
       "L 124.125832 -159.137343 \n",
       "L 123.809052 -161.098541 \n",
       "L 123.531785 -163.059738 \n",
       "L 123.304424 -165.020936 \n",
       "L 123.136636 -166.982133 \n",
       "L 123.036875 -168.943331 \n",
       "L 123.011896 -170.904528 \n",
       "L 123.066316 -172.865725 \n",
       "L 123.202272 -174.826923 \n",
       "L 123.419246 -176.78812 \n",
       "L 123.714103 -178.749318 \n",
       "L 124.081361 -180.710515 \n",
       "L 124.513671 -182.671713 \n",
       "L 125.002451 -184.63291 \n",
       "L 125.538567 -186.594107 \n",
       "L 126.112959 -188.555305 \n",
       "L 126.717109 -190.516502 \n",
       "L 127.343276 -192.4777 \n",
       "L 127.98449 -194.438897 \n",
       "L 128.634337 -196.400094 \n",
       "L 129.286609 -198.361292 \n",
       "L 129.934937 -200.322489 \n",
       "L 130.572508 -202.283687 \n",
       "L 131.191952 -204.244884 \n",
       "L 131.78545 -206.206082 \n",
       "L 132.345047 -208.167279 \n",
       "L 132.863139 -210.128476 \n",
       "L 133.333034 -212.089674 \n",
       "L 133.74951 -214.050871 \n",
       "L 134.109278 -216.012069 \n",
       "L 134.411273 -217.973266 \n",
       "L 134.656752 -219.934464 \n",
       "L 134.849187 -221.895661 \n",
       "L 134.993961 -223.856858 \n",
       "L 135.09793 -225.818056 \n",
       "L 135.168888 -227.779253 \n",
       "L 135.21499 -229.740451 \n",
       "L 135.244196 -231.701648 \n",
       "L 135.263765 -233.662846 \n",
       "L 135.279863 -235.624043 \n",
       "L 135.297284 -237.58524 \n",
       "L 135.319317 -239.546438 \n",
       "L 135.347756 -241.507635 \n",
       "L 135.383034 -243.468833 \n",
       "L 135.424467 -245.43003 \n",
       "L 135.470558 -247.391227 \n",
       "L 135.519333 -249.352425 \n",
       "L 135.568658 -251.313622 \n",
       "L 135.616518 -253.27482 \n",
       "L 135.661217 -255.236017 \n",
       "L 135.701501 -257.197215 \n",
       "L 135.736605 -259.158412 \n",
       "L 135.766231 -261.119609 \n",
       "L 135.790476 -263.080807 \n",
       "L 135.931399 -263.080807 \n",
       "L 135.931399 -263.080807 \n",
       "L 135.955644 -261.119609 \n",
       "L 135.98527 -259.158412 \n",
       "L 136.020374 -257.197215 \n",
       "L 136.060658 -255.236017 \n",
       "L 136.105357 -253.27482 \n",
       "L 136.153217 -251.313622 \n",
       "L 136.202542 -249.352425 \n",
       "L 136.251317 -247.391227 \n",
       "L 136.297408 -245.43003 \n",
       "L 136.338841 -243.468833 \n",
       "L 136.374119 -241.507635 \n",
       "L 136.402558 -239.546438 \n",
       "L 136.424591 -237.58524 \n",
       "L 136.442012 -235.624043 \n",
       "L 136.45811 -233.662846 \n",
       "L 136.477679 -231.701648 \n",
       "L 136.506885 -229.740451 \n",
       "L 136.552987 -227.779253 \n",
       "L 136.623945 -225.818056 \n",
       "L 136.727914 -223.856858 \n",
       "L 136.872688 -221.895661 \n",
       "L 137.065123 -219.934464 \n",
       "L 137.310602 -217.973266 \n",
       "L 137.612597 -216.012069 \n",
       "L 137.972365 -214.050871 \n",
       "L 138.388841 -212.089674 \n",
       "L 138.858736 -210.128476 \n",
       "L 139.376828 -208.167279 \n",
       "L 139.936425 -206.206082 \n",
       "L 140.529923 -204.244884 \n",
       "L 141.149367 -202.283687 \n",
       "L 141.786938 -200.322489 \n",
       "L 142.435266 -198.361292 \n",
       "L 143.087538 -196.400094 \n",
       "L 143.737385 -194.438897 \n",
       "L 144.378599 -192.4777 \n",
       "L 145.004766 -190.516502 \n",
       "L 145.608916 -188.555305 \n",
       "L 146.183308 -186.594107 \n",
       "L 146.719424 -184.63291 \n",
       "L 147.208204 -182.671713 \n",
       "L 147.640514 -180.710515 \n",
       "L 148.007772 -178.749318 \n",
       "L 148.302629 -176.78812 \n",
       "L 148.519603 -174.826923 \n",
       "L 148.655559 -172.865725 \n",
       "L 148.709979 -170.904528 \n",
       "L 148.685 -168.943331 \n",
       "L 148.585239 -166.982133 \n",
       "L 148.417451 -165.020936 \n",
       "L 148.19009 -163.059738 \n",
       "L 147.912823 -161.098541 \n",
       "L 147.596043 -159.137343 \n",
       "L 147.250402 -157.176146 \n",
       "L 146.88637 -155.214949 \n",
       "L 146.513818 -153.253751 \n",
       "L 146.141614 -151.292554 \n",
       "L 145.777257 -149.331356 \n",
       "L 145.42654 -147.370159 \n",
       "L 145.093297 -145.408962 \n",
       "L 144.779243 -143.447764 \n",
       "L 144.483948 -141.486567 \n",
       "L 144.204964 -139.525369 \n",
       "L 143.938108 -137.564172 \n",
       "L 143.6779 -135.602974 \n",
       "L 143.418125 -133.641777 \n",
       "L 143.152489 -131.68058 \n",
       "L 142.875294 -129.719382 \n",
       "L 142.582101 -127.758185 \n",
       "L 142.270265 -125.796987 \n",
       "L 141.939305 -123.83579 \n",
       "L 141.591031 -121.874592 \n",
       "L 141.22941 -119.913395 \n",
       "L 140.860158 -117.952198 \n",
       "L 140.490118 -115.991 \n",
       "L 140.126486 -114.029803 \n",
       "L 139.776011 -112.068605 \n",
       "L 139.444262 -110.107408 \n",
       "L 139.13509 -108.146211 \n",
       "L 138.850332 -106.185013 \n",
       "L 138.589821 -104.223816 \n",
       "L 138.351661 -102.262618 \n",
       "L 138.132716 -100.301421 \n",
       "L 137.929229 -98.340223 \n",
       "L 137.73746 -96.379026 \n",
       "L 137.554238 -94.417829 \n",
       "L 137.377361 -92.456631 \n",
       "L 137.205796 -90.495434 \n",
       "L 137.039662 -88.534236 \n",
       "L 136.880045 -86.573039 \n",
       "L 136.728678 -84.611841 \n",
       "L 136.587567 -82.650644 \n",
       "L 136.458631 -80.689447 \n",
       "L 136.343398 -78.728249 \n",
       "L 136.242806 -76.767052 \n",
       "L 136.157113 -74.805854 \n",
       "L 136.085913 -72.844657 \n",
       "L 136.028233 -70.88346 \n",
       "L 135.982679 -68.922262 \n",
       "z\n",
       "\" style=\"stroke: #7a7a7a; stroke-width: 1.5\"/>\n",
       "    </defs>\n",
       "    <g clip-path=\"url(#pe3887c7f71)\">\n",
       "     <use xlink:href=\"#m66211c402e\" x=\"0\" y=\"351.993622\" style=\"fill: #cccccc; stroke: #7a7a7a; stroke-width: 1.5\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_1\">\n",
       "    <g id=\"xtick_1\">\n",
       "     <g id=\"line2d_1\">\n",
       "      <defs>\n",
       "       <path id=\"mb911c42676\" d=\"M 0 0 \n",
       "L 0 3.5 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#mb911c42676\" x=\"80.060938\" y=\"299.518125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_1\">\n",
       "      <!-- False -->\n",
       "      <g transform=\"translate(60.836472 329.644936) rotate(-45) scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-46\" d=\"M 628 4666 \n",
       "L 3309 4666 \n",
       "L 3309 4134 \n",
       "L 1259 4134 \n",
       "L 1259 2759 \n",
       "L 3109 2759 \n",
       "L 3109 2228 \n",
       "L 1259 2228 \n",
       "L 1259 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-61\" d=\"M 2194 1759 \n",
       "Q 1497 1759 1228 1600 \n",
       "Q 959 1441 959 1056 \n",
       "Q 959 750 1161 570 \n",
       "Q 1363 391 1709 391 \n",
       "Q 2188 391 2477 730 \n",
       "Q 2766 1069 2766 1631 \n",
       "L 2766 1759 \n",
       "L 2194 1759 \n",
       "z\n",
       "M 3341 1997 \n",
       "L 3341 0 \n",
       "L 2766 0 \n",
       "L 2766 531 \n",
       "Q 2569 213 2275 61 \n",
       "Q 1981 -91 1556 -91 \n",
       "Q 1019 -91 701 211 \n",
       "Q 384 513 384 1019 \n",
       "Q 384 1609 779 1909 \n",
       "Q 1175 2209 1959 2209 \n",
       "L 2766 2209 \n",
       "L 2766 2266 \n",
       "Q 2766 2663 2505 2880 \n",
       "Q 2244 3097 1772 3097 \n",
       "Q 1472 3097 1187 3025 \n",
       "Q 903 2953 641 2809 \n",
       "L 641 3341 \n",
       "Q 956 3463 1253 3523 \n",
       "Q 1550 3584 1831 3584 \n",
       "Q 2591 3584 2966 3190 \n",
       "Q 3341 2797 3341 1997 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-6c\" d=\"M 603 4863 \n",
       "L 1178 4863 \n",
       "L 1178 0 \n",
       "L 603 0 \n",
       "L 603 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-73\" d=\"M 2834 3397 \n",
       "L 2834 2853 \n",
       "Q 2591 2978 2328 3040 \n",
       "Q 2066 3103 1784 3103 \n",
       "Q 1356 3103 1142 2972 \n",
       "Q 928 2841 928 2578 \n",
       "Q 928 2378 1081 2264 \n",
       "Q 1234 2150 1697 2047 \n",
       "L 1894 2003 \n",
       "Q 2506 1872 2764 1633 \n",
       "Q 3022 1394 3022 966 \n",
       "Q 3022 478 2636 193 \n",
       "Q 2250 -91 1575 -91 \n",
       "Q 1294 -91 989 -36 \n",
       "Q 684 19 347 128 \n",
       "L 347 722 \n",
       "Q 666 556 975 473 \n",
       "Q 1284 391 1588 391 \n",
       "Q 1994 391 2212 530 \n",
       "Q 2431 669 2431 922 \n",
       "Q 2431 1156 2273 1281 \n",
       "Q 2116 1406 1581 1522 \n",
       "L 1381 1569 \n",
       "Q 847 1681 609 1914 \n",
       "Q 372 2147 372 2553 \n",
       "Q 372 3047 722 3315 \n",
       "Q 1072 3584 1716 3584 \n",
       "Q 2034 3584 2315 3537 \n",
       "Q 2597 3491 2834 3397 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-65\" d=\"M 3597 1894 \n",
       "L 3597 1613 \n",
       "L 953 1613 \n",
       "Q 991 1019 1311 708 \n",
       "Q 1631 397 2203 397 \n",
       "Q 2534 397 2845 478 \n",
       "Q 3156 559 3463 722 \n",
       "L 3463 178 \n",
       "Q 3153 47 2828 -22 \n",
       "Q 2503 -91 2169 -91 \n",
       "Q 1331 -91 842 396 \n",
       "Q 353 884 353 1716 \n",
       "Q 353 2575 817 3079 \n",
       "Q 1281 3584 2069 3584 \n",
       "Q 2775 3584 3186 3129 \n",
       "Q 3597 2675 3597 1894 \n",
       "z\n",
       "M 3022 2063 \n",
       "Q 3016 2534 2758 2815 \n",
       "Q 2500 3097 2075 3097 \n",
       "Q 1594 3097 1305 2825 \n",
       "Q 1016 2553 972 2059 \n",
       "L 3022 2063 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-46\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-61\" x=\"48.394531\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-6c\" x=\"109.673828\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-73\" x=\"137.457031\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-65\" x=\"189.556641\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"xtick_2\">\n",
       "     <g id=\"line2d_2\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#mb911c42676\" x=\"135.860938\" y=\"299.518125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_2\">\n",
       "      <!-- True -->\n",
       "      <g transform=\"translate(119.375406 326.906002) rotate(-45) scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-54\" d=\"M -19 4666 \n",
       "L 3928 4666 \n",
       "L 3928 4134 \n",
       "L 2272 4134 \n",
       "L 2272 0 \n",
       "L 1638 0 \n",
       "L 1638 4134 \n",
       "L -19 4134 \n",
       "L -19 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-72\" d=\"M 2631 2963 \n",
       "Q 2534 3019 2420 3045 \n",
       "Q 2306 3072 2169 3072 \n",
       "Q 1681 3072 1420 2755 \n",
       "Q 1159 2438 1159 1844 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1341 3275 1631 3429 \n",
       "Q 1922 3584 2338 3584 \n",
       "Q 2397 3584 2469 3576 \n",
       "Q 2541 3569 2628 3553 \n",
       "L 2631 2963 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-75\" d=\"M 544 1381 \n",
       "L 544 3500 \n",
       "L 1119 3500 \n",
       "L 1119 1403 \n",
       "Q 1119 906 1312 657 \n",
       "Q 1506 409 1894 409 \n",
       "Q 2359 409 2629 706 \n",
       "Q 2900 1003 2900 1516 \n",
       "L 2900 3500 \n",
       "L 3475 3500 \n",
       "L 3475 0 \n",
       "L 2900 0 \n",
       "L 2900 538 \n",
       "Q 2691 219 2414 64 \n",
       "Q 2138 -91 1772 -91 \n",
       "Q 1169 -91 856 284 \n",
       "Q 544 659 544 1381 \n",
       "z\n",
       "M 1991 3584 \n",
       "L 1991 3584 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-54\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-72\" x=\"46.333984\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-75\" x=\"87.447266\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-65\" x=\"150.826172\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_3\">\n",
       "     <!-- Leukemia -->\n",
       "     <g transform=\"translate(83.903125 342.713935) scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-4c\" d=\"M 628 4666 \n",
       "L 1259 4666 \n",
       "L 1259 531 \n",
       "L 3531 531 \n",
       "L 3531 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6b\" d=\"M 581 4863 \n",
       "L 1159 4863 \n",
       "L 1159 1991 \n",
       "L 2875 3500 \n",
       "L 3609 3500 \n",
       "L 1753 1863 \n",
       "L 3688 0 \n",
       "L 2938 0 \n",
       "L 1159 1709 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6d\" d=\"M 3328 2828 \n",
       "Q 3544 3216 3844 3400 \n",
       "Q 4144 3584 4550 3584 \n",
       "Q 5097 3584 5394 3201 \n",
       "Q 5691 2819 5691 2113 \n",
       "L 5691 0 \n",
       "L 5113 0 \n",
       "L 5113 2094 \n",
       "Q 5113 2597 4934 2840 \n",
       "Q 4756 3084 4391 3084 \n",
       "Q 3944 3084 3684 2787 \n",
       "Q 3425 2491 3425 1978 \n",
       "L 3425 0 \n",
       "L 2847 0 \n",
       "L 2847 2094 \n",
       "Q 2847 2600 2669 2842 \n",
       "Q 2491 3084 2119 3084 \n",
       "Q 1678 3084 1418 2786 \n",
       "Q 1159 2488 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1356 3278 1631 3431 \n",
       "Q 1906 3584 2284 3584 \n",
       "Q 2666 3584 2933 3390 \n",
       "Q 3200 3197 3328 2828 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-69\" d=\"M 603 3500 \n",
       "L 1178 3500 \n",
       "L 1178 0 \n",
       "L 603 0 \n",
       "L 603 3500 \n",
       "z\n",
       "M 603 4863 \n",
       "L 1178 4863 \n",
       "L 1178 4134 \n",
       "L 603 4134 \n",
       "L 603 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-4c\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"53.962891\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-75\" x=\"115.486328\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6b\" x=\"178.865234\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"233.150391\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6d\" x=\"294.673828\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-69\" x=\"392.085938\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-61\" x=\"419.869141\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"matplotlib.axis_2\">\n",
       "    <g id=\"ytick_1\">\n",
       "     <g id=\"line2d_3\">\n",
       "      <defs>\n",
       "       <path id=\"m6fb0b76956\" d=\"M 0 0 \n",
       "L -3.5 0 \n",
       "\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </defs>\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6fb0b76956\" x=\"52.160938\" y=\"299.518125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_4\">\n",
       "      <!-- −2.0 -->\n",
       "      <g transform=\"translate(20.878125 303.317344) scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-2212\" d=\"M 678 2272 \n",
       "L 4684 2272 \n",
       "L 4684 1741 \n",
       "L 678 1741 \n",
       "L 678 2272 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-32\" d=\"M 1228 531 \n",
       "L 3431 531 \n",
       "L 3431 0 \n",
       "L 469 0 \n",
       "L 469 531 \n",
       "Q 828 903 1448 1529 \n",
       "Q 2069 2156 2228 2338 \n",
       "Q 2531 2678 2651 2914 \n",
       "Q 2772 3150 2772 3378 \n",
       "Q 2772 3750 2511 3984 \n",
       "Q 2250 4219 1831 4219 \n",
       "Q 1534 4219 1204 4116 \n",
       "Q 875 4013 500 3803 \n",
       "L 500 4441 \n",
       "Q 881 4594 1212 4672 \n",
       "Q 1544 4750 1819 4750 \n",
       "Q 2544 4750 2975 4387 \n",
       "Q 3406 4025 3406 3419 \n",
       "Q 3406 3131 3298 2873 \n",
       "Q 3191 2616 2906 2266 \n",
       "Q 2828 2175 2409 1742 \n",
       "Q 1991 1309 1228 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-2e\" d=\"M 684 794 \n",
       "L 1344 794 \n",
       "L 1344 0 \n",
       "L 684 0 \n",
       "L 684 794 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-30\" d=\"M 2034 4250 \n",
       "Q 1547 4250 1301 3770 \n",
       "Q 1056 3291 1056 2328 \n",
       "Q 1056 1369 1301 889 \n",
       "Q 1547 409 2034 409 \n",
       "Q 2525 409 2770 889 \n",
       "Q 3016 1369 3016 2328 \n",
       "Q 3016 3291 2770 3770 \n",
       "Q 2525 4250 2034 4250 \n",
       "z\n",
       "M 2034 4750 \n",
       "Q 2819 4750 3233 4129 \n",
       "Q 3647 3509 3647 2328 \n",
       "Q 3647 1150 3233 529 \n",
       "Q 2819 -91 2034 -91 \n",
       "Q 1250 -91 836 529 \n",
       "Q 422 1150 422 2328 \n",
       "Q 422 3509 836 4129 \n",
       "Q 1250 4750 2034 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-2212\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-32\" x=\"83.789062\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"147.412109\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"179.199219\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_2\">\n",
       "     <g id=\"line2d_4\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6fb0b76956\" x=\"52.160938\" y=\"253.318125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_5\">\n",
       "      <!-- −1.5 -->\n",
       "      <g transform=\"translate(20.878125 257.117344) scale(0.1 -0.1)\">\n",
       "       <defs>\n",
       "        <path id=\"DejaVuSans-31\" d=\"M 794 531 \n",
       "L 1825 531 \n",
       "L 1825 4091 \n",
       "L 703 3866 \n",
       "L 703 4441 \n",
       "L 1819 4666 \n",
       "L 2450 4666 \n",
       "L 2450 531 \n",
       "L 3481 531 \n",
       "L 3481 0 \n",
       "L 794 0 \n",
       "L 794 531 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "        <path id=\"DejaVuSans-35\" d=\"M 691 4666 \n",
       "L 3169 4666 \n",
       "L 3169 4134 \n",
       "L 1269 4134 \n",
       "L 1269 2991 \n",
       "Q 1406 3038 1543 3061 \n",
       "Q 1681 3084 1819 3084 \n",
       "Q 2600 3084 3056 2656 \n",
       "Q 3513 2228 3513 1497 \n",
       "Q 3513 744 3044 326 \n",
       "Q 2575 -91 1722 -91 \n",
       "Q 1428 -91 1123 -41 \n",
       "Q 819 9 494 109 \n",
       "L 494 744 \n",
       "Q 775 591 1075 516 \n",
       "Q 1375 441 1709 441 \n",
       "Q 2250 441 2565 725 \n",
       "Q 2881 1009 2881 1497 \n",
       "Q 2881 1984 2565 2268 \n",
       "Q 2250 2553 1709 2553 \n",
       "Q 1456 2553 1204 2497 \n",
       "Q 953 2441 691 2322 \n",
       "L 691 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       </defs>\n",
       "       <use xlink:href=\"#DejaVuSans-2212\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-31\" x=\"83.789062\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"147.412109\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"179.199219\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_3\">\n",
       "     <g id=\"line2d_5\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6fb0b76956\" x=\"52.160938\" y=\"207.118125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_6\">\n",
       "      <!-- −1.0 -->\n",
       "      <g transform=\"translate(20.878125 210.917344) scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-2212\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-31\" x=\"83.789062\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"147.412109\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"179.199219\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_4\">\n",
       "     <g id=\"line2d_6\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6fb0b76956\" x=\"52.160938\" y=\"160.918125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_7\">\n",
       "      <!-- −0.5 -->\n",
       "      <g transform=\"translate(20.878125 164.717344) scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-2212\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"83.789062\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"147.412109\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"179.199219\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_5\">\n",
       "     <g id=\"line2d_7\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6fb0b76956\" x=\"52.160938\" y=\"114.718125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_8\">\n",
       "      <!-- 0.0 -->\n",
       "      <g transform=\"translate(29.257813 118.517344) scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_6\">\n",
       "     <g id=\"line2d_8\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6fb0b76956\" x=\"52.160938\" y=\"68.518125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_9\">\n",
       "      <!-- 0.5 -->\n",
       "      <g transform=\"translate(29.257813 72.317344) scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-30\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-35\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"ytick_7\">\n",
       "     <g id=\"line2d_9\">\n",
       "      <g>\n",
       "       <use xlink:href=\"#m6fb0b76956\" x=\"52.160938\" y=\"22.318125\" style=\"stroke: #000000; stroke-width: 0.8\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "     <g id=\"text_10\">\n",
       "      <!-- 1.0 -->\n",
       "      <g transform=\"translate(29.257813 26.117344) scale(0.1 -0.1)\">\n",
       "       <use xlink:href=\"#DejaVuSans-31\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-2e\" x=\"63.623047\"/>\n",
       "       <use xlink:href=\"#DejaVuSans-30\" x=\"95.410156\"/>\n",
       "      </g>\n",
       "     </g>\n",
       "    </g>\n",
       "    <g id=\"text_11\">\n",
       "     <!-- gene effect -->\n",
       "     <g transform=\"translate(14.798438 189.386094) rotate(-90) scale(0.1 -0.1)\">\n",
       "      <defs>\n",
       "       <path id=\"DejaVuSans-67\" d=\"M 2906 1791 \n",
       "Q 2906 2416 2648 2759 \n",
       "Q 2391 3103 1925 3103 \n",
       "Q 1463 3103 1205 2759 \n",
       "Q 947 2416 947 1791 \n",
       "Q 947 1169 1205 825 \n",
       "Q 1463 481 1925 481 \n",
       "Q 2391 481 2648 825 \n",
       "Q 2906 1169 2906 1791 \n",
       "z\n",
       "M 3481 434 \n",
       "Q 3481 -459 3084 -895 \n",
       "Q 2688 -1331 1869 -1331 \n",
       "Q 1566 -1331 1297 -1286 \n",
       "Q 1028 -1241 775 -1147 \n",
       "L 775 -588 \n",
       "Q 1028 -725 1275 -790 \n",
       "Q 1522 -856 1778 -856 \n",
       "Q 2344 -856 2625 -561 \n",
       "Q 2906 -266 2906 331 \n",
       "L 2906 616 \n",
       "Q 2728 306 2450 153 \n",
       "Q 2172 0 1784 0 \n",
       "Q 1141 0 747 490 \n",
       "Q 353 981 353 1791 \n",
       "Q 353 2603 747 3093 \n",
       "Q 1141 3584 1784 3584 \n",
       "Q 2172 3584 2450 3431 \n",
       "Q 2728 3278 2906 2969 \n",
       "L 2906 3500 \n",
       "L 3481 3500 \n",
       "L 3481 434 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-6e\" d=\"M 3513 2113 \n",
       "L 3513 0 \n",
       "L 2938 0 \n",
       "L 2938 2094 \n",
       "Q 2938 2591 2744 2837 \n",
       "Q 2550 3084 2163 3084 \n",
       "Q 1697 3084 1428 2787 \n",
       "Q 1159 2491 1159 1978 \n",
       "L 1159 0 \n",
       "L 581 0 \n",
       "L 581 3500 \n",
       "L 1159 3500 \n",
       "L 1159 2956 \n",
       "Q 1366 3272 1645 3428 \n",
       "Q 1925 3584 2291 3584 \n",
       "Q 2894 3584 3203 3211 \n",
       "Q 3513 2838 3513 2113 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-20\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-66\" d=\"M 2375 4863 \n",
       "L 2375 4384 \n",
       "L 1825 4384 \n",
       "Q 1516 4384 1395 4259 \n",
       "Q 1275 4134 1275 3809 \n",
       "L 1275 3500 \n",
       "L 2222 3500 \n",
       "L 2222 3053 \n",
       "L 1275 3053 \n",
       "L 1275 0 \n",
       "L 697 0 \n",
       "L 697 3053 \n",
       "L 147 3053 \n",
       "L 147 3500 \n",
       "L 697 3500 \n",
       "L 697 3744 \n",
       "Q 697 4328 969 4595 \n",
       "Q 1241 4863 1831 4863 \n",
       "L 2375 4863 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-63\" d=\"M 3122 3366 \n",
       "L 3122 2828 \n",
       "Q 2878 2963 2633 3030 \n",
       "Q 2388 3097 2138 3097 \n",
       "Q 1578 3097 1268 2742 \n",
       "Q 959 2388 959 1747 \n",
       "Q 959 1106 1268 751 \n",
       "Q 1578 397 2138 397 \n",
       "Q 2388 397 2633 464 \n",
       "Q 2878 531 3122 666 \n",
       "L 3122 134 \n",
       "Q 2881 22 2623 -34 \n",
       "Q 2366 -91 2075 -91 \n",
       "Q 1284 -91 818 406 \n",
       "Q 353 903 353 1747 \n",
       "Q 353 2603 823 3093 \n",
       "Q 1294 3584 2113 3584 \n",
       "Q 2378 3584 2631 3529 \n",
       "Q 2884 3475 3122 3366 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "       <path id=\"DejaVuSans-74\" d=\"M 1172 4494 \n",
       "L 1172 3500 \n",
       "L 2356 3500 \n",
       "L 2356 3053 \n",
       "L 1172 3053 \n",
       "L 1172 1153 \n",
       "Q 1172 725 1289 603 \n",
       "Q 1406 481 1766 481 \n",
       "L 2356 481 \n",
       "L 2356 0 \n",
       "L 1766 0 \n",
       "Q 1100 0 847 248 \n",
       "Q 594 497 594 1153 \n",
       "L 594 3053 \n",
       "L 172 3053 \n",
       "L 172 3500 \n",
       "L 594 3500 \n",
       "L 594 4494 \n",
       "L 1172 4494 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      </defs>\n",
       "      <use xlink:href=\"#DejaVuSans-67\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"63.476562\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-6e\" x=\"125\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"188.378906\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-20\" x=\"249.902344\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"281.689453\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"343.212891\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-66\" x=\"378.417969\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-65\" x=\"413.623047\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-63\" x=\"475.146484\"/>\n",
       "      <use xlink:href=\"#DejaVuSans-74\" x=\"530.126953\"/>\n",
       "     </g>\n",
       "    </g>\n",
       "   </g>\n",
       "   <g id=\"patch_3\">\n",
       "    <path d=\"M 52.160938 299.518125 \n",
       "L 52.160938 22.318125 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_4\">\n",
       "    <path d=\"M 163.760938 299.518125 \n",
       "L 163.760938 22.318125 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_5\">\n",
       "    <path d=\"M 52.160938 299.518125 \n",
       "L 163.760938 299.518125 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"patch_6\">\n",
       "    <path d=\"M 52.160938 22.318125 \n",
       "L 163.760938 22.318125 \n",
       "\" style=\"fill: none; stroke: #000000; stroke-width: 0.8; stroke-linejoin: miter; stroke-linecap: square\"/>\n",
       "   </g>\n",
       "   <g id=\"text_12\">\n",
       "    <!-- DHODH -->\n",
       "    <g transform=\"translate(84.973438 16.318125) scale(0.12 -0.12)\">\n",
       "     <defs>\n",
       "      <path id=\"DejaVuSans-44\" d=\"M 1259 4147 \n",
       "L 1259 519 \n",
       "L 2022 519 \n",
       "Q 2988 519 3436 956 \n",
       "Q 3884 1394 3884 2338 \n",
       "Q 3884 3275 3436 3711 \n",
       "Q 2988 4147 2022 4147 \n",
       "L 1259 4147 \n",
       "z\n",
       "M 628 4666 \n",
       "L 1925 4666 \n",
       "Q 3281 4666 3915 4102 \n",
       "Q 4550 3538 4550 2338 \n",
       "Q 4550 1131 3912 565 \n",
       "Q 3275 0 1925 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-48\" d=\"M 628 4666 \n",
       "L 1259 4666 \n",
       "L 1259 2753 \n",
       "L 3553 2753 \n",
       "L 3553 4666 \n",
       "L 4184 4666 \n",
       "L 4184 0 \n",
       "L 3553 0 \n",
       "L 3553 2222 \n",
       "L 1259 2222 \n",
       "L 1259 0 \n",
       "L 628 0 \n",
       "L 628 4666 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "      <path id=\"DejaVuSans-4f\" d=\"M 2522 4238 \n",
       "Q 1834 4238 1429 3725 \n",
       "Q 1025 3213 1025 2328 \n",
       "Q 1025 1447 1429 934 \n",
       "Q 1834 422 2522 422 \n",
       "Q 3209 422 3611 934 \n",
       "Q 4013 1447 4013 2328 \n",
       "Q 4013 3213 3611 3725 \n",
       "Q 3209 4238 2522 4238 \n",
       "z\n",
       "M 2522 4750 \n",
       "Q 3503 4750 4090 4092 \n",
       "Q 4678 3434 4678 2328 \n",
       "Q 4678 1225 4090 567 \n",
       "Q 3503 -91 2522 -91 \n",
       "Q 1538 -91 948 565 \n",
       "Q 359 1222 359 2328 \n",
       "Q 359 3434 948 4092 \n",
       "Q 1538 4750 2522 4750 \n",
       "z\n",
       "\" transform=\"scale(0.015625)\"/>\n",
       "     </defs>\n",
       "     <use xlink:href=\"#DejaVuSans-44\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-48\" x=\"77.001953\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-4f\" x=\"152.197266\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-44\" x=\"230.908203\"/>\n",
       "     <use xlink:href=\"#DejaVuSans-48\" x=\"307.910156\"/>\n",
       "    </g>\n",
       "   </g>\n",
       "  </g>\n",
       " </g>\n",
       " <defs>\n",
       "  <clipPath id=\"pe3887c7f71\">\n",
       "   <rect x=\"52.160938\" y=\"22.318125\" width=\"111.6\" height=\"277.2\"/>\n",
       "  </clipPath>\n",
       " </defs>\n",
       "</svg>\n"
      ],
      "text/plain": [
       "<Figure size 200x500 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig, ax = plt.subplots(figsize=(2,5))\n",
    "\n",
    "sns.violinplot(\n",
    "    data=dep_df, \n",
    "    x='Leukemia',y=\"dependency\",inner=None, color=\".8\", fill=False,\n",
    "    ax=ax\n",
    ")\n",
    "\n",
    "# sns.stripplot(\n",
    "#     data=dep_df.query('name != \"other\"').copy(), \n",
    "#     jitter=True,x=\"Leukemia\", y=\"dependency\", size=10,linewidth=.3,\n",
    "#     hue=\"name\", alpha=.8,\n",
    "#     ax=ax\n",
    "# )\n",
    "\n",
    "ax.set_xticklabels(ax.get_xticklabels(), rotation=45, ha='right')\n",
    "\n",
    "ax.set_ylim(bottom=-2,top=1)\n",
    "\n",
    "ax.set_title(gene_name)\n",
    "ax.set_ylabel('gene effect')\n",
    "# plt.savefig('aml-panel-gene-expression.pdf')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "04876db1-d134-46e8-988a-4459b2ee0949",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Our CRISPRi screens\n",
    "DNMT1 inhibitor drugs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "83c6e69e-ae73-492d-b048-d263e064d77b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# scrns = load_data(screens=True,wd=wd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c1541493-4550-4899-882a-02b01b61db83",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# scrns.keys()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0fb5b422-adee-49b4-881e-fd9f677fc31a",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# res,sen = find_top(scrns['hl60_exp1_DAC_rho'].astype(float), \n",
    "#          'rho score',0.2,\n",
    "#          'Mann-Whitney p-value',0.01)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "225db192-0e7c-4267-b0b1-e05c0b457bea",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# '\",\"'.join(sen.index.to_list())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "67d4c20c-cea4-455d-b13e-5f913d8401f8",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# df = scrns['hl60_exp1_DAC_rho'].astype(float)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "cf48e65d-14ea-4931-87e2-9358a113ebbb",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# rho_df = merge_screen_data('hl60', ['exp1','exp2'], ['GSK','DAC'], 'rho', data=scrns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "be93398d-08e3-42c1-84a0-47e3b12d6a52",
   "metadata": {
    "tags": []
   },
   "source": [
    "# Our meRIP-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6bec1d8b-e568-4ceb-a716-3d3e590f61c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "hl60_RNAme_df = pd.read_csv('meRIP-seq/radar/result.sig.txt',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50c616a2-4040-4bea-90bf-6b4fcc116f8f",
   "metadata": {},
   "outputs": [],
   "source": [
    "hl60_RNAme_df[hl60_RNAme_df.name.isin(['MAT1A', 'MAT2A', 'MAT2B','METTL14'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5a4d190-e9ca-42c5-ae87-f72b722d3c9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genes1 = genes.copy()\n",
    "# genes2 = genes.copy()\n",
    "\n",
    "# genes1['hyper_m6A'] = set(m6A_hl60_df[\n",
    "#     (m6A_hl60_df.logFC > 2) & (m6A_hl60_df.p_value < 0.001)\n",
    "# ].name.tolist())\n",
    "\n",
    "# genes2['hypo_m6A']  = set(m6A_hl60_df[\n",
    "#     (m6A_hl60_df.logFC < -2) & (m6A_hl60_df.p_value < 0.001)\n",
    "# ].name.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99f12999-0a1e-48d2-81b0-b265cf1db539",
   "metadata": {},
   "outputs": [],
   "source": [
    "# fontsize=8\n",
    "\n",
    "# # Set up the plot\n",
    "# f, (ax1,ax2) = plt.subplots(1,2,figsize=(8,5))\n",
    "\n",
    "# venn(genes1,fontsize=fontsize,legend_loc=\"lower left\",ax=ax1)\n",
    "# ax1.set_title('\\n\\nDecitabine', size = 15)\n",
    "\n",
    "# # Set up the plot\n",
    "# venn(genes2,fontsize=fontsize,legend_loc=\"lower right\",ax=ax2)\n",
    "# ax2.set_title('\\n\\nDMSO', size = 15)\n",
    "\n",
    "# f.suptitle(\"HL60 DNA & RNA methylation genes\",fontsize = 25)\n",
    "# plt.tight_layout()\n",
    "# f.subplots_adjust(wspace=0, hspace=0)\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5945b96f-8ca5-4929-9d46-980b97d10c48",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reduce(set.intersection, (set(val) for val in genes1.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "259c604e-a501-40e6-b380-92d790588c8d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# reduce(set.intersection, (set(val) for val in genes2.values()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6be4e96b",
   "metadata": {},
   "outputs": [],
   "source": [
    "def intersection(lst1, lst2):\n",
    "    return list(set(lst1) & set(lst2))\n",
    "\n",
    "\n",
    "def pretty_print_attr(obj):\n",
    "    attr = []\n",
    "    ls_attr = []\n",
    "    meth = []\n",
    "    for i in dir(obj):\n",
    "        if \"_\" != i[0]:\n",
    "            if type(getattr(obj, i)) == str or type(getattr(obj, i)) == int:\n",
    "                attr.append(i)\n",
    "            elif type(getattr(obj, i)) == list:\n",
    "                ls_attr.append(i)\n",
    "            else:\n",
    "                meth.append(i)\n",
    "\n",
    "    print(\"Attributes:\\n\")\n",
    "    for i in attr: print(i+\":\", getattr(obj, i))\n",
    "    for i in ls_attr: print(i+\" list first item:\", getattr(obj, i)[0])\n",
    "    for i in ls_attr: print(i+\" length:\", len(getattr(obj, i)))\n",
    "    print(\"\\nMethods:\\n\")\n",
    "    for i in meth: print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "73dad444",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mt_wt_objs(cancer, genes):\n",
    "    # Cell lines with mutation\n",
    "    mt_list = cancer.mutated(genes)\n",
    "    mt = can.CellLineCluster(mt_list) \n",
    "\n",
    "    # CellLineCluster ojbect must be instantiated with a mutable sequence\n",
    "    # I use set operations to get wild type cell line ids and convert to a list\n",
    "    wt_list = list(set(lu.depmap_ids) - set(mt_list)) \n",
    "\n",
    "    wt = can.CellLineCluster(wt_list) \n",
    "    print (f'#of mutated cell lines:\\n\\t{len(mt.depmap_ids)}')\n",
    "    print (f'#of wildtype cell lines:\\n\\t{len(wt.depmap_ids)}')\n",
    "    \n",
    "    return mt, wt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f19f9fe4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def gene_effect_heatmap(obj1, obj2, genes, name = None):\n",
    "    #Make Figure appropriate size, dpi, and font\n",
    "    plt.rcParams.update({\"figure.figsize\": (16, 6),\n",
    "                        \"savefig.dpi\": 300,\n",
    "                        \"font.size\": 12\n",
    "                        })\n",
    "\n",
    "    #One figure with one subplot\n",
    "    fig, ax = plt.subplots(1,1)\n",
    "\n",
    "    #Construcat matrix to make heatmap and cell line labels\n",
    "    data = pd.concat([obj1.effect_of(genes), obj2.effect_of(genes)], axis=1)\n",
    "    names = can.data.cell_lines.loc[data.columns, \"cell_line_name\"]\n",
    "\n",
    "    # We want to show all ticks...\n",
    "    ax.set_xticks(np.arange(len(names)))\n",
    "    ax.set_yticks(np.arange(len(genes)))\n",
    "    # ... and label them with the respective list entries\n",
    "    ax.set_xticklabels(names)\n",
    "    ax.set_yticklabels(genes)\n",
    "\n",
    "    #make heatmap\n",
    "    im = ax.imshow(data, cmap=\"RdBu\")\n",
    "\n",
    "    #Make colorbar scale to axis\n",
    "    divider = make_axes_locatable(ax)\n",
    "    cax = divider.append_axes(\"right\", size=\"5%\", pad=0.1)\n",
    "    cbar = ax.figure.colorbar(im, ax = ax, cax = cax)\n",
    "    cbar.ax.set_ylabel(\"Gene Effect\", rotation=-90, va=\"bottom\")\n",
    "\n",
    "    #Draw Dividing line btween mutant and\n",
    "    ax.axvline(x=obj1.gene_effect.shape[1] - 0.5, c = \"black\", linewidth = 3)\n",
    "    plt.setp(ax.get_xticklabels(), rotation=-90, ha=\"left\", va=\"center\",\n",
    "         rotation_mode=\"anchor\")\n",
    "    plt.tight_layout()\n",
    "    plt.show()\n",
    "\n",
    "    if name:\n",
    "        fig.savefig(name, dpi=300)\n",
    "\n",
    "        \n",
    "def gene_effect_scatter(mt, wt, genes=None, name_scatter='', return_effect=False, plot=True, name=None):\n",
    "\n",
    "    # Average Gene Effect \n",
    "    mt_effect = mt.gene_dependency.mean(1)\n",
    "    wt_effect = wt.gene_dependency.mean(1)\n",
    "\n",
    "    if return_effect:\n",
    "        return mt_effect, wt_effect\n",
    "    \n",
    "    if plot:\n",
    "        #Make Figure appropriate size, dpi, and font\n",
    "        plt.rcParams.update({\"figure.figsize\": (8, 8),\n",
    "                            \"savefig.dpi\": 300,\n",
    "                            \"font.family\": \"sans-serif\",\n",
    "                            \"font.size\": 12\n",
    "                            })\n",
    "\n",
    "        #Generate Figure and Axis objects\n",
    "        fig, ax = plt.subplots(1,1)\n",
    "\n",
    "        #Label Axes\n",
    "        ax.set_xlabel(f\"{name_scatter} MT Average Gene Effect (CERES Score)\")\n",
    "        ax.set_ylabel(f\"{name_scatter} WT Average Gene Effect (CERES Score)\")\n",
    "\n",
    "        #Draw Line at median common essential value\n",
    "        ax.axhline(y = 0.50,\n",
    "                   c = \"black\",\n",
    "                   linewidth=0.5,\n",
    "                   label = \"Minimun Gene Dependencey Probability\"\n",
    "                  )\n",
    "\n",
    "        ax.axvline(x = 0.50,\n",
    "                   c= \"black\",\n",
    "                   linewidth=0.5)\n",
    "\n",
    "        #Plot all genes\n",
    "        ax.scatter(mt_effect,\n",
    "                   wt_effect,\n",
    "                   c = \"#2166ac\",\n",
    "                   alpha = 0.1,\n",
    "                   s = 50\n",
    "                  )\n",
    "\n",
    "        ax.legend()\n",
    "\n",
    "        if genes:\n",
    "            #For Labeling\n",
    "            if type(genes) != list:\n",
    "                mt_lab = mt_effect.loc[[genes]]\n",
    "                wt_lab = wt_effect.loc[[genes]]\n",
    "            else:\n",
    "                mt_lab = mt_effect.loc[genes]\n",
    "                wt_lab = wt_effect.loc[genes]\n",
    "            # Outline Genes To label\n",
    "            ax.scatter(mt_lab,\n",
    "                       wt_lab,\n",
    "                       c = \"#2166ac\",\n",
    "                       s = 50,\n",
    "                       edgecolor = \"black\",\n",
    "                       linewidth = 2,\n",
    "                       alpha = 0.7\n",
    "                      )\n",
    "        #     for i in range(mt_lab.shape[0]):\n",
    "        #         text = list(mt_lab.index)\n",
    "        #         ax.annotate(text[i],\n",
    "        #                     xy = (mt_lab[i], wt_lab[i]),\n",
    "        #                     xytext = label[i],\n",
    "        #                     xycoords = \"data\",\n",
    "        #                     arrowprops = {\"arrowstyle\": \"-\"}\n",
    "        #                    )\n",
    "            texts = []\n",
    "            for x, y, s in zip(mt_lab, wt_lab, mt_lab.index.tolist()):\n",
    "                texts.append(plt.text(x, y, s))\n",
    "\n",
    "            adjust_text(\n",
    "                texts, force_points=0.2, force_text=0.2,\n",
    "                expand_points=(2, 2), \n",
    "                expand_text=(2, 2),\n",
    "                arrowprops=dict(arrowstyle=\"-\", color='black', lw=0.2)\n",
    "            )\n",
    "\n",
    "        plt.show()\n",
    "\n",
    "        if name:\n",
    "            fig.savefig(name, dpi=300)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b572d859",
   "metadata": {},
   "outputs": [],
   "source": [
    "cell_lines = can.data.cell_lines\n",
    "genes = can.data.genes.index.tolist()\n",
    "\n",
    "m6A = pd.read_csv('m6A-genes.tsv',sep='\\t')\n",
    "m6A_genes = [g.split(' ')[0] for g in m6A.Names if g in genes]\n",
    "\n",
    "DNMT_genes = [g for g in genes if 'DNMT' in g]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f03cc6c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "set(m6A.Type)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7e1a8101",
   "metadata": {},
   "outputs": [],
   "source": [
    "m6A_writer_genes = [g.split(' ')[0] for g in m6A[m6A.Type == 'm6A writer'].Names if g in genes] \n",
    "m6A_eraser_genes = [g.split(' ')[0] for g in m6A[m6A.Type == 'm6A eraser'].Names if g in genes] "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be37f4e0",
   "metadata": {},
   "outputs": [],
   "source": [
    "can.data.mutations.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a3c2333",
   "metadata": {},
   "outputs": [],
   "source": [
    "list(set(can.data.mutations.Protein_Change))[1:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a3f192c2-6e82-43f4-924d-1512987b13a8",
   "metadata": {},
   "source": [
    "### Analyzing Global Gene Dependency\n",
    "\n",
    "\n",
    "> #### [Note-about-Gene-Effect-Scores:-Dependency-vs-Essentiality:](https://candi.readthedocs.io/en/latest/kras_egfr_scatter.html#Note-about-Gene-Effect-Scores:-Dependency-vs-Essentiality) \n",
    "A more negative gene effect means more dependent. A gene effect of -1.0 is the median gene effect of all common essential genes. If a gene has a gene effect of -1.0 or lower it then that gene is essential. A cell line can still be dependent on a gene with a lower gene effect if knocking out that gene slows growth/proliferation.\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0028bb6-3913-4195-94bd-8d8370074218",
   "metadata": {},
   "source": [
    "### m6A writer mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d89412b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "m6A_mt, m6A_wt = mt_wt_objs(lu, m6A_writer_genes + ['TP53'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "794f0b6d",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_effect_heatmap(m6A_mt, m6A_wt,m6A_writer_genes + DNMT_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "35cc2db7",
   "metadata": {},
   "outputs": [],
   "source": [
    "m6A_mt_effect, m6A_wt_effect = gene_effect_scatter(\n",
    "    m6A_mt, m6A_wt,return_effect=True,plot=False\n",
    ")\n",
    "\n",
    "labels = intersection (\n",
    "    m6A_mt_effect[m6A_mt_effect <= 0.45].index.tolist(),m6A_wt_effect[m6A_wt_effect >= 0.7].index.tolist()\n",
    ") + intersection (m6A_mt_effect[m6A_mt_effect >= 0.65].index.tolist(),m6A_wt_effect[m6A_wt_effect <= 0.3].index.tolist())\n",
    "\n",
    "gene_effect_scatter(\n",
    "    m6A_mt, m6A_wt,\n",
    "    genes=labels + ['DNMT1'] + ['INTS5'] + \n",
    "    [g for g in m6A_writer_genes if g in m6A_mt_effect.index.tolist()],\n",
    "    name_scatter='m6A writer',\n",
    "    name='DNA-RNA-mtyl-m6A-scatter.pdf'\n",
    ")\n",
    "\n",
    "run_enrichment(labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a24459e0-4031-4f1f-b2a9-f809e0ad2473",
   "metadata": {},
   "source": [
    "### INTS5 or m6A writer mutation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "108381ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt, wt = mt_wt_objs(lu, m6A_writer_genes+['INTS5','TP53'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5ba8f3ae",
   "metadata": {},
   "outputs": [],
   "source": [
    "# gene_effect_heatmap(DNMT_mt, DNMT_wt,m6A_writer_genes + DNMT_genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a815a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "mt_effect, wt_effect = gene_effect_scatter(\n",
    "    mt, wt,return_effect=True,plot=False\n",
    ")\n",
    "\n",
    "labels = intersection (\n",
    "    mt_effect[mt_effect < 0.35].index.tolist(),wt_effect[wt_effect > 0.65].index.tolist()\n",
    ") + intersection (mt_effect[mt_effect > 0.7].index.tolist(),wt_effect[wt_effect < 0.4].index.tolist())\n",
    "\n",
    "gene_effect_scatter(\n",
    "    mt, wt,\n",
    "    genes=labels + ['DNMT1'] + ['INTS5', 'SQLE'] +\n",
    "    [g for g in m6A_writer_genes if g in mt_effect.index.tolist()],\n",
    "    name_scatter='DNMT family',\n",
    "#     name='DNA-RNA-mtyl-DNmt-scatter.pdf'\n",
    ")\n",
    "\n",
    "run_enrichment(labels)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "806c1669-7785-4017-a657-51bf4ed878c6",
   "metadata": {},
   "source": [
    "### DESeq analysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2528f39c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_counts_coldata(obj1, obj2, condition, factor1, factor2):\n",
    "\n",
    "    counts1 = obj1.rnaseq_reads\n",
    "    coldat1 = pd.Series(counts1.shape[1] * [factor1], index = counts1.columns, name = condition)\n",
    "\n",
    "    counts2 = obj2.rnaseq_reads\n",
    "    coldat2 = pd.Series(counts2.shape[1] * [factor2], index = counts2.columns, name = condition)\n",
    "\n",
    "    #Concatenate Column Data\n",
    "    coldat = pd.concat([coldat1, coldat2], axis = 0)\n",
    "    #Concatenate read count data\n",
    "    counts_mat = pd.concat([counts1, counts2], axis = 1)\n",
    "    #Sum duplicate indeces\n",
    "    counts_mat = counts_mat.groupby(counts_mat.index).sum().astype(int)\n",
    "\n",
    "    return counts_mat, coldat"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "758b468a",
   "metadata": {},
   "outputs": [],
   "source": [
    "counts_mat, coldat = make_counts_coldata(m6A_mt, m6A_wt,'mutation','mt','wt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8a40ebc2",
   "metadata": {},
   "outputs": [],
   "source": [
    "coldat[coldat == 'mt']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "54414a25-e271-4ba4-bc07-3225c8457b41",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "8a0c3f66-d1c8-46b7-9397-b17b7c5e2724",
   "metadata": {
    "tags": []
   },
   "source": [
    "## CCLE DNA Methylation data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5ccab04b-4f9d-441a-bb5d-33a31e0e8d5f",
   "metadata": {},
   "source": [
    "HL-60 raw data - https://trace.ncbi.nlm.nih.gov/Traces/sra/?run=SRR8633931"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "97b45437-6c07-41ac-a049-69e98e84144b",
   "metadata": {},
   "outputs": [],
   "source": [
    "depmap_data_path = str(pathlib.Path(can.__file__).resolve().parent.parent)+'/setup/data/depmap'\n",
    "\n",
    "cgi_CpG_clusters = pd.read_csv(\n",
    "    f'{depmap_data_path}/CCLE_RRBS_cgi_CpG_clusters_20181119.txt.gz',\n",
    "    sep='\\t', quotechar='\"',low_memory=False\n",
    ")\n",
    "\n",
    "enh_CpG_clusters = pd.read_csv(\n",
    "    f'{depmap_data_path}/CCLE_RRBS_enh_CpG_clusters_20181119.txt.gz',\n",
    "    sep='\\t', quotechar='\"',low_memory=False\n",
    ")\n",
    "\n",
    "tss_CpG_clusters = pd.read_csv(\n",
    "    f'{depmap_data_path}/CCLE_RRBS_tss_CpG_clusters_20181022.txt.gz',\n",
    "    sep='\\t', quotechar='\"',low_memory=False\n",
    ")\n",
    "\n",
    "TSS1kb = pd.read_csv(\n",
    "    f'{depmap_data_path}/CCLE_RRBS_TSS1kb_20181022.txt.gz',\n",
    "    sep='\\t', quotechar='\"',low_memory=False\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f077becc-f6b0-4071-8688-b8f3b1115d45",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cgi_CpG_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6fa9bcf1-89e4-408d-92b2-09840333a3dd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# enh_CpG_clusters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7afa5129-5d76-4e1e-9d0e-2daec9554b9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes = {}"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c34b138a-9d37-4f30-b23c-472d1286636f",
   "metadata": {},
   "source": [
    "### TSS1kb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "232cf928-82c5-4197-9519-1ee4f71962b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss1kb_hl60_df = TSS1kb.loc[\n",
    "    :,['locus_id','CpG_sites_hg19','avg_coverage']+\n",
    "    [col for col in TSS1kb.columns for aml in [\n",
    "    'HL60_',#'OCIAML2','OCIAML2','KG1_'\n",
    "] if aml in col]].dropna()\n",
    "# .unique()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38a932d8-ea1c-49c5-b550-c825b21b0c6b",
   "metadata": {},
   "outputs": [],
   "source": [
    "data1 = tss1kb_hl60_df.loc[:,['locus_id','HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE']].set_index('locus_id').astype('double')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c6c7ffd9-dd1a-487f-b1b8-64753a51b944",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes['tss1kb'],_,_,_ = zip(*\n",
    "    data1[data1.iloc[:,0] > 0.95].index.str.split('_').tolist()\n",
    ")\n",
    "genes['tss1kb'] = set(tss1kb_hl60_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fcb5d166-8b2f-4d90-a762-a6f5884ee329",
   "metadata": {},
   "source": [
    "### CpG"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e2c1abb-c5a5-43f5-8ff1-a53e8accc589",
   "metadata": {},
   "outputs": [],
   "source": [
    "tss_CpG_hl60_df = tss_CpG_clusters.loc[\n",
    "    :,['cluster_id','CpG_sites_hg19','avg_coverage']+\n",
    "    [col for col in TSS1kb.columns for aml in [\n",
    "    'HL60_',#'OCIAML2','OCIAML2','KG1_'\n",
    "] if aml in col]].dropna()\n",
    "# .unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "82f8a790-d849-446b-b2db-09638820481a",
   "metadata": {},
   "outputs": [],
   "source": [
    "data2 = tss_CpG_hl60_df.loc[:,['cluster_id','HL60_HAEMATOPOIETIC_AND_LYMPHOID_TISSUE']].set_index('cluster_id').astype('double')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "be3ef3c9-136d-4fa3-ac42-b7be4fb81a28",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Set up the plot\n",
    "f, (ax1,ax2) = plt.subplots(1,2,figsize=(8,2))\n",
    "\n",
    "ax1.hist(\n",
    "    data1, histtype='step', stacked=False, fill=False, bins = 100, \n",
    ")\n",
    "ax1.set_title('HL60 TSS1kb', size = 12)\n",
    "ax1.set_ylabel('log(Frequency)', size= 10)\n",
    "ax1.set_ylim(0,10000)\n",
    "\n",
    "# Set up the plot\n",
    "ax2 = plt.subplot(1, 2, 2)\n",
    "\n",
    "ax2.hist(\n",
    "    data2, histtype='step', stacked=False, fill=False, bins = 100, \n",
    ")\n",
    "ax2.set_title('HL60 TSS CpG', size = 12)\n",
    "ax2.set_ylim(0,10000)\n",
    "# Title and labels\n",
    "\n",
    "# f.set_yscale('log')\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5b2250bb-4aed-4940-85d6-1bda35f7e7c2",
   "metadata": {},
   "outputs": [],
   "source": [
    "genes['tss_CpG'] = set([x[0] for x in data[data.iloc[:,0] > 0.95].index.str.split('_').tolist()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1999cab-98a5-4ef9-9a9b-1a703583cbde",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f6bf1c8d-1ab4-4971-b4cb-807d4cf801d5",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0dd3c27b-ac65-496a-9045-4dc8eb8f29d1",
   "metadata": {
    "tags": []
   },
   "source": [
    "## CCLE metabolomics\n",
    "\n",
    "    CCLE_metabolomics_20190502.csv\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19aa7275-11ea-42e3-90a0-397dabe714eb",
   "metadata": {
    "tags": []
   },
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7352f7ea",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "56ff40c3-8508-4280-9bb6-ed3a7b8ba00d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:candi-kg]",
   "language": "python",
   "name": "conda-env-candi-kg-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
