{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "98f43882-1b7c-4127-b37e-374603b7b2ee",
   "metadata": {},
   "source": [
    "https://f1000research.com/articles/6-2055/v2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40c096cf-eb4e-428d-a70d-0e184e5cc828",
   "metadata": {},
   "source": [
    "> #### Resolution of the DNA methylation state of single CpG dyads using in silico strand annealing and WGBS data\n",
    "Nature protocol | https://www.nature.com/articles/s41596-018-0090-x\n",
    "\n",
    "> - Whole-genome bisulfite sequencing (WGBS) has been widely used to quantify cytosine DNA methylation frequency in an expanding array of cell and tissue types. Because of the denaturing conditions used, this method ultimately leads to the measurement of methylation frequencies at single cytosines. \n",
    "> - Hence, the methylation frequency of CpG dyads (two complementary CG dinucleotides) can be only indirectly inferred by overlaying the methylation frequency of two cytosines measured independently. \n",
    "> - Furthermore, hemi-methylated CpGs (hemiCpGs) have not been previously analyzed in WGBS studies. We recently developed in silico strand annealing (iSA), a bioinformatics method applicable to WGBS data, to resolve the methylation status of CpG dyads into __unmethylated__, __hemi-methylated__, and __methylated__. \n",
    "> - HemiCpGs account for 4â€“20% of the DNA methylome in different cell types, and some can be inherited across cell divisions, suggesting a role as a stable epigenetic mark. Therefore, it is important to resolve hemiCpGs from fully methylated CpGs in WGBS studies. This protocol describes step-by-step commands to accomplish this task, including dividing alignments by strand, pairing alignments between strands, and extracting single-fragment methylation calls. \n",
    "> - The versatility of iSA enables its application downstream of other WGBS-related methods such as nasBS-seq (nascent DNA bisulfite sequencing), ChIP-BSseq (ChIP followed by bisulfite sequencing), TAB-seq, oxBS-seq, and fCAB-seq. iSA is also tunable for analyzing the methylation status of cytosines in any sequence context. \n",
    "> - We exemplify this flexibility by uncovering the single-fragment non-CpG methylome. \n",
    "> - This protocol provides enough details for users with little experience in bioinformatic analysis and takes 2â€“7 h."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "757203cb-1bec-49eb-b3d8-dbbb4b25d563",
   "metadata": {},
   "source": [
    "> #### Extended-representation bisulfite sequencing of gene regulatory elements in multiplexed samples and single cells\n",
    "https://www.nature.com/articles/s41587-021-00910-x\n",
    "\n",
    "<!-- > - Before alignment, primer dimers were filtered using Cutadapt version 2.7 and the following parameters: `--discard -a GCTCTTCCGATCT`. Short read pairs were trimmed using Trim Galore version 0.6.5 and the following parameters: `--paired --illumina --nextseq 20`. \n",
    "> - High-quality sequencing reads were then aligned to an in silico bisulfite-converted reference genome (hg38 and mm10) using methylCtools version 1.0.0 (https://github.com/hovestadt/methylCtools, ref. 49) and bwa mem version 0.7.17. \n",
    "> - Sorted alignments were further processed to only maintain uniquely mapped read pairs with a mapping score â‰¥1, that were mapping to an MspI cut site and that had an insert size between 20â€‰bp and 600â€‰bp. Putative PCR duplicates were removed by considering the outer mapping position of both paired-end reads (read 2 being located at the MspI cut site and read 1 being located at variable positions), as well as the random hexamer sequence that was trimmed before alignment and functions as a UMI. For library complexity analysis, alignments were downsampled before this step. \n",
    "> - We note that multiple random hexamer priming events during the second-strand synthesis step might lead to additional sequencing reads from the same original fragment that cannot be identified using this approach. DNA methylation calling was performed using methylCtools bcall and the --trimPE parameter. Detailed quality metrics for each library are provided in Supplementary Table 1. DNA methylation values were deposited in the GEO (GSE149954) for all samples reported in this study.\n",
    "\n",
    "\n",
    " -->\n",
    " \n",
    " This is simply called RRBS protocol :)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "29fd4217-9355-4f9b-afba-65fda8e3ae97",
   "metadata": {},
   "source": [
    "# SRA Data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "97f876cd-7c3b-43ac-b2eb-0fd9db38a394",
   "metadata": {},
   "source": [
    "my `sra` env"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d1a2641-48f9-48d0-9c2b-8c561765b797",
   "metadata": {},
   "source": [
    "`fastq-dump`\n",
    "\n",
    "https://www.biostars.org/p/222122/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f6118f9-78af-4b47-8693-67fcbcd0e15e",
   "metadata": {},
   "source": [
    "<!--     fastq-dump --outdir fastq --gzip --split-3 sra/SRR11711273.sra\n",
    "\n",
    "    Read 17522209 spots for sra/SRR11711273.sra\n",
    "    Written 17522209 spots for sra/SRR11711273.sra\n",
    "\n",
    "\n",
    "___\n",
    "\n",
    "    parallel-fastq-dump --threads 12 --outdir fastq/ --split-files --gzip sra/SRR11711272.sra -->\n",
    "    \n",
    "\n",
    "<!--     fasterq-dump --split-files --include-technical --threads 1 \n",
    "    --temp . --outfile HL60_10ng_dmso.fastq --progress SRR11711272\n",
    "\n",
    "\n",
    "    fasterq-dump --split-files --include-technical --threads 1 \n",
    "    --temp . --outfile HL60_10ng_decitabine.fastq --progress SRR11711273 -->\n",
    "\n",
    "\n",
    "    wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR117/072/SRR11711272/SRR11711272_1.fastq.gz\n",
    "    wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR117/072/SRR11711272/SRR11711272_2.fastq.gz\n",
    "    \n",
    "    wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR117/073/SRR11711273/SRR11711273_1.fastq.gz\n",
    "    wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR117/073/SRR11711273/SRR11711273_2.fastq.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f90b8e33-5771-48ec-81c9-8559a6ebd64e",
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir fastq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "3d70549a-c677-4852-96cc-143eb9e348ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "â€˜SRR11711272_1.fastq.gzâ€™ -> â€˜fastq/SRR11711272_1.fastq.gzâ€™\n",
      "â€˜SRR11711272_2.fastq.gzâ€™ -> â€˜fastq/SRR11711272_2.fastq.gzâ€™\n",
      "â€˜SRR11711273_1.fastq.gzâ€™ -> â€˜fastq/SRR11711273_1.fastq.gzâ€™\n",
      "â€˜SRR11711273_2.fastq.gzâ€™ -> â€˜fastq/SRR11711273_2.fastq.gzâ€™\n"
     ]
    }
   ],
   "source": [
    "!mv -v *fastq.gz fastq/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2e4ab690-799b-4f9f-9933-b2ce1f528a79",
   "metadata": {},
   "source": [
    "# Reading and annotating the methylation counts"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b0ea6a90-4ab8-4083-8bce-e067df3c411a",
   "metadata": {},
   "source": [
    "## Trimming"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54ad0b92-8944-410e-8c2f-bc2ba483d2b0",
   "metadata": {},
   "source": [
    "<!-- > Before alignment, primer dimers were filtered using Cutadapt version 2.7 and the following parameters: `--discard -a GCTCTTCCGATCT`. Short read pairs were trimmed using Trim Galore version 0.6.5 and the following parameters: `--paired --illumina --nextseq 20`. -->"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "277afb82-de71-4c3b-a827-3f42c07bf0b9",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p logs/\n",
    "!mkdir -p logs/trim_galore/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "4f1bf1cd-ddbd-4fed-b8c0-1b3d8f7ba8ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "trim_galore --core 3 --paired --rrbs -o fastq fastq/SRR11711272_1.fastq.gz fastq/SRR11711272_2.fastq.gz\n",
      "trim_galore --core 3 --paired --rrbs -o fastq fastq/SRR11711273_1.fastq.gz fastq/SRR11711273_2.fastq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "for fq1 in fastq/*1.fastq.gz; do\n",
    "    fq2=${fq1/1.fastq/2.fastq}\n",
    "    b=`basename $fq1`\n",
    "    log_file=${b/_1.fastq.gz/.log}\n",
    "    cm=\"trim_galore --core 3 --paired --rrbs -o fastq $fq1 $fq2\"\n",
    "    echo $cm\n",
    "    $cm &> logs/trim_galore/$log_file;\n",
    "    wait\n",
    "done"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "7f809031-d391-4742-a500-8dfac79140ea",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "â€˜fastq/SRR11711272_1.fastq.gz_trimming_report.txtâ€™ -> â€˜logs/trim_galore/SRR11711272_1.fastq.gz_trimming_report.txtâ€™\n",
      "â€˜fastq/SRR11711272_2.fastq.gz_trimming_report.txtâ€™ -> â€˜logs/trim_galore/SRR11711272_2.fastq.gz_trimming_report.txtâ€™\n",
      "â€˜fastq/SRR11711273_1.fastq.gz_trimming_report.txtâ€™ -> â€˜logs/trim_galore/SRR11711273_1.fastq.gz_trimming_report.txtâ€™\n",
      "â€˜fastq/SRR11711273_2.fastq.gz_trimming_report.txtâ€™ -> â€˜logs/trim_galore/SRR11711273_2.fastq.gz_trimming_report.txtâ€™\n"
     ]
    }
   ],
   "source": [
    "mv -v fastq/*trimming_report.txt logs/trim_galore/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "b7979419-d6f7-4acf-bb3d-324297a28b23",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "  \u001b[34m/\u001b[0m\u001b[32m/\u001b[0m\u001b[31m/\u001b[0m \u001b]8;id=765597;https://multiqc.info\u001b\\\u001b[1mMultiQC\u001b[0m\u001b]8;;\u001b\\ ðŸ” \u001b[2m| v1.12\u001b[0m\n",
      "\n",
      "\u001b[34m|           multiqc\u001b[0m | Search path : /data_gilbert/home/aarab/Projects/Decitabine-treatment/Bisulfite-seq/logs/trim_galore\n",
      "\u001b[2K\u001b[34m|\u001b[0m         \u001b[34msearching\u001b[0m | \u001b[90mâ”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”â”\u001b[0m \u001b[35m100%\u001b[0m \u001b[32m6/6\u001b[0m  xt\u001b[0mlogs/trim_galore/SRR11711273.log\u001b[0m\n",
      "\u001b[?25h\u001b[34m|          cutadapt\u001b[0m | Found 4 reports\n",
      "\u001b[34m|           multiqc\u001b[0m | Compressing plot data\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mDeleting    : multiqc-trim.html   (-f was specified)\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | \u001b[33mDeleting    : multiqc-trim_data   (-f was specified)\u001b[0m\n",
      "\u001b[34m|           multiqc\u001b[0m | Report      : multiqc-trim.html\n",
      "\u001b[34m|           multiqc\u001b[0m | Data        : multiqc-trim_data\n",
      "\u001b[34m|           multiqc\u001b[0m | MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "!multiqc logs/trim_galore/ -f -n multiqc-trim"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dd05bb2-8b33-42db-a691-9770aa5c1b0e",
   "metadata": {},
   "source": [
    "## Processing the BS-seq FASTQ files with Bismark"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e379aacb-9805-473b-aba1-fcfaf4d31556",
   "metadata": {},
   "source": [
    "Bismark - \n",
    "https://github.com/FelixKrueger/Bismark\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "40aa1dff-17c7-48e3-88b1-b98e6e31cb55",
   "metadata": {},
   "source": [
    "It faild maybe because of the genome annotation - [there's no output result #34\n",
    "](https://github.com/FelixKrueger/Bismark/issues/34)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e77fd44-8c15-4ec0-a941-a8083477e440",
   "metadata": {},
   "source": [
    "I've discussed my issues here https://github.com/FelixKrueger/Bismark/issues/475"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "56722c56-5d4e-41cd-893e-6418b5aba07a",
   "metadata": {},
   "source": [
    "We solved the issue together! https://github.com/FelixKrueger/Bismark/issues/475#issuecomment-1061147127"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b60a0aa5-74e2-40df-baab-4089bacdbbe1",
   "metadata": {},
   "source": [
    "### Index genome"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42879a82-5889-4899-9bac-82dad9885ede",
   "metadata": {},
   "source": [
    "__(I) Running bismark_genome_preparation__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "id": "054d804d-0d2e-48af-b722-9e59b7608aff",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "# !bismark_genome_preparation ~/genomes/hg38/gencode.v34/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "id": "3c938d55-55fa-43b5-bea6-2e54eaa16f5e",
   "metadata": {},
   "outputs": [],
   "source": [
    "!gzip -d /data_gilbert/home/aarab/genomes/hg38/chromosomes/*.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "38f65a4e-a994-4407-9452-56f34127c50f",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path to genome folder specified as: /data_gilbert/home/aarab/genomes/hg38/chromosomes/\n",
      "Aligner to be used: >> Bowtie 2 << (default)\n",
      "Writing bisulfite genomes out into a single MFA (multi FastA) file\n",
      "\n",
      "Bismark Genome Preparation - Step I: Preparing folders\n",
      "\n",
      "Bisulfite Genome Indexer version v0.23.1 (last modified: 27 Jan 2021)\n",
      "Created Bisulfite Genome folder /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/\n",
      "Created Bisulfite Genome folder /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/CT_conversion/\n",
      "Created Bisulfite Genome folder /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/GA_conversion/\n",
      "\n",
      "Step I - Prepare genome folders - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step II: Bisulfite converting reference genome\n",
      "\n",
      "conversions performed:\n",
      "chromosome\tC->T\tG->A\n",
      "chr1\t48055043\t48111528\n",
      "chr10\t27639505\t27719976\n",
      "chr11\t27903257\t27981801\n",
      "chr12\t27092804\t27182678\n",
      "chr13\t18839192\t18933605\n",
      "chr14\t18423758\t18559033\n",
      "chr15\t17752941\t17825903\n",
      "chr16\t18172742\t18299976\n",
      "chr17\t18723944\t18851500\n",
      "chr18\t15794455\t16061651\n",
      "chr19\t13954580\t14061132\n",
      "chr2\t48318180\t48450903\n",
      "chr20\t13916133\t14094472\n",
      "chr21\t8185244\t8226381\n",
      "chr22\t9160652\t9246186\n",
      "chr3\t39233483\t39344259\n",
      "chr4\t36236976\t36331025\n",
      "chr5\t35731600\t35879674\n",
      "chr6\t33646690\t33713330\n",
      "chr7\t32317984\t32378859\n",
      "chr8\t29030173\t29103787\n",
      "chr9\t25099811\t25170662\n",
      "chrM\t5181\t2169\n",
      "chrX\t30523780\t30697741\n",
      "chrY\t5285789\t5286894\n",
      "\n",
      "Total number of conversions performed:\n",
      "C->T:\t599043897\n",
      "G->A:\t601515125\n",
      "\n",
      "Step II - Genome bisulfite conversions - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step III: Launching the Bowtie 2 indexer\n",
      "Please be aware that this process can - depending on genome size - take several hours!\n",
      "\n",
      "Preparing indexing of CT converted genome in /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/CT_conversion/\n",
      "Parent process: Starting to index C->T converted genome with the following command:\n",
      "\n",
      "bowtie2-build  -f genome_mfa.CT_conversion.fa BS_CT\n",
      "\n",
      "Settings:\n",
      "  Output files: \"BS_CT.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.CT_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "Preparing indexing of GA converted genome in /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/GA_conversion/\n",
      "Child process: Starting to index G->A converted genome with the following command:\n",
      "\n",
      "bowtie2-build  -f genome_mfa.GA_conversion.fa BS_GA\n",
      "\n",
      "Settings:\n",
      "  Output files: \"BS_GA.*.bt2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Max bucket size: default\n",
      "  Max bucket size, sqrt multiplier: default\n",
      "  Max bucket size, len divisor: 4\n",
      "  Difference-cover sample period: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.GA_conversion.fa\n",
      "Building a SMALL index\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:17\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time reading reference sizes: 00:00:17\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:16\n",
      "bmax according to bmaxDivN setting: 734413920\n",
      "Using parameters --bmax 550810440 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Time to join reference sequences: 00:00:16\n",
      "bmax according to bmaxDivN setting: 734413920\n",
      "Using parameters --bmax 550810440 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 550810440 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  Passed!  Constructing with these parameters: --bmax 550810440 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:01:35\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  V-Sorting samples time: 00:01:37\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:28\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Ranking v-sort output time: 00:00:29\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:42\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:42\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.19665e+08 (target: 550810439)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (550810440) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.67207e+08 (target: 550810439)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (550810440) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 506437153 for bucket 1\n",
      "  (Using difference cover)\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 265840023 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:03:32\n",
      "Returning block of 265840024 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (550810440) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 313712311 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:09:16\n",
      "Returning block of 506437154 for bucket 1\n",
      "  Sorting block time: 00:04:24\n",
      "Returning block of 313712312 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (550810440) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "Getting block 2 of 7\n",
      "  Reserving size (550810440) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 2: 10%\n",
      "  bucket 3: 90%\n",
      "  bucket 2: 20%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 269510198 for bucket 3\n",
      "  (Using difference cover)\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 447885829 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:03:48\n",
      "Returning block of 269510199 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (550810440) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 416425897 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:18\n",
      "Returning block of 447885830 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (550810440) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 539608788 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:06:03\n",
      "Returning block of 416425898 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (550810440) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 480655197 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:12\n",
      "Returning block of 480655198 for bucket 5\n",
      "  Sorting block time: 00:10:41\n",
      "Returning block of 539608789 for bucket 3\n",
      "Getting block 6 of 8\n",
      "  Reserving size (550810440) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 367604136 for bucket 6\n",
      "  (Using difference cover)\n",
      "Getting block 4 of 7\n",
      "  Reserving size (550810440) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 384792641 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:05:38\n",
      "Returning block of 367604137 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (550810440) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 547961096 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:14\n",
      "Returning block of 384792642 for bucket 4\n",
      "Getting block 5 of 7\n",
      "  Reserving size (550810440) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 459841719 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:14\n",
      "Returning block of 547961097 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (550810440) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 275946816 for bucket 8\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:57\n",
      "Returning block of 459841720 for bucket 5\n",
      "Getting block 6 of 7\n",
      "  Reserving size (550810440) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 328692208 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:04:02\n",
      "Returning block of 275946817 for bucket 8\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1468669118\n",
      "fchr[G]: 2067713015\n",
      "fchr[T]: 2067713015\n",
      "fchr[$]: 2937655681\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 983425384 bytes to primary EBWT file: BS_GA.1.bt2.tmp\n",
      "Wrote 734413928 bytes to secondary EBWT file: BS_GA.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2937655681\n",
      "    bwtLen: 2937655682\n",
      "    sz: 734413921\n",
      "    bwtSz: 734413921\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 183603481\n",
      "    offsSz: 734413924\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 15300291\n",
      "    numLines: 15300291\n",
      "    ebwtTotLen: 979218624\n",
      "    ebwtTotSz: 979218624\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 01:03:58\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:13\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:16\n",
      "  Time to reverse reference sequence: 00:00:02\n",
      "bmax according to bmaxDivN setting: 734413920\n",
      "Using parameters --bmax 550810440 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 550810440 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:01:36\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:29\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:42\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Sorting block time: 00:06:25\n",
      "Returning block of 328692209 for bucket 6\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.67207e+08 (target: 550810439)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (550810440) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 393309086 for bucket 1\n",
      "  (Using difference cover)\n",
      "Getting block 7 of 7\n",
      "  Reserving size (550810440) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 270397337 for bucket 7\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:05:32\n",
      "Returning block of 393309087 for bucket 1\n",
      "  Sorting block time: 00:05:16\n",
      "Returning block of 270397338 for bucket 7\n",
      "Getting block 2 of 8\n",
      "  Reserving size (550810440) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "Exited Ebwt loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 867153993\n",
      "fchr[G]: 867153993\n",
      "fchr[T]: 1468669118\n",
      "fchr[$]: 2937655681\n",
      "Exiting Ebwt::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 983425384 bytes to primary EBWT file: BS_CT.1.bt2.tmp\n",
      "Wrote 734413928 bytes to secondary EBWT file: BS_CT.2.bt2.tmp\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from Ebwt constructor\n",
      "Headers:\n",
      "    len: 2937655681\n",
      "    bwtLen: 2937655682\n",
      "    sz: 734413921\n",
      "    bwtSz: 734413921\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 20\n",
      "    eftabSz: 80\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 183603481\n",
      "    offsSz: 734413924\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideBwtSz: 48\n",
      "    sideBwtLen: 192\n",
      "    numSides: 15300291\n",
      "    numLines: 15300291\n",
      "    ebwtTotLen: 979218624\n",
      "    ebwtTotSz: 979218624\n",
      "    color: 0\n",
      "    reverse: 0\n",
      "Total time for call to driver() for forward index: 01:16:51\n",
      "Reading reference sizes\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 469629838 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Time reading reference sizes: 00:00:14\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:15\n",
      "  Time to reverse reference sequence: 00:00:02\n",
      "bmax according to bmaxDivN setting: 734413920\n",
      "Using parameters --bmax 550810440 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 550810440 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:01:38\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:29\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:39\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.19665e+08 (target: 550810439)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering Ebwt loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (550810440) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 519023662 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:06:16\n",
      "Returning block of 469629839 for bucket 2\n",
      "Getting block 3 of 8\n",
      "  Reserving size (550810440) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 433807892 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:34\n",
      "Returning block of 519023663 for bucket 1\n",
      "  Sorting block time: 00:05:11\n",
      "Returning block of 433807893 for bucket 3\n",
      "Getting block 2 of 7\n",
      "  Reserving size (550810440) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 461139791 for bucket 2\n",
      "  (Using difference cover)\n",
      "Getting block 4 of 8\n",
      "  Reserving size (550810440) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 157321038 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:01:46\n",
      "Returning block of 157321039 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (550810440) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 435850394 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:09:19\n",
      "Returning block of 461139792 for bucket 2\n",
      "  Sorting block time: 00:07:26\n",
      "Returning block of 435850395 for bucket 5\n",
      "Getting block 3 of 7\n",
      "  Reserving size (550810440) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 409562682 for bucket 3\n",
      "  (Using difference cover)\n",
      "Getting block 6 of 8\n",
      "  Reserving size (550810440) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 374208978 for bucket 6\n",
      "  (Using difference cover)\n"
     ]
    }
   ],
   "source": [
    "!bismark_genome_preparation --verbose ~/genomes/hg38/chromosomes/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9df2b93b-323f-453d-800b-4d3ef2dc857b",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Path to genome folder specified as: /data_gilbert/home/aarab/genomes/hg38/chromosomes/\n",
      "Aligner to be used: >> HISAT2 <<\n",
      "Writing bisulfite genomes out into a single MFA (multi FastA) file\n",
      "\n",
      "Bismark Genome Preparation - Step I: Preparing folders\n",
      "\n",
      "Bisulfite Genome Indexer version v0.23.1 (last modified: 27 Jan 2021)\n",
      "\n",
      "A directory called /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/ already exists. Already existing converted sequences and/or already existing Bowtie 2 or HISAT2) indices will be overwritten!\n",
      "\n",
      "\n",
      "Step I - Prepare genome folders - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step II: Bisulfite converting reference genome\n",
      "\n",
      "conversions performed:\n",
      "chromosome\tC->T\tG->A\n",
      "chr1\t48055043\t48111528\n",
      "chr10\t27639505\t27719976\n",
      "chr11\t27903257\t27981801\n",
      "chr12\t27092804\t27182678\n",
      "chr13\t18839192\t18933605\n",
      "chr14\t18423758\t18559033\n",
      "chr15\t17752941\t17825903\n",
      "chr16\t18172742\t18299976\n",
      "chr17\t18723944\t18851500\n",
      "chr18\t15794455\t16061651\n",
      "chr19\t13954580\t14061132\n",
      "chr2\t48318180\t48450903\n",
      "chr20\t13916133\t14094472\n",
      "chr21\t8185244\t8226381\n",
      "chr22\t9160652\t9246186\n",
      "chr3\t39233483\t39344259\n",
      "chr4\t36236976\t36331025\n",
      "chr5\t35731600\t35879674\n",
      "chr6\t33646690\t33713330\n",
      "chr7\t32317984\t32378859\n",
      "chr8\t29030173\t29103787\n",
      "chr9\t25099811\t25170662\n",
      "chrM\t5181\t2169\n",
      "chrX\t30523780\t30697741\n",
      "chrY\t5285789\t5286894\n",
      "\n",
      "Total number of conversions performed:\n",
      "C->T:\t599043897\n",
      "G->A:\t601515125\n",
      "\n",
      "Step II - Genome bisulfite conversions - completed\n",
      "\n",
      "\n",
      "Bismark Genome Preparation - Step III: Launching the HISAT2 indexer\n",
      "Please be aware that this process can - depending on genome size - take several hours!\n",
      "\n",
      "Preparing indexing of CT converted genome in /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/CT_conversion/\n",
      "Parent process: Starting to index C->T converted genome with the following command:\n",
      "\n",
      "hisat2-build  -f genome_mfa.CT_conversion.fa BS_CT\n",
      "\n",
      "Settings:\n",
      "  Output files: \"BS_CT.*.ht2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Local offset rate: 3 (one in 8)\n",
      "  Local fTable chars: 6\n",
      "  Local sequence length: 57344\n",
      "  Local sequence overlap between two consecutive indexes: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.CT_conversion.fa\n",
      "Reading reference sizes\n",
      "Preparing indexing of GA converted genome in /data_gilbert/home/aarab/genomes/hg38/chromosomes/Bisulfite_Genome/GA_conversion/\n",
      "Child process: Starting to index G->A converted genome with the following command:\n",
      "\n",
      "hisat2-build  -f genome_mfa.GA_conversion.fa BS_GA\n",
      "\n",
      "Settings:\n",
      "  Output files: \"BS_GA.*.ht2\"\n",
      "  Line rate: 6 (line is 64 bytes)\n",
      "  Lines per side: 1 (side is 64 bytes)\n",
      "  Offset rate: 4 (one in 16)\n",
      "  FTable chars: 10\n",
      "  Strings: unpacked\n",
      "  Local offset rate: 3 (one in 8)\n",
      "  Local fTable chars: 6\n",
      "  Local sequence length: 57344\n",
      "  Local sequence overlap between two consecutive indexes: 1024\n",
      "  Endianness: little\n",
      "  Actual local endianness: little\n",
      "  Sanity checking: disabled\n",
      "  Assertions: disabled\n",
      "  Random seed: 0\n",
      "  Sizeofs: void*:8, int:4, long:8, size_t:8\n",
      "Input files DNA, FASTA:\n",
      "  genome_mfa.GA_conversion.fa\n",
      "Reading reference sizes\n",
      "  Time reading reference sizes: 00:00:16\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time reading reference sizes: 00:00:18\n",
      "Calculating joined length\n",
      "Writing header\n",
      "Reserving space for joined string\n",
      "Joining reference sequences\n",
      "  Time to join reference sequences: 00:00:13\n",
      "  Time to read SNPs and splice sites: 00:00:00\n",
      "Using parameters --bmax 550810440 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Passed!  Constructing with these parameters: --bmax 550810440 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Time to join reference sequences: 00:00:13\n",
      "  Time to read SNPs and splice sites: 00:00:00\n",
      "Using parameters --bmax 550810440 --dcv 1024\n",
      "  Doing ahead-of-time memory usage test\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  Passed!  Constructing with these parameters: --bmax 550810440 --dcv 1024\n",
      "Constructing suffix-array element generator\n",
      "Building DifferenceCoverSample\n",
      "  Building sPrime\n",
      "  Building sPrimeOrder\n",
      "  V-Sorting samples\n",
      "  V-Sorting samples time: 00:01:19\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  V-Sorting samples time: 00:01:18\n",
      "  Allocating rank array\n",
      "  Ranking v-sort output\n",
      "  Ranking v-sort output time: 00:00:21\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Ranking v-sort output time: 00:00:21\n",
      "  Invoking Larsson-Sadakane on ranks\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:37\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "  Invoking Larsson-Sadakane on ranks time: 00:00:36\n",
      "  Sanity-checking and returning\n",
      "Building samples\n",
      "Reserving space for 12 sample suffixes\n",
      "Generating random suffixes\n",
      "QSorting 12 sample offsets, eliminating duplicates\n",
      "QSorting sample offsets, eliminating duplicates time: 00:00:00\n",
      "Multikey QSorting 12 samples\n",
      "  (Using difference cover)\n",
      "  Multikey QSorting samples time: 00:00:00\n",
      "Calculating bucket sizes\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 4.19665e+08 (target: 550810439)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering GFM loop\n",
      "Getting block 1 of 7\n",
      "  Reserving size (550810440) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "Splitting and merging\n",
      "  Splitting and merging time: 00:00:00\n",
      "Avg bucket size: 3.67207e+08 (target: 550810439)\n",
      "Converting suffix-array elements to index image\n",
      "Allocating ftab, absorbFtab\n",
      "Entering GFM loop\n",
      "Getting block 1 of 8\n",
      "  Reserving size (550810440) for bucket 1\n",
      "  Calculating Z arrays for bucket 1\n",
      "  Entering block accumulator loop for bucket 1:\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 10%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 20%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 30%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 40%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 50%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 60%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 70%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 80%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 90%\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 506437153 for bucket 1\n",
      "  (Using difference cover)\n",
      "  bucket 1: 100%\n",
      "  Sorting block of length 265840023 for bucket 1\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:02:54\n",
      "Returning block of 265840024 for bucket 1\n",
      "Getting block 2 of 8\n",
      "  Reserving size (550810440) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "  bucket 2: 20%\n",
      "  bucket 2: 30%\n",
      "  bucket 2: 40%\n",
      "  bucket 2: 50%\n",
      "  bucket 2: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 313712311 for bucket 2\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:07:12\n",
      "Returning block of 506437154 for bucket 1\n",
      "  Sorting block time: 00:03:38\n",
      "Returning block of 313712312 for bucket 2\n",
      "Getting block 2 of 7\n",
      "  Reserving size (550810440) for bucket 2\n",
      "  Calculating Z arrays for bucket 2\n",
      "  Entering block accumulator loop for bucket 2:\n",
      "  bucket 2: 10%\n",
      "Getting block 3 of 8\n",
      "  Reserving size (550810440) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 2: 20%\n",
      "  bucket 3: 10%\n",
      "  bucket 2: 30%\n",
      "  bucket 3: 20%\n",
      "  bucket 2: 40%\n",
      "  bucket 3: 30%\n",
      "  bucket 2: 50%\n",
      "  bucket 3: 40%\n",
      "  bucket 2: 60%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 2: 70%\n",
      "  bucket 3: 70%\n",
      "  bucket 2: 80%\n",
      "  bucket 3: 80%\n",
      "  bucket 2: 90%\n",
      "  bucket 3: 90%\n",
      "  bucket 2: 100%\n",
      "  Sorting block of length 447885829 for bucket 2\n",
      "  (Using difference cover)\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 269510198 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:03:07\n",
      "Returning block of 269510199 for bucket 3\n",
      "Getting block 4 of 8\n",
      "  Reserving size (550810440) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "  bucket 4: 30%\n",
      "  bucket 4: 40%\n",
      "  bucket 4: 50%\n",
      "  bucket 4: 60%\n",
      "  bucket 4: 70%\n",
      "  bucket 4: 80%\n",
      "  bucket 4: 90%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 416425897 for bucket 4\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:06:28\n",
      "Returning block of 447885830 for bucket 2\n",
      "Getting block 3 of 7\n",
      "  Reserving size (550810440) for bucket 3\n",
      "  Calculating Z arrays for bucket 3\n",
      "  Entering block accumulator loop for bucket 3:\n",
      "  bucket 3: 10%\n",
      "  bucket 3: 20%\n",
      "  bucket 3: 30%\n",
      "  bucket 3: 40%\n",
      "  bucket 3: 50%\n",
      "  bucket 3: 60%\n",
      "  bucket 3: 70%\n",
      "  bucket 3: 80%\n",
      "  bucket 3: 90%\n",
      "  bucket 3: 100%\n",
      "  Sorting block of length 539608788 for bucket 3\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:04:54\n",
      "Returning block of 416425898 for bucket 4\n",
      "Getting block 5 of 8\n",
      "  Reserving size (550810440) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 480655197 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:08:17\n",
      "Returning block of 539608789 for bucket 3\n",
      "  Sorting block time: 00:05:46\n",
      "Returning block of 480655198 for bucket 5\n",
      "Getting block 4 of 7\n",
      "  Reserving size (550810440) for bucket 4\n",
      "  Calculating Z arrays for bucket 4\n",
      "  Entering block accumulator loop for bucket 4:\n",
      "  bucket 4: 10%\n",
      "  bucket 4: 20%\n",
      "Getting block 6 of 8\n",
      "  Reserving size (550810440) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 4: 30%\n",
      "  bucket 6: 10%\n",
      "  bucket 4: 40%\n",
      "  bucket 6: 20%\n",
      "  bucket 4: 50%\n",
      "  bucket 6: 30%\n",
      "  bucket 4: 60%\n",
      "  bucket 6: 40%\n",
      "  bucket 4: 70%\n",
      "  bucket 6: 50%\n",
      "  bucket 4: 80%\n",
      "  bucket 6: 60%\n",
      "  bucket 4: 90%\n",
      "  bucket 6: 70%\n",
      "  bucket 4: 100%\n",
      "  Sorting block of length 384792641 for bucket 4\n",
      "  (Using difference cover)\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 367604136 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:04:21\n",
      "Returning block of 367604137 for bucket 6\n",
      "Getting block 7 of 8\n",
      "  Reserving size (550810440) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  Sorting block time: 00:05:36\n",
      "Returning block of 384792642 for bucket 4\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 547961096 for bucket 7\n",
      "  (Using difference cover)\n",
      "Getting block 5 of 7\n",
      "  Reserving size (550810440) for bucket 5\n",
      "  Calculating Z arrays for bucket 5\n",
      "  Entering block accumulator loop for bucket 5:\n",
      "  bucket 5: 10%\n",
      "  bucket 5: 20%\n",
      "  bucket 5: 30%\n",
      "  bucket 5: 40%\n",
      "  bucket 5: 50%\n",
      "  bucket 5: 60%\n",
      "  bucket 5: 70%\n",
      "  bucket 5: 80%\n",
      "  bucket 5: 90%\n",
      "  bucket 5: 100%\n",
      "  Sorting block of length 459841719 for bucket 5\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:06:31\n",
      "Returning block of 547961097 for bucket 7\n",
      "Getting block 8 of 8\n",
      "  Reserving size (550810440) for bucket 8\n",
      "  Calculating Z arrays for bucket 8\n",
      "  Entering block accumulator loop for bucket 8:\n",
      "  bucket 8: 10%\n",
      "  bucket 8: 20%\n",
      "  Sorting block time: 00:06:57\n",
      "Returning block of 459841720 for bucket 5\n",
      "  bucket 8: 30%\n",
      "  bucket 8: 40%\n",
      "  bucket 8: 50%\n",
      "  bucket 8: 60%\n",
      "  bucket 8: 70%\n",
      "  bucket 8: 80%\n",
      "  bucket 8: 90%\n",
      "  bucket 8: 100%\n",
      "  Sorting block of length 275946816 for bucket 8\n",
      "  (Using difference cover)\n",
      "Getting block 6 of 7\n",
      "  Reserving size (550810440) for bucket 6\n",
      "  Calculating Z arrays for bucket 6\n",
      "  Entering block accumulator loop for bucket 6:\n",
      "  bucket 6: 10%\n",
      "  bucket 6: 20%\n",
      "  bucket 6: 30%\n",
      "  bucket 6: 40%\n",
      "  bucket 6: 50%\n",
      "  bucket 6: 60%\n",
      "  bucket 6: 70%\n",
      "  bucket 6: 80%\n",
      "  bucket 6: 90%\n",
      "  bucket 6: 100%\n",
      "  Sorting block of length 328692208 for bucket 6\n",
      "  (Using difference cover)\n",
      "  Sorting block time: 00:03:16\n",
      "Returning block of 275946817 for bucket 8\n",
      "Exited GFM loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 1468669118\n",
      "fchr[G]: 2067713015\n",
      "fchr[T]: 2067713015\n",
      "fchr[$]: 2937655681\n",
      "Exiting GFM::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 983425404 bytes to primary GFM file: BS_GA.1.ht2\n",
      "Wrote 734413928 bytes to secondary GFM file: BS_GA.2.ht2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from GFM constructor\n",
      "  Sorting block time: 00:04:58\n",
      "Returning block of 328692209 for bucket 6\n",
      "Getting block 7 of 7\n",
      "  Reserving size (550810440) for bucket 7\n",
      "  Calculating Z arrays for bucket 7\n",
      "  Entering block accumulator loop for bucket 7:\n",
      "  bucket 7: 10%\n",
      "  bucket 7: 20%\n",
      "  bucket 7: 30%\n",
      "  bucket 7: 40%\n",
      "  bucket 7: 50%\n",
      "  bucket 7: 60%\n",
      "  bucket 7: 70%\n",
      "  bucket 7: 80%\n",
      "  bucket 7: 90%\n",
      "  bucket 7: 100%\n",
      "  Sorting block of length 270397337 for bucket 7\n",
      "  (Using difference cover)\n",
      "Returning from initFromVector\n",
      "Wrote 1290699801 bytes to primary GFM file: BS_GA.5.ht2\n",
      "Wrote 747869486 bytes to secondary GFM file: BS_GA.6.ht2\n",
      "Re-opening _in5 and _in5 as input streams\n",
      "Returning from HGFM constructor\n",
      "Headers:\n",
      "    len: 2937655681\n",
      "    gbwtLen: 2937655682\n",
      "    nodes: 2937655682\n",
      "    sz: 734413921\n",
      "    gbwtSz: 734413921\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 0\n",
      "    eftabSz: 0\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 183603481\n",
      "    offsSz: 734413924\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideGbwtSz: 48\n",
      "    sideGbwtLen: 192\n",
      "    numSides: 15300291\n",
      "    numLines: 15300291\n",
      "    gbwtTotLen: 979218624\n",
      "    gbwtTotSz: 979218624\n",
      "    reverse: 0\n",
      "    linearFM: Yes\n",
      "Total time for call to driver() for forward index: 00:58:15\n",
      "  Sorting block time: 00:04:26\n",
      "Returning block of 270397338 for bucket 7\n",
      "Exited GFM loop\n",
      "fchr[A]: 0\n",
      "fchr[C]: 867153993\n",
      "fchr[G]: 867153993\n",
      "fchr[T]: 1468669118\n",
      "fchr[$]: 2937655681\n",
      "Exiting GFM::buildToDisk()\n",
      "Returning from initFromVector\n",
      "Wrote 983425404 bytes to primary GFM file: BS_CT.1.ht2\n",
      "Wrote 734413928 bytes to secondary GFM file: BS_CT.2.ht2\n",
      "Re-opening _in1 and _in2 as input streams\n",
      "Returning from GFM constructor\n",
      "Returning from initFromVector\n",
      "Wrote 1290699801 bytes to primary GFM file: BS_CT.5.ht2\n",
      "Wrote 747869486 bytes to secondary GFM file: BS_CT.6.ht2\n",
      "Re-opening _in5 and _in5 as input streams\n",
      "Returning from HGFM constructor\n",
      "Headers:\n",
      "    len: 2937655681\n",
      "    gbwtLen: 2937655682\n",
      "    nodes: 2937655682\n",
      "    sz: 734413921\n",
      "    gbwtSz: 734413921\n",
      "    lineRate: 6\n",
      "    offRate: 4\n",
      "    offMask: 0xfffffff0\n",
      "    ftabChars: 10\n",
      "    eftabLen: 0\n",
      "    eftabSz: 0\n",
      "    ftabLen: 1048577\n",
      "    ftabSz: 4194308\n",
      "    offsLen: 183603481\n",
      "    offsSz: 734413924\n",
      "    lineSz: 64\n",
      "    sideSz: 64\n",
      "    sideGbwtSz: 48\n",
      "    sideGbwtLen: 192\n",
      "    numSides: 15300291\n",
      "    numLines: 15300291\n",
      "    gbwtTotLen: 979218624\n",
      "    gbwtTotSz: 979218624\n",
      "    reverse: 0\n",
      "    linearFM: Yes\n",
      "Total time for call to driver() for forward index: 01:07:19\n"
     ]
    }
   ],
   "source": [
    "!bismark_genome_preparation --verbose --hisat2 ~/genomes/hg38/chromosomes/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "31c09197-9296-4c42-8468-5427b9bb9166",
   "metadata": {},
   "source": [
    "### Analyze samples"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "28fb90c5-5e40-43b4-9dee-3ca6d824a5b6",
   "metadata": {},
   "source": [
    "__(II) Running bismark__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d22b24b0-7772-4e0e-9268-7e1c93a9923e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bismark --gzip --genome /data_gilbert/home/aarab/genomes/hg38/chromosomes/ -1 SRR11711272_1_val_1.fq.gz -2 SRR11711272_2_val_2.fq.gz\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "genDIR=\"/data_gilbert/home/aarab/genomes/hg38/chromosomes/\"\n",
    "\n",
    "cd fastq\n",
    "for fq1 in *_val_1.fq.gz; do \n",
    "    fq2=${fq1/_1_val_1.fq/_2_val_2.fq}\n",
    "    b=`basename $fq1`\n",
    "    sample=${b/_1_val_1.fq.gz/}\n",
    "    cm=\"bismark --gzip --genome $genDIR -1 $fq1 -2 $fq2\"\n",
    "    # --score_min L,0,-0.4\n",
    "    echo $cm\n",
    "    $cm &> ${sample}.log;\n",
    "    wait\n",
    "done\n",
    "cd ../"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d3699bee-f8eb-4146-a207-0fd463154f3c",
   "metadata": {},
   "source": [
    "__(III) Running deduplicate_bismark__\n",
    "> should not be used for reduced representation libraries such as RRBS, amplicon or target enrichment libraries."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e9cf7429-1d78-44da-9485-92d0926d1c6d",
   "metadata": {},
   "source": [
    "however, \n",
    "\n",
    "https://pubmed.ncbi.nlm.nih.gov/33958785/\n",
    "\n",
    "> Alignments were futher filtered to only include reads that are uniquely aligned, map to an MspI cut site, and do not represent PCR duplicates (taking into account the hexamer sequence as a unique molecular identifier)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "896eac02-c873-4b64-83e9-891c9361c56e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Please provide one or more Bismark output files for deduplication\n",
      "\n",
      "\n",
      "===============================================================================================================\n",
      "\n",
      "This script is supposed to remove alignments to the same position in the genome from the Bismark mapping output\n",
      "(both single and paired-end SAM/BAM files), which can arise by e.g. excessive PCR amplification. If sequences align\n",
      "to the same genomic position but on different strands they will be scored individually.\n",
      "\n",
      "Note that deduplication is not recommended for RRBS-type experiments!\n",
      "\n",
      "In the default mode, the first alignment to a given position will be used irrespective of its methylation call\n",
      "(as the alignments are not ordered in any way this is also near enough random).\n",
      "\n",
      "For single-end alignments only the chromosome, start coordinate and strand of a read will be used for deduplication.\n",
      "\n",
      "For paired-end alignments the chromosome, strand alignments, and start as well as end coordinate of the entire fragment\n",
      "will be used for deduplication. >This script expects the Bismark output to be in SAM format<.\n",
      "\n",
      "*** Please note that for paired-end BAM files the deduplication script expects Read1 and Read2 to\n",
      "follow each other in consecutive lines! If the file has been sorted by position make sure that you resort it\n",
      "by read name first (e.g. using samtools sort -n)  ***\n",
      "\n",
      "===============================================================================================================\n",
      "\n",
      ">>> USAGE: ./deduplicate_bismark [options] filename(s) <<<\n",
      "\n",
      "-s/--single\t\tdeduplicate single-end BAM/SAM Bismark files. Default: [AUTO-DETECT]\n",
      "-p/--paired\t\tdeduplicate paired-end BAM/SAM Bismark files. Default: [AUTO-DETECT]\n",
      "\n",
      "-o/--outfile [filename]\tThe basename of a desired output file. This basename is modified to end in '.deduplicated.bam',\n",
      "\t\t\tor '.multiple.deduplicated.bam' in '--multiple' mode, for consistency reasons.\n",
      "\n",
      "--output_dir [path]\tOutput directory, either relative or absolute. Output is written to the current directory if not\n",
      "\t\t\tspecified explicitly.\n",
      "\n",
      "--barcode\t\tIn addition to chromosome, start position and orientation this will also take a potential barcode into\n",
      "\t\t\tconsideration while deduplicating. The barcode needs to be the last element of the read ID and separated\n",
      "                        by a ':', e.g.: MISEQ:14:000000000-A55D0:1:1101:18024:2858_1:N:0:CTCCT\n",
      "\n",
      "--bam\t\t\tThe output will be written out in BAM format. This script will attempt to use the path to Samtools\n",
      "\t\t\tthat was specified with '--samtools_path', or, if it hasn't been specified, attempt to find Samtools\n",
      "\t\t\tin the PATH. If no installation of Samtools can be found, a GZIP compressed output is written out\n",
      "\t\t\tinstead (yielding a .sam.gz output file). Default: ON.\n",
      "\n",
      "--sam\t\t\tThe output will be written out in SAM format. Default: OFF.\n",
      "\n",
      "--multiple\t\tAll specified input files are treated as one sample and concatenated together for deduplication.\n",
      "\t\t\tThis uses Unix 'cat' for SAM files and 'samtools cat' for BAM files. Additional notes for BAM files:\n",
      "\t\t\tAlthough this works on either BAM or CRAM, all input files must be the same format as each other.\n",
      "\t\t\tThe sequence dictionary of each input file must be identical, although this command does not check this.\n",
      "\t\t\tBy default the header is taken from the first file to be concatenated.\n",
      "\n",
      "--samtools_path [path]\tThe path to your Samtools installation, e.g. /home/user/samtools/. Does not need to be specified\n",
      "\t\t\texplicitly if Samtools is in the PATH already\n",
      "\n",
      "--version\t\tPrint version information and exit\n",
      "===============================================================================================================\n",
      "\n",
      "This script was last modified on 30 October 2020\n",
      "\n"
     ]
    }
   ],
   "source": [
    "!deduplicate_bismark"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "43c346fd-9c65-41b0-8539-99d106848fc5",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "@SRR11711272.1 NS500400:843:HVTWWBGXC:1:11101:3311:1050:TCGCTAGA+GGAGAGTA:GGATGGTT+ACTCTN/1\n",
      "CGGGATATTTGGTTGTTNNAAATAATATG\n",
      "+\n",
      "EA6EEEEEEEEEEAEEE##EEEEEEEEEE\n",
      "@SRR11711272.2 NS500400:843:HVTWWBGXC:1:11101:11560:1050:TCGCTAGA+GGAGAGTA:GGATGGTT+CCCCAN/1\n",
      "TGGTTGATTGTTTAGTTNNATGTAGGTTA\n",
      "+\n",
      "EEAEEEEEEEEEEEAEE##/EEE/EEEEE\n",
      "@SRR11711272.3 NS500400:843:HVTWWBGXC:1:11101:2495:1050:TCGCTAGA+GGAGAGTA:GGATGGTT+CAACCN/1\n",
      "CGGTTATTATTTTAGTTNNGATTAGAATTTAT\n",
      "\n",
      "gzip: stdout: Broken pipe\n"
     ]
    }
   ],
   "source": [
    "!zcat bismark/SRR11711272_1_val_1.fq.gz | head"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "aec8408f-0734-405f-b624-ad1fba8e6423",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bismark/SRR11711272_1_val_1_bismark_bt2_pe.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Neither -s (single-end) nor -p (paired-end) selected for deduplication. Trying to extract this information for each file separately from the @PG line of the SAM/BAM file\n",
      "Processing single-end Bismark output file(s) (SAM format):\n",
      "bismark/SRR11711272_1_val_1_bismark_bt2_pe.bam\n",
      "\n",
      "\n",
      "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
      "\n",
      "Checking file >>bismark/SRR11711272_1_val_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
      "\n",
      "Trying to determine the type of mapping from the SAM header line\n",
      "Treating file as paired-end data (extracted from @PG line)\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...passed!\n",
      "Output file is: SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Total number of alignments analysed in bismark/SRR11711272_1_val_1_bismark_bt2_pe.bam:\t24516791\n",
      "Total number duplicated alignments removed:\t3437275 (14.02%)\n",
      "Duplicated alignments were found at:\t2529677 different position(s)\n",
      "\n",
      "Total count of deduplicated leftover sequences: 21079516 (85.98% of total)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now testing Bismark result file bismark/SRR11711272_1_val_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\tskipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
      "skipping header line:\t@SQ\tSN:chr1\tLN:248956422\n",
      "skipping header line:\t@SQ\tSN:chr10\tLN:133797422\n",
      "skipping header line:\t@SQ\tSN:chr11\tLN:135086622\n",
      "skipping header line:\t@SQ\tSN:chr12\tLN:133275309\n",
      "skipping header line:\t@SQ\tSN:chr13\tLN:114364328\n",
      "skipping header line:\t@SQ\tSN:chr14\tLN:107043718\n",
      "skipping header line:\t@SQ\tSN:chr15\tLN:101991189\n",
      "skipping header line:\t@SQ\tSN:chr16\tLN:90338345\n",
      "skipping header line:\t@SQ\tSN:chr17\tLN:83257441\n",
      "skipping header line:\t@SQ\tSN:chr18\tLN:80373285\n",
      "skipping header line:\t@SQ\tSN:chr19\tLN:58617616\n",
      "skipping header line:\t@SQ\tSN:chr2\tLN:242193529\n",
      "skipping header line:\t@SQ\tSN:chr20\tLN:64444167\n",
      "skipping header line:\t@SQ\tSN:chr21\tLN:46709983\n",
      "skipping header line:\t@SQ\tSN:chr22\tLN:50818468\n",
      "skipping header line:\t@SQ\tSN:chr3\tLN:198295559\n",
      "skipping header line:\t@SQ\tSN:chr4\tLN:190214555\n",
      "skipping header line:\t@SQ\tSN:chr5\tLN:181538259\n",
      "skipping header line:\t@SQ\tSN:chr6\tLN:170805979\n",
      "skipping header line:\t@SQ\tSN:chr7\tLN:159345973\n",
      "skipping header line:\t@SQ\tSN:chr8\tLN:145138636\n",
      "skipping header line:\t@SQ\tSN:chr9\tLN:138394717\n",
      "skipping header line:\t@SQ\tSN:chrM\tLN:16569\n",
      "skipping header line:\t@SQ\tSN:chrX\tLN:156040895\n",
      "skipping header line:\t@SQ\tSN:chrY\tLN:57227415\n",
      "skipping header line:\t@PG\tID:Bismark\tVN:v0.23.1\tCL:\"bismark --gzip --genome /data_gilbert/home/aarab/genomes/hg38/chromosomes/ -1 SRR11711272_1_val_1.fq.gz -2 SRR11711272_2_val_2.fq.gz\"\n",
      "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -bSh -\n",
      "skipping header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -h --threads 1 bismark/SRR11711272_1_val_1_bismark_bt2_pe.bam\n",
      "bismark/SRR11711273_1_val_1_bismark_bt2_pe.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Neither -s (single-end) nor -p (paired-end) selected for deduplication. Trying to extract this information for each file separately from the @PG line of the SAM/BAM file\n",
      "Processing single-end Bismark output file(s) (SAM format):\n",
      "bismark/SRR11711273_1_val_1_bismark_bt2_pe.bam\n",
      "\n",
      "\n",
      "If there are several alignments to a single position in the genome the first alignment will be chosen. Since the input files are not in any way sorted this is a near-enough random selection of reads.\n",
      "\n",
      "Checking file >>bismark/SRR11711273_1_val_1_bismark_bt2_pe.bam<< for signs of file truncation...\n",
      "\n",
      "Trying to determine the type of mapping from the SAM header line\n",
      "Treating file as paired-end data (extracted from @PG line)\n",
      "\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...passed!\n",
      "Output file is: SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Total number of alignments analysed in bismark/SRR11711273_1_val_1_bismark_bt2_pe.bam:\t6784977\n",
      "Total number duplicated alignments removed:\t379032 (5.59%)\n",
      "Duplicated alignments were found at:\t309602 different position(s)\n",
      "\n",
      "Total count of deduplicated leftover sequences: 6405945 (94.41% of total)\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now testing Bismark result file bismark/SRR11711273_1_val_1_bismark_bt2_pe.bam for positional sorting (which would be bad...)\tskipping header line:\t@HD\tVN:1.0\tSO:unsorted\n",
      "skipping header line:\t@SQ\tSN:chr1\tLN:248956422\n",
      "skipping header line:\t@SQ\tSN:chr10\tLN:133797422\n",
      "skipping header line:\t@SQ\tSN:chr11\tLN:135086622\n",
      "skipping header line:\t@SQ\tSN:chr12\tLN:133275309\n",
      "skipping header line:\t@SQ\tSN:chr13\tLN:114364328\n",
      "skipping header line:\t@SQ\tSN:chr14\tLN:107043718\n",
      "skipping header line:\t@SQ\tSN:chr15\tLN:101991189\n",
      "skipping header line:\t@SQ\tSN:chr16\tLN:90338345\n",
      "skipping header line:\t@SQ\tSN:chr17\tLN:83257441\n",
      "skipping header line:\t@SQ\tSN:chr18\tLN:80373285\n",
      "skipping header line:\t@SQ\tSN:chr19\tLN:58617616\n",
      "skipping header line:\t@SQ\tSN:chr2\tLN:242193529\n",
      "skipping header line:\t@SQ\tSN:chr20\tLN:64444167\n",
      "skipping header line:\t@SQ\tSN:chr21\tLN:46709983\n",
      "skipping header line:\t@SQ\tSN:chr22\tLN:50818468\n",
      "skipping header line:\t@SQ\tSN:chr3\tLN:198295559\n",
      "skipping header line:\t@SQ\tSN:chr4\tLN:190214555\n",
      "skipping header line:\t@SQ\tSN:chr5\tLN:181538259\n",
      "skipping header line:\t@SQ\tSN:chr6\tLN:170805979\n",
      "skipping header line:\t@SQ\tSN:chr7\tLN:159345973\n",
      "skipping header line:\t@SQ\tSN:chr8\tLN:145138636\n",
      "skipping header line:\t@SQ\tSN:chr9\tLN:138394717\n",
      "skipping header line:\t@SQ\tSN:chrM\tLN:16569\n",
      "skipping header line:\t@SQ\tSN:chrX\tLN:156040895\n",
      "skipping header line:\t@SQ\tSN:chrY\tLN:57227415\n",
      "skipping header line:\t@PG\tID:Bismark\tVN:v0.23.1\tCL:\"bismark --gzip --genome /data_gilbert/home/aarab/genomes/hg38/chromosomes/ -1 SRR11711273_1_val_1.fq.gz -2 SRR11711273_2_val_2.fq.gz\"\n",
      "skipping header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -bSh -\n",
      "skipping header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -h --threads 1 bismark/SRR11711273_1_val_1_bismark_bt2_pe.bam\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "for bam in bismark/*_bt2_pe.bam; do \n",
    "    echo $bam\n",
    "    deduplicate_bismark --bam $bam\n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "141bb633-e407-47a6-b04b-fca027425cb8",
   "metadata": {},
   "source": [
    "__(IV) Running bismark_methylation_extractor__\n",
    "extract context-dependent (CpG/CHG/CHH) methylation\n",
    "\n",
    "This will produce three methytlation output files:\n",
    "\n",
    "    CpG_context_test_dataset_bismark_bt2.txt.gz\n",
    "    CHG_context_test_dataset_bismark_bt2.txt.gz\n",
    "    CHH_context_test_dataset_bismark_bt2.txt.gz\n",
    "as well as a bedGraph and a Bismark coverage file. For more on these files and their formats please see below."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c3f24b98-d37d-43f6-935d-dbf529eac025",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      " *** Bismark methylation extractor version v0.23.1 ***\n",
      "\n",
      "Trying to determine the type of mapping from the SAM header line of file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "Treating file(s) as paired-end data (as extracted from @PG line)\n",
      "\n",
      "Setting option '--no_overlap' since this is (normally) the right thing to do for paired-end data\n",
      "\n",
      "Summarising Bismark methylation extractor parameters:\n",
      "===============================================================\n",
      "Bismark paired-end SAM format specified (default)\n",
      "Number of cores to be used: 10\n",
      "Output will be written to the current directory ('/data_gilbert/home/aarab/Projects/Decitabine-treatment/DAC/Bisulfite-seq/bismark')\n",
      "\n",
      "\n",
      "Summarising bedGraph parameters:\n",
      "===============================================================\n",
      "Generating additional output in bedGraph and coverage format\n",
      "bedGraph format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage>\n",
      "coverage format:\t<Chromosome> <Start Position> <End Position> <Methylation Percentage> <count methylated> <count non-methylated>\n",
      "\n",
      "Using a cutoff of 1 read(s) to report cytosine positions\n",
      "Reporting and sorting cytosine methylation information in CpG context only (default)\n",
      "The bedGraph UNIX sort command will use the following memory setting:\t'2G'. Temporary directory used for sorting is the output directory\n",
      "\n",
      "Checking file >>SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam<< for signs of file truncation...\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "...passed!\n",
      "Writing result file containing methylation information for C in CpG context from the original top strand to CpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CpG context from the complementary to original top strand to CpG_CTOT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CpG context from the complementary to original bottom strand to CpG_CTOB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CpG context from the original bottom strand to CpG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "\n",
      "Writing result file containing methylation information for C in CHG context from the original top strand to CHG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CHG context from the complementary to original top strand to CHG_CTOT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CHG context from the complementary to original bottom strand to CHG_CTOB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CHG context from the original bottom strand to CHG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "\n",
      "Writing result file containing methylation information for C in CHH context from the original top strand to CHH_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CHH context from the complementary to original top strand to CHH_CTOT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CHH context from the complementary to original bottom strand to CHH_CTOB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Writing result file containing methylation information for C in CHH context from the original bottom strand to CHH_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "Now reading in Bismark result file SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\n",
      "skipping SAM header line:\t@HD\tVN:1.0\tSO:unsorted\n",
      "skipping SAM header line:\t@SQ\tSN:chr1\tLN:248956422\n",
      "skipping SAM header line:\t@SQ\tSN:chr10\tLN:133797422\n",
      "skipping SAM header line:\t@SQ\tSN:chr11\tLN:135086622\n",
      "skipping SAM header line:\t@SQ\tSN:chr12\tLN:133275309\n",
      "skipping SAM header line:\t@SQ\tSN:chr13\tLN:114364328\n",
      "skipping SAM header line:\t@SQ\tSN:chr14\tLN:107043718\n",
      "skipping SAM header line:\t@SQ\tSN:chr15\tLN:101991189\n",
      "skipping SAM header line:\t@SQ\tSN:chr16\tLN:90338345\n",
      "skipping SAM header line:\t@SQ\tSN:chr17\tLN:83257441\n",
      "skipping SAM header line:\t@SQ\tSN:chr18\tLN:80373285\n",
      "skipping SAM header line:\t@SQ\tSN:chr19\tLN:58617616\n",
      "skipping SAM header line:\t@SQ\tSN:chr2\tLN:242193529\n",
      "skipping SAM header line:\t@SQ\tSN:chr20\tLN:64444167\n",
      "skipping SAM header line:\t@SQ\tSN:chr21\tLN:46709983\n",
      "skipping SAM header line:\t@SQ\tSN:chr22\tLN:50818468\n",
      "skipping SAM header line:\t@SQ\tSN:chr3\tLN:198295559\n",
      "skipping SAM header line:\t@SQ\tSN:chr4\tLN:190214555\n",
      "skipping SAM header line:\t@SQ\tSN:chr5\tLN:181538259\n",
      "skipping SAM header line:\t@SQ\tSN:chr6\tLN:170805979\n",
      "skipping SAM header line:\t@SQ\tSN:chr7\tLN:159345973\n",
      "skipping SAM header line:\t@SQ\tSN:chr8\tLN:145138636\n",
      "skipping SAM header line:\t@SQ\tSN:chr9\tLN:138394717\n",
      "skipping SAM header line:\t@SQ\tSN:chrM\tLN:16569\n",
      "skipping SAM header line:\t@SQ\tSN:chrX\tLN:156040895\n",
      "skipping SAM header line:\t@SQ\tSN:chrY\tLN:57227415\n",
      "skipping SAM header line:\t@PG\tID:Bismark\tVN:v0.23.1\tCL:\"bismark --gzip --genome /data_gilbert/home/aarab/genomes/hg38/chromosomes/ -1 SRR11711272_1_val_1.fq.gz -2 SRR11711272_2_val_2.fq.gz\"\n",
      "skipping SAM header line:\t@PG\tID:samtools\tPN:samtools\tPP:Bismark\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -bSh -\n",
      "skipping SAM header line:\t@PG\tID:samtools.1\tPN:samtools\tPP:samtools\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -h --threads 1 bismark/SRR11711272_1_val_1_bismark_bt2_pe.bam\n",
      "skipping SAM header line:\t@PG\tID:samtools.2\tPN:samtools\tPP:samtools.1\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -bS --threads 1 -\n",
      "skipping SAM header line:\t@PG\tID:samtools.3\tPN:samtools\tPP:samtools.2\tVN:1.15\tCL:/data_gilbert/home/aarab/anaconda3/envs/DNAme/bin/samtools view -h SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 500000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1000000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 1500000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2000000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 2500000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3000000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 3500000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4000000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 4500000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5000000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 5500000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6000000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 6500000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7000000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 7500000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8000000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 8500000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9000000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 9500000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10000000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 10500000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11000000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 11500000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12000000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 12500000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13000000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 13500000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14000000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 14500000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15000000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 15500000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16000000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 16500000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17000000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 17500000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18000000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 18500000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19000000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 19500000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20000000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 20500000\n",
      "Processed lines: 21000000\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 21000000\n",
      "Processed lines: 21000000\n",
      "Now waiting for all child processes to complete\n",
      "Processed lines: 21000000\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 21000000\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Finished processing child process. Exiting..\n",
      "Processed lines: 21000000\n",
      "Finished processing child process. Exiting..\n",
      "\n",
      "Merging individual splitting reports into overall report: 'SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt'\n",
      "Merging from these individual files:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Now testing Bismark result file >SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam< for positional sorting (which would be bad...)\tSRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.1\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.2\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.3\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.4\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.5\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.6\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.7\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.8\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.9\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.10\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Processed 21079516 lines in total\n",
      "Total number of methylation call strings processed: 42159032\n",
      "\n",
      "Final Cytosine Methylation Report\n",
      "=================================\n",
      "Total number of C's analysed:\t292164183\n",
      "\n",
      "Total methylated C's in CpG context:\t25968503\n",
      "Total methylated C's in CHG context:\t784307\n",
      "Total methylated C's in CHH context:\t2155710\n",
      "\n",
      "Total C to T conversions in CpG context:\t15713878\n",
      "Total C to T conversions in CHG context:\t65131428\n",
      "Total C to T conversions in CHH context:\t182410357\n",
      "\n",
      "C methylated in CpG context:\t62.3%\n",
      "C methylated in CHG context:\t1.2%\n",
      "C methylated in CHH context:\t1.2%\n",
      "\n",
      "\n",
      "\n",
      "Merging individual M-bias reports into overall M-bias statistics from these 10 individual files:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.1.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.2.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.3.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.4.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.5.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.6.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.7.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.8.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.9.mbias\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt.10.mbias\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Determining maximum read lengths for M-Bias plots\n",
      "Maximum read length of Read 1: 32\n",
      "Maximum read length of Read 2: 28\n",
      "\n",
      "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
      "Determining maximum read lengths for M-Bias plots\n",
      "Maximum read length of Read 1: 32\n",
      "Maximum read length of Read 2: 28\n",
      "\n",
      "Perl module GD::Graph::lines is not installed, skipping drawing M-bias plots (only writing out M-bias plot table)\n",
      "Deleting unused files ...\n",
      "\n",
      "CpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
      "CpG_CTOT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz was empty ->\tdeleted\n",
      "CpG_CTOB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz was empty ->\tdeleted\n",
      "CpG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
      "CHG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
      "CHG_CTOT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz was empty ->\tdeleted\n",
      "CHG_CTOB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz was empty ->\tdeleted\n",
      "CHG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
      "CHH_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
      "CHH_CTOT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz was empty ->\tdeleted\n",
      "CHH_CTOB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz was empty ->\tdeleted\n",
      "CHH_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz contains data ->\tkept\n",
      "\n",
      "\n",
      "Using these input files: CpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz CpG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz CHG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz CHH_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "\n",
      "Summary of parameters for bismark2bedGraph conversion:\n",
      "======================================================\n",
      "bedGraph output:\t\tSRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
      "output directory:\t\t><\n",
      "remove whitespaces:\t\tno\n",
      "CX context:\t\t\tno (CpG context only, default)\n",
      "No-header selected:\t\tno\n",
      "Sorting method:\t\t\tUnix sort-based (smaller memory footprint, but slower)\n",
      "Sort buffer size:\t\t2G\n",
      "Coverage threshold:\t\t1\n",
      "=============================================================================\n",
      "Methylation information will now be written into a bedGraph and coverage file\n",
      "=============================================================================\n",
      "\n",
      "Using the following files as Input:\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data_gilbert/home/aarab/Projects/Decitabine-treatment/DAC/Bisulfite-seq/bismark/CpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\t/data_gilbert/home/aarab/Projects/Decitabine-treatment/DAC/Bisulfite-seq/bismark/CpG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Writing bedGraph to file: SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\n",
      "Also writing out a coverage file including counts methylated and unmethylated residues to file: SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\n",
      "\n",
      "Now writing methylation information for file >>CpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n",
      "Finished writing out individual chromosome files for CpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\n",
      "Now writing methylation information for file >>CpG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz<< to individual files for each chromosome\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd bismark/\n",
    "for bam in *deduplicated.bam; do \n",
    "    bismark_methylation_extractor --multicore 10 --gzip --bedGraph $bam\n",
    "done\n",
    "cd ../"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "552074a1-1ba9-4994-ae7d-9215704265f8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "bismark_summary_report.html\n",
      "bismark_summary_report.txt\n",
      "\u001b[0m\u001b[38;5;9mCHG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHG_OB_SRR11711272_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHG_OB_SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHG_OB_SRR11711273_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHG_OT_SRR11711272_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHG_OT_SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHG_OT_SRR11711273_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OB_SRR11711272_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OB_SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OB_SRR11711273_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OT_SRR11711272_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OT_SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCHH_OT_SRR11711273_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OB_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OB_SRR11711272_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OB_SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OB_SRR11711273_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OT_SRR11711272_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OT_SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mCpG_OT_SRR11711273_1_val_1_bismark_bt2_pe.txt.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711272_1.fastq.gz\u001b[0m\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.bam\n",
      "\u001b[38;5;9mSRR11711272_1_val_1_bismark_bt2_pe.bedGraph.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711272_1_val_1_bismark_bt2_pe.bismark.cov.gz\u001b[0m\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\u001b[38;5;9mSRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\u001b[0m\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.M-bias.txt\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.deduplication_report.txt\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.M-bias.txt\n",
      "SRR11711272_1_val_1_bismark_bt2_PE_report.txt\n",
      "SRR11711272_1_val_1_bismark_bt2_pe_splitting_report.txt\n",
      "SRR11711272_1_val_1_bismark_bt2_pe.srt.bam\n",
      "\u001b[38;5;9mSRR11711272_1_val_1.fq.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711272_2.fastq.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711272_2_val_2.fq.gz\u001b[0m\n",
      "SRR11711272.log\n",
      "\u001b[38;5;9mSRR11711273_1.fastq.gz\u001b[0m\n",
      "SRR11711273_1_val_1_bismark_bt2_pe.bam\n",
      "\u001b[38;5;9mSRR11711273_1_val_1_bismark_bt2_pe.bedGraph.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711273_1_val_1_bismark_bt2_pe.bismark.cov.gz\u001b[0m\n",
      "SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.bam\n",
      "\u001b[38;5;9mSRR11711273_1_val_1_bismark_bt2_pe.deduplicated.bedGraph.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711273_1_val_1_bismark_bt2_pe.deduplicated.bismark.cov.gz\u001b[0m\n",
      "SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.M-bias.txt\n",
      "SRR11711273_1_val_1_bismark_bt2_pe.deduplicated_splitting_report.txt\n",
      "SRR11711273_1_val_1_bismark_bt2_pe.deduplication_report.txt\n",
      "SRR11711273_1_val_1_bismark_bt2_pe.M-bias.txt\n",
      "SRR11711273_1_val_1_bismark_bt2_PE_report.txt\n",
      "SRR11711273_1_val_1_bismark_bt2_pe_splitting_report.txt\n",
      "SRR11711273_1_val_1_bismark_bt2_pe.srt.bam\n",
      "\u001b[38;5;9mSRR11711273_1_val_1.fq.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711273_2.fastq.gz\u001b[0m\n",
      "\u001b[38;5;9mSRR11711273_2_val_2.fq.gz\u001b[0m\n",
      "SRR11711273.log\n"
     ]
    }
   ],
   "source": [
    "ls bismark/"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e36995cc-7017-4031-9886-bd95ccd1496d",
   "metadata": {},
   "source": [
    "<!-- __(V) Running bismark2report__ -->\n",
    "\n",
    "__(VI) Running bismark2summary__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "a3344109-f982-4555-8e2c-8874b5714c19",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "No Bismark/Bowtie2 single-end BAM files detected\n",
      "Found Bismark/Bowtie2 paired-end files\n",
      "No Bismark/HISAT2 single-end BAM files detected\n",
      "No Bismark/HISAT2 paired-end BAM files detected\n",
      "\n",
      "Generating Bismark summary report from 2 Bismark BAM file(s)...\n",
      ">> Reading from Bismark report: SRR11711272_1_val_1_bismark_bt2_PE_report.txt\n",
      ">> Reading from Bismark report: SRR11711273_1_val_1_bismark_bt2_PE_report.txt\n",
      "\n",
      "Wrote Bismark project summary to >> bismark_summary_report.html <<\n",
      "\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd bismark/ \n",
    "bismark2summary\n",
    "cd ../"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "17cb1228-12a5-4c75-ab4f-b205c5f2d78f",
   "metadata": {},
   "source": [
    "___\n",
    "for methylkit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "dfa103f5-ae40-45f1-a156-cb6460005370",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools sort -@ 24 -o bismark/SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.srt.bam bismark/SRR11711272_1_val_1_bismark_bt2_pe.deduplicated.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 24 in-memory blocks...\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "samtools sort -@ 24 -o bismark/SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.srt.bam bismark/SRR11711273_1_val_1_bismark_bt2_pe.deduplicated.bam\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 0 files and 24 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "for bam in bismark/*deduplicated.bam; do \n",
    "    out=${bam/.bam/.srt.bam}\n",
    "    cm=\"samtools sort -@ 24 -o $out $bam\"\n",
    "    echo $cm\n",
    "    $cm; \n",
    "    wait\n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0a35f3de-10b1-46c8-b032-12b1352d4f70",
   "metadata": {},
   "source": [
    "# "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "bd47dbec-e492-4af7-9143-2cec90d2d61a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Oct 19 13:34:29 PDT 2022\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:DNAme]",
   "language": "python",
   "name": "conda-env-DNAme-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
