{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a033638b-a236-41ed-9252-14a39bac7711",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "# Run `JunctionSeq`"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "750b49d3-ebfb-4be8-97f7-bb303204d70a",
   "metadata": {},
   "source": [
    "> A Utility for Detection of Differential Exon and Splice-Junction Usage in RNA-Seq data\n",
    "\n",
    "http://bioconductor.org/packages/3.10/bioc/vignettes/JunctionSeq/inst/doc/JunctionSeq.pdf\n",
    "\n",
    "http://hartleys.github.io/JunctionSeq/doc/example-walkthrough.pdf\n",
    "\n",
    "http://www.bioconductor.org/packages/devel/workflows/vignettes/rnaseqDTU/inst/doc/rnaseqDTU.html\n",
    "\n",
    "\n",
    "<!-- 1. Alignment \n",
    "- https://sailfish.readthedocs.io/en/master/sailfish.html\n",
    "\n",
    "2. Splicing \n",
    "- https://www.biostars.org/p/200963/\n",
    "\n",
    "Papers: \n",
    "- https://academic.oup.com/nar/article/44/15/e127/2457696\n",
    "- https://academic.oup.com/bib/article/21/6/2052/5648232\n",
    "- https://doi.org/10.1371/journal.pcbi.1007925 \n",
    "> Improved methods for RNAseq-based alternative splicing analysis\n",
    "https://www.nature.com/articles/s41598-021-89938-2\n",
    "-->\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "206a08ba-3ec9-46df-aa85-b95d47a915a0",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Alignment"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "72c74cbb-33e9-4388-bc5c-8e05c4a2b594",
   "metadata": {},
   "source": [
    "> QoRTs is extremelly memory-hungry for coordinate-sorted files, despite its manual not making this point clear. If you have lots of memory, you my try increasing the JVM memory with -Xmx, for example -Xmx64g or even higher. Or sort your bam files by name (`samtools sort -n -O bam -o file.sort.bam file.bam`) and use the `--nameSorted` QoRTs parameter - for name-sorted bam files, memory usage is indeed low.\n",
    "https://www.biostars.org/p/278221/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "00d909c8-5edd-4e9e-ab4e-75dd932bbc64",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hl60_120h_t_1\n",
      "==========================================\n",
      "hl60_120h_t_2\n",
      "==========================================\n",
      "hl60_120h_u_1\n",
      "==========================================\n",
      "hl60_120h_u_2\n",
      "==========================================\n",
      "hl60_6h_t_1\n",
      "==========================================\n",
      "hl60_6h_t_2\n",
      "==========================================\n",
      "hl60_6h_u_1\n",
      "==========================================\n",
      "hl60_6h_u_2\n",
      "==========================================\n",
      "hl60_72h_t_1\n",
      "==========================================\n",
      "hl60_72h_t_2\n",
      "==========================================\n",
      "hl60_72h_u_1\n",
      "==========================================\n",
      "hl60_72h_u_2\n",
      "==========================================\n",
      "kg1_t_1\n",
      "==========================================\n",
      "kg1_t_2\n",
      "==========================================\n",
      "kg1_t_3\n",
      "==========================================\n",
      "kg1_u_1\n",
      "==========================================\n",
      "kg1_u_2\n",
      "==========================================\n",
      "kg1_u_3\n",
      "==========================================\n",
      "molm14_t_1\n",
      "==========================================\n",
      "molm14_t_2\n",
      "==========================================\n",
      "molm14_t_3\n",
      "==========================================\n",
      "molm14_u_1\n",
      "==========================================\n",
      "molm14_u_2\n",
      "==========================================\n",
      "molm14_u_3\n",
      "==========================================\n",
      "ociaml2_t_1\n",
      "==========================================\n",
      "ociaml2_t_2\n",
      "==========================================\n",
      "ociaml2_t_3\n",
      "==========================================\n",
      "ociaml2_u_1\n",
      "==========================================\n",
      "ociaml2_u_2\n",
      "==========================================\n",
      "ociaml2_u_3\n",
      "==========================================\n",
      "ociaml3_t_1\n",
      "==========================================\n",
      "ociaml3_t_2\n",
      "==========================================\n",
      "ociaml3_t_3\n",
      "==========================================\n",
      "ociaml3_u_1\n",
      "==========================================\n",
      "ociaml3_u_2\n",
      "==========================================\n",
      "ociaml3_u_3\n",
      "==========================================\n",
      "thp1_t_1\n",
      "==========================================\n",
      "thp1_t_2\n",
      "==========================================\n",
      "thp1_t_3\n",
      "==========================================\n",
      "thp1_u_1\n",
      "==========================================\n",
      "thp1_u_2\n",
      "==========================================\n",
      "thp1_u_3\n",
      "==========================================\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[bam_sort_core] merging from 24 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 36 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 36 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 24 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 24 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 24 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 0 files and 12 in-memory blocks...\n",
      "[bam_sort_core] merging from 12 files and 12 in-memory blocks...\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "for bam in bam/*bam; do\n",
    "    sampleID=`basename $bam`\n",
    "    sampleID=${sampleID/.bam/}\n",
    "    out=${bam/.bam/.sort.bam}\n",
    "    echo $sampleID\n",
    "    \n",
    "    cm=\"samtools sort -@ 12 -n -O bam -o \"$out\" \"$bam\n",
    "    # echo $cm\n",
    "    $cm\n",
    "    echo \"==========================================\"\n",
    "    \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6541d070-07ed-4d92-b70c-df4184dbc6af",
   "metadata": {
    "tags": []
   },
   "source": [
    "## Generate count files \n",
    "\n",
    "https://anaconda.org/bioconda/qorts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "d1e3ecb5",
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p iso\n",
    "!mkdir -p iso/rawCts\n",
    "!mkdir -p iso/annoFiles"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "805d26db-7047-49be-81d8-416f189204a9",
   "metadata": {},
   "source": [
    "> #### Alignment\n",
    "> QoRTs [4], which is used to generate read counts, is designed to run on pairedend or single-end next-gen RNA-Seq data. The data must first be aligned (or\n",
    "\"mapped\") to a reference genome. RNA-Star [5], GSNAP [6], and TopHat2 [7]\n",
    "are all popular and effective aligners for use with RNA-Seq data. The use of\n",
    "short-read or unspliced aligners such as BowTie, ELAND, BWA, or Novoalign\n",
    "is NOT recommended."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "db5a8c21-6dff-4ca8-b547-d42b521ee676",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[38;5;9m/data_gilbert/home/aarab/Workflows/QoRTs-STABLE.jar\u001b[0m\n"
     ]
    }
   ],
   "source": [
    "ls ~/Workflows/QoRTs-STABLE.jar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "774cfea4-793b-42c0-b933-79bc56810645",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "%%bash\n",
    "\n",
    "GTF='~/genomes/hg38/gencode.v34/gencode.v34.annotation.gtf'\n",
    "\n",
    "touch QoRTs.sh\n",
    "for bam in bam/*sort.bam; do\n",
    "    sampleID=`basename $bam`\n",
    "    sampleID=${sampleID/.sort.bam/}\n",
    "    \n",
    "    # echo $sampleID\n",
    "    cm=\"java -Xmx60G -jar ~/Workflows/QoRTs-STABLE.jar QC --stranded_fr_secondstrand --singleEnded --nameSorted \"$bam\" \"$GTF\" iso/rawCts/\"$sampleID\"/ &> iso/rawCts/\"${sampleID}\".log\"\n",
    "    # --stranded --singleEnded \n",
    "    echo $cm >> QoRTs.sh;\n",
    "    # $cm;\n",
    "    # wait\n",
    "    \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f567d43-3da0-44bf-b29a-1e1d848bdd97",
   "metadata": {},
   "source": [
    "> The above command will generate a count file `rawCts/sampleID/QC.spliceJunctionAndExonCounts.forJunctionSeq.txt.gz`.\n",
    "This file contains both gene-level coverage counts, as well as coverage counts\n",
    "for transcript subunits such as exons and splice junction loci. It will also generate two additional secondary count files (one with all known splice counts,\n",
    "the other with all novel splice counts) which may be needed if you wish to add\n",
    "novel splice junctions to your analysis (see Section 4.4)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "3b15e6f5-2631-405b-beea-8bf96ebee705",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[38;5;27mhl60_120h_t_1\u001b[0m/     hl60_72h_u_1.log  \u001b[38;5;27mmolm14_u_1\u001b[0m/      ociaml3_t_2.log\n",
      "hl60_120h_t_1.log  \u001b[38;5;27mhl60_72h_u_2\u001b[0m/     molm14_u_1.log   \u001b[38;5;27mociaml3_t_3\u001b[0m/\n",
      "\u001b[38;5;27mhl60_120h_t_2\u001b[0m/     hl60_72h_u_2.log  \u001b[38;5;27mmolm14_u_2\u001b[0m/      ociaml3_t_3.log\n",
      "hl60_120h_t_2.log  \u001b[38;5;27mkg1_t_1\u001b[0m/          molm14_u_2.log   \u001b[38;5;27mociaml3_u_1\u001b[0m/\n",
      "\u001b[38;5;27mhl60_120h_u_1\u001b[0m/     kg1_t_1.log       \u001b[38;5;27mmolm14_u_3\u001b[0m/      ociaml3_u_1.log\n",
      "hl60_120h_u_1.log  \u001b[38;5;27mkg1_t_2\u001b[0m/          molm14_u_3.log   \u001b[38;5;27mociaml3_u_2\u001b[0m/\n",
      "\u001b[38;5;27mhl60_120h_u_2\u001b[0m/     kg1_t_2.log       \u001b[38;5;27mociaml2_t_1\u001b[0m/     ociaml3_u_2.log\n",
      "hl60_120h_u_2.log  \u001b[38;5;27mkg1_t_3\u001b[0m/          ociaml2_t_1.log  \u001b[38;5;27mociaml3_u_3\u001b[0m/\n",
      "\u001b[38;5;27mhl60_6h_t_1\u001b[0m/       kg1_t_3.log       \u001b[38;5;27mociaml2_t_2\u001b[0m/     ociaml3_u_3.log\n",
      "hl60_6h_t_1.log    \u001b[38;5;27mkg1_u_1\u001b[0m/          ociaml2_t_2.log  \u001b[38;5;27mthp1_t_1\u001b[0m/\n",
      "\u001b[38;5;27mhl60_6h_t_2\u001b[0m/       kg1_u_1.log       \u001b[38;5;27mociaml2_t_3\u001b[0m/     thp1_t_1.log\n",
      "hl60_6h_t_2.log    \u001b[38;5;27mkg1_u_2\u001b[0m/          ociaml2_t_3.log  \u001b[38;5;27mthp1_t_2\u001b[0m/\n",
      "\u001b[38;5;27mhl60_6h_u_1\u001b[0m/       kg1_u_2.log       \u001b[38;5;27mociaml2_u_1\u001b[0m/     thp1_t_2.log\n",
      "hl60_6h_u_1.log    \u001b[38;5;27mkg1_u_3\u001b[0m/          ociaml2_u_1.log  \u001b[38;5;27mthp1_t_3\u001b[0m/\n",
      "\u001b[38;5;27mhl60_6h_u_2\u001b[0m/       kg1_u_3.log       \u001b[38;5;27mociaml2_u_2\u001b[0m/     thp1_t_3.log\n",
      "hl60_6h_u_2.log    \u001b[38;5;27mmolm14_t_1\u001b[0m/       ociaml2_u_2.log  \u001b[38;5;27mthp1_u_1\u001b[0m/\n",
      "\u001b[38;5;27mhl60_72h_t_1\u001b[0m/      molm14_t_1.log    \u001b[38;5;27mociaml2_u_3\u001b[0m/     thp1_u_1.log\n",
      "hl60_72h_t_1.log   \u001b[38;5;27mmolm14_t_2\u001b[0m/       ociaml2_u_3.log  \u001b[38;5;27mthp1_u_2\u001b[0m/\n",
      "\u001b[38;5;27mhl60_72h_t_2\u001b[0m/      molm14_t_2.log    \u001b[38;5;27mociaml3_t_1\u001b[0m/     thp1_u_2.log\n",
      "hl60_72h_t_2.log   \u001b[38;5;27mmolm14_t_3\u001b[0m/       ociaml3_t_1.log  \u001b[38;5;27mthp1_u_3\u001b[0m/\n",
      "\u001b[38;5;27mhl60_72h_u_1\u001b[0m/      molm14_t_3.log    \u001b[38;5;27mociaml3_t_2\u001b[0m/     thp1_u_3.log\n"
     ]
    }
   ],
   "source": [
    "ls iso/rawCts/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7cfd25f1-d6e2-4a98-8041-b9ecde33d6fd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5ae13ba-ec4a-45e8-ab2a-f36cb0bc9f1b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5fb108d0-b9d9-45c7-ae89-72e925e521d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "03f31c4c-e2c8-46be-b6d2-35679252c5c6",
   "metadata": {},
   "source": [
    "### 4.3 (Option 1) Including Only Annotated Splice Junction Loci"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f318f590-08b7-4e77-b4be-6b5b3bec8c83",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/data_gilbert/home/aarab/genomes/hg38/gencode.v34/gencode.v34.annotation.gtf\n"
     ]
    }
   ],
   "source": [
    "ls ~/genomes/hg38/gencode.v34/gencode.v34.annotation.gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "9ce507c6-6b84-4cae-a4de-19364daea3ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "# !rm -rv annoFiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "084c9f92-f84f-446d-a32a-57199fce7fc2",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Starting QoRTs v1.3.6 (Compiled Tue Sep 25 11:21:46 EDT 2018)\n",
      "Starting time: (Tue Jan 25 20:07:09 PST 2022)\n",
      "INPUT_COMMAND(makeFlatGff)\n",
      "  INPUT_ARG(infile)=/data_gilbert/home/aarab/genomes/hg38/gencode.v34/gencode.v34.annotation.gtf\n",
      "  INPUT_ARG(outfile)=iso/annoFiles/JunctionSeq.flat.gff.gz\n",
      "  INPUT_ARG(stranded)=true\n",
      "FlatteningGtf: done.(2022-01-25 20:08:42)\n",
      "Done. (Tue Jan 25 20:08:42 PST 2022)\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "GTF='/data_gilbert/home/aarab/genomes/hg38/gencode.v34/gencode.v34.annotation.gtf'\n",
    "GFF='iso/annoFiles/JunctionSeq.flat.gff.gz'\n",
    "\n",
    "java -jar ~/Workflows/QoRTs-STABLE.jar makeFlatGff --stranded $GTF $GFF\n",
    "# --stranded_fr_secondstrand --singleEnded"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9f3dd720-85f1-4e27-bb35-7807a4825fce",
   "metadata": {},
   "source": [
    "### 4.4 (Option 2) Including Novel Splice Junction Loci"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ce19ffff-f578-4be6-ada7-140a75ed3114",
   "metadata": {},
   "source": [
    "### multiQC"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "c85bbe1e-4788-4d7f-a218-cc0830374754",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[INFO   ]         multiqc : This is MultiQC v1.9\n",
      "[INFO   ]         multiqc : Template    : default\n",
      "[INFO   ]         multiqc : Searching   : /data_gilbert/home/aarab/Projects/Decitabine-treatment/RNA-seq/iso/rawCts\n",
      "[INFO   ]           qorts : Found 42 logs\n",
      "[INFO   ]         multiqc : Compressing plot data\n",
      "[INFO   ]         multiqc : Report      : QoRTs-multiQC.html\n",
      "[INFO   ]         multiqc : Data        : QoRTs-multiQC_data\n",
      "[INFO   ]         multiqc : MultiQC complete\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd iso \n",
    "multiqc -n QoRTs-multiQC rawCts/\n",
    "cd ../"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "22cd07a0-fc4e-45de-a93a-cff1bb84fd8f",
   "metadata": {},
   "source": [
    "It looks good to me!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "544b79be-4ae1-4054-9393-8c2a7628ab7d",
   "metadata": {},
   "source": [
    "## Process count files "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cd15874f-f1e1-4275-92b9-da8985f4edd0",
   "metadata": {},
   "source": [
    "First I made `iso2` conda env: \n",
    "    \n",
    "    conda create -n iso2 -c bioconda bioconductor-genomicranges=1.38\n",
    "    conda activate iso2\n",
    "    \n",
    "Then, I installed `JunctionSeq` locally using ...\n",
    "\n",
    "    source(\"http://hartleys.github.io/JunctionSeq/install/JS.install.R\");\n",
    "    JS.install();\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2f1da8b6-659b-4968-a6a1-34e403475537",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "args <- commandArgs(trailingOnly = TRUE)\n",
      "\n",
      "PDIR <- args[1]\n",
      "samplesheetPATH = args[2]\n",
      "COND <- args[3]\n",
      "refCOND <- args[4]\n",
      "name <- args[5]\n",
      "\n",
      "setwd(PDIR)\n",
      "suppressMessages(suppressWarnings(library(tidyverse)))\n",
      "suppressMessages(suppressWarnings(library(JunctionSeq)))\n",
      "message ('R libraries loaded!')\n",
      "suppressMessages(suppressWarnings(dir.create('iso')))\n",
      "suppressMessages(suppressWarnings(dir.create(paste0(\"iso/\",name))))\n",
      "suppressMessages(suppressWarnings(dir.create(paste0(\"iso/\",name,\"/jscs\"))))\n",
      "suppressMessages(suppressWarnings(dir.create(paste0(\"iso/\",name,\"/plots/\"))))\n",
      "\n",
      "loadRData <- function(fileName){\n",
      "    ## loads an RData file, and returns it \n",
      "    # https://stackoverflow.com/questions/5577221/how-can-i-load-an-object-into-a-variable-name-that-i-specify-from-an-r-data-file\n",
      "    load(fileName)\n",
      "    get(ls()[ls() != \"fileName\"])\n",
      "}\n",
      "\n",
      "if (file.exists(paste0(\"iso/\",name,\"/jscs/jscs.RData\"))){\n",
      "    jscs <- loadRData(paste0(\"iso/\",name,\"/jscs/jscs.RData\"))\n",
      "\n",
      "    message ('`jscs.RData` exists!')\n",
      "    \n",
      "} else {\n",
      "    ###################################################################################################\n",
      "    message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "    message ('Part 1. Read genome annotations, count data and samplesheet')\n",
      "    message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "    message (date())\n",
      "\n",
      "    GTF = '~/genomes/hg38/gencode.v38/gencode.v38.annotation.gtf'\n",
      "    # annoFiles ='~/genomes/hg38/gencode.v38/JunctionSeq.flat.gff.gz'\n",
      "    annoFiles ='/data_gilbert/home/aarab/Projects/Decitabine-treatment/RNA-seq/iso/annoFiles/JunctionSeq.flat.gff.gz'\n",
      "\n",
      "    gtf <- rtracklayer::import(GTF)\n",
      "\n",
      "    gene2name <- gtf[gtf$type == \"gene\"] %>% data.frame %>% \n",
      "        column_to_rownames('gene_id') %>% dplyr::select('gene_name') %>% \n",
      "        rownames_to_column('geneID')\n",
      "\n",
      "    txpt2name <- gtf[gtf$type == \"transcript\"] %>% data.frame %>% \n",
      "        column_to_rownames('transcript_id') %>% \n",
      "        dplyr::select(c('transcript_name','gene_name'))\n",
      "\n",
      "    message ('annotations loaded!')\n",
      "\n",
      "    samplesheet = read.table(\n",
      "        samplesheetPATH,header=TRUE,stringsAsFactors=FALSE,sep=','\n",
      "    )\n",
      "\n",
      "    countFiles <- paste0(\n",
      "        \"iso/rawCts/\",\n",
      "        samplesheet$Sample,\n",
      "        \"/QC.spliceJunctionAndExonCounts.forJunctionSeq.txt.gz\"\n",
      "    )\n",
      "    files = countFiles %>% as.list\n",
      "\n",
      "    message (\"samples:\")\n",
      "    message (samplesheet[,1])\n",
      "    message (\"conditions:\")\n",
      "    message (samplesheet[,1])\n",
      "\n",
      "    ###################################################################################################\n",
      "    message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "    message ('Part 2. Run JunctionSeq')\n",
      "    message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "    message (date())\n",
      "\n",
      "    jscs <- runJunctionSeqAnalyses(    \n",
      "        sample.files = files,\n",
      "        sample.names = samplesheet[,1],\n",
      "        condition=relevel(factor(samplesheet[,COND]),ref=refCOND),\n",
      "        # use.novel.junctions = FALSE,\n",
      "        flat.gff.file = annoFiles,\n",
      "        gene.names = gene2name,\n",
      "        nCores = 15,\n",
      "        analysis.type = \"junctionsAndExons\"\n",
      "    )\n",
      "    message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "    message ('Part 3. Save results')\n",
      "    message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "    message (date())\n",
      "\n",
      "    writeCompleteResults(\n",
      "        jscs,\n",
      "        outfile.prefix=paste0(\"iso/\",name,\"/jscs/\"),\n",
      "        save.jscs = TRUE\n",
      "    )\n",
      "}\n",
      "\n",
      "message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "message ('Last Part. Draw plots')\n",
      "message ('@@@@@@@@@@@@@@@@@@@@@@@@@@@@@')\n",
      "message (date())\n",
      "buildAllPlots(\n",
      "    jscs=jscs,\n",
      "    outfile.prefix = paste0(\"iso/\",name,\"/plots/\"),\n",
      "    use.plotting.device = \"svg\",\n",
      "    FDR.threshold = 0.01\n",
      ")\n",
      "\n",
      "message (date())\n"
     ]
    }
   ],
   "source": [
    "!cat run-JunctionSeq.R"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "42c14765-0154-40f1-8830-d0943073794d",
   "metadata": {},
   "source": [
    "# Downstream "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "76d9c8e8-6a1d-426d-a199-eb89264ffc71",
   "metadata": {},
   "source": [
    "## `genewiseResults`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "e5f60b05-cf85-4f39-b194-b9967f92cf31",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd \n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "id": "a42ab1b2-badb-4925-b4c3-c14293a1061a",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys \n",
    "sys.path.append('../')\n",
    "from util import cleanAxes\n",
    "from mpl_toolkits.axes_grid1 import make_axes_locatable\n",
    "from adjustText import adjust_text\n",
    "import seaborn as sns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 535,
   "id": "fce7b1f5-0cfa-4c3e-be4f-eefb8268cb62",
   "metadata": {},
   "outputs": [],
   "source": [
    "def rank_plot_pvalue(df,ax, title=None,genes=None):\n",
    "    \n",
    "    df = df.sort_values(['geneWisePadj'])\n",
    "    epsilon = df['geneWisePadj'][df['geneWisePadj'] != 0].min()\n",
    "    \n",
    "    df['geneWisePadj_rank'] = df.sort_values(['geneWisePadj'],ascending=False).reset_index().index + 1 \n",
    "    df['geneWisePadj_log'] = -np.log10(df['geneWisePadj'] + epsilon)\n",
    "    df = df.set_index('geneName')\n",
    "\n",
    "    X = df['geneWisePadj_rank']\n",
    "    Y = df['geneWisePadj_log']\n",
    "\n",
    "    ax.scatter(X,\n",
    "               Y,\n",
    "               c = \"teal\",\n",
    "               s = 10,\n",
    "               # edgecolor = \"black\",\n",
    "               # linewidth = 2,\n",
    "               alpha = 0.7\n",
    "              )\n",
    "\n",
    "    # ax.legend()\n",
    "    if genes:\n",
    "        if type(genes)==int:\n",
    "            n = genes \n",
    "            genes = list(X.sort_values().iloc[0:n].index)\n",
    "        #For Labeling\n",
    "        if type(genes) != list:\n",
    "            x_lab = X.loc[[genes]]\n",
    "            y_lab = Y.loc[[genes]]\n",
    "        else:\n",
    "            x_lab = X.loc[genes]\n",
    "            y_lab = Y.loc[genes]\n",
    "        # Outline Genes To label\n",
    "        ax.scatter(x_lab,\n",
    "                   y_lab,\n",
    "                   c = \"teal\",\n",
    "                   s = 50,\n",
    "                   edgecolor = \"lightblue\",\n",
    "                   linewidth = 1,\n",
    "                   alpha = 0.7\n",
    "                  )\n",
    "\n",
    "        texts = []\n",
    "        for x, y, s in zip(x_lab, y_lab, x_lab.index.tolist()):\n",
    "            texts.append(ax.text(x, y, s))\n",
    "\n",
    "        adjust_text(\n",
    "            texts, force_points=0.2, force_text=0.2,\n",
    "            expand_points=(2, 2), \n",
    "            expand_text=(2, 2),\n",
    "            arrowprops=dict(arrowstyle=\"-\", color='black', lw=0.2),\n",
    "            ax=ax,\n",
    "        )\n",
    "    \n",
    "    ax.set_xlabel(None)\n",
    "    ax.set_yscale('symlog')\n",
    "    cleanAxes(ax)\n",
    "    ax.set_title(title,size=14)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 536,
   "id": "0c23a5e4-8426-4916-85c1-872a2a7e5e3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_6h   = pd.read_csv('iso/hl60_6h_delta_iso/jscs/sigGenes.genewiseResults.txt.gz',sep='\\t')\n",
    "df_72h  = pd.read_csv('iso/hl60_72h_delta_iso/jscs/sigGenes.genewiseResults.txt.gz',sep='\\t')\n",
    "df_120h = pd.read_csv('iso/hl60_120h_delta_iso/jscs/sigGenes.genewiseResults.txt.gz',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 537,
   "id": "1b777f6a-da0c-45bd-8c0e-9c2b6bda6e9a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAagAAADQCAYAAABStPXYAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAriElEQVR4nO3de3xU5Z348c+TKyEhQQyQwHC/aQgSxkBAsVKiq6KBKLXi4la20+3Pem231bWgVbuNrLW21WqvZhesCrZaURSrFcUupFBYIBCgYBICBAIkQAjJhMlcnt8fM4nJzCSEZJIzc/J9v17zIueZM+F7kifznfNcldYaIYQQItxEGR2AEEIIEYwkKCGEEGFJEpQQQoiwJAlKCCFEWJIEJYQQIixJghJCCBGWJEEZQClVoZT6ntFxiPCllNJKqa8YHYcIX0qp0b56km10LD1FElQPUEqlK6VWKqWqlVLnlVJ7lVLXGh2X6B2+DyA6yON9pVSsUuoZpdQupVSDUqpKKfW6Umqk0XGL0FFKfUkp9a5S6qjvd7+k1XOdqgNKqXil1C+UUjW+895VSll6/WIMJAkqxJRSA4FNgAJuBi4HHgBOGhiW6F3TgfRWDyuggT8A/X3HBb5/FwAjgD8rpWIMiVb0hCSgBHgIaPR7rrN14OfAQuBO4BogGXhPKRXdo5GHE621PEL4AJ4GNl3gnArgMeA3QB1QCTxsdOzy6LE6sQyoBfq383wG3gQ2pVWZBr4J/BFoAMqBu4y+Fnl06fdfDyy5wDlt6gCQAjQBi1udMwLwADf4jkf7XrMQ+AtgB/YC1xt9zaF6yB1U6OUDW5RSbyilTiqldiql7ldKKb/zvgPsxvsJ6hngx0qpWb0cq+hhvt+7DXhVa21v57Rk379n/Mp/ALwDTAXeAP5bKTWqRwIVRvOvA1cCscBHzSdorY8A+4Cr/F5bALyAt55sBVYrpZJ6NNpeIgkq9MYC9+L9xHsD8DzwX8B9fud9pLV+UWtdqrX+BVAK5PZqpKI3XA+MAV4O9qRSKg54Dlirta70e/r3WutXtdalwOOAC29TjzCRdupAGuAGavxOP+F7rrWfaa3Xaq0/B5YCg4Csnou490ibd+hFAdu01t/3He9QSk3Am6BebHXeLr/XHQOG9EJ8onf9G7BVa73T/wlff8OrwEBgfpDXttQRrbVLKVWN1BFT6UQdCHgJ3ma91lq/lxzz/WuKeiJ3UKFXhbcduLV9gP8oLaffsUZ+H6ailBqCtwP8d0GeiwFWAVcAuVrrU0G+hdQRE7tAHTgORAOpfi8bgvcuqrWWeqJ9nVOYpJ6Y4iLCzCZgkl/ZROCQAbEIYy0BHMDq1oVKqVi8fUpXAF/WWh/v/dCEkTpRB/4Pb+K5vtVrLHhHBRf1VpxGkya+0PsZUKSUWoa3Ak4DHsTbNiz6CN/giG8Aq7XW51qVx+AdmTcdyAO0Uqq5T+Gs1tp/SLKIQL5BCuN9h1HASKVUFnAabzNch3VAa31WKVUIPKuUOgmcAn6Ktznv4967EmPJHVSIaa234h3J91W88yAK8HZw/9LAsETvmwNMILB5z4K32W8Y3k/JVa0ed/RifKJnZQM7fI8E4Cnf1z+k83XgO8Cf8H7Q3YR3uHqe1trdO5dgPPVFk6UQQggRPuQOSgghRFiSBCWEECIsSYISQggRliRBCSGECEthkaBuvPFGjXcSojzM9+gSqROmfnSJ1AlTP4IKiwRVU+O/3JTo66ROCH9SJ/qesEhQQgghhD9JUEIIIcJS2CUot0dzosHBwdoGTjQ4cHu63GQthBAX7fjx4yxatIhx48aRkZHBvHnzOHDgAAA/+9nP6NevH2fPnm053263s3jxYqZMmUJmZiazZ8+mvr4egOjoaLKyssjMzOT222/Hbre3fJ/JkyeTmZnJnXfeyfnz53v/QiNAWCWoOoeT90sr+Z+du/jp3/7G/+zcxfulldQ5/Bd1FkKI0NNac+uttzJnzhzKysrYu3cvTz/9NCdOeBcQX7VqFdOnT+ftt99uec3zzz/P0KFD2b17NyUlJRQWFhIbGwtAQkICO3fupKSkhLi4OH79619z9OhRXnjhBbZt20ZJSQlut5vVq1cHjaevC5vFYt0ezYZDx3lr7052Hj/K/upqHB4PowdeyvzLruDVnZu5J/tKCnJlTz8hRM/49NNPiY2N5Z577mkpy8rKAqCsrIz6+nqeffZZnn76aZYsWQJAVVUVo0aNajl/0iT/zQy8rrnmGnbt8m7d5HK5aGxsJDY2FrvdzrBhw3rmgiJc2NxB1TQ2sbf6ODuPH2XXiRM4PB4AKg6Vc/j0SSwpg3h640aWrFljbKBCCNMqKSnhyiuvDPrcqlWruPPOO7nmmmvYv38/J0+eBODrX/86zzzzDLNmzeKxxx7j888/D3ity+Xigw8+YMqUKQwfPpzvfe97jBw5kvT0dFJSUvinf/qnHr2uSBU2CcrudHG07iw1DQ1tn4iL4+DnB4g5cQKOHeO1Dz9kxfvvs23btpZb5Lq6OmOCFkL0GatXr2bRokVERUVx22238cc//hHw3mGVl5fz8MMPc/r0aaZPn86+ffsAaGxsJCsri+zsbEaOHInNZuPMmTO88847HDx4kGPHjtHQ0MCrr75q5KWFrbBp4usfG8Pw5BRSExM55utgBCAhgXGXT6b0VDXgIiY6msYhQ8jOzga8FeDIkSMtnZgjR45kyBBT7HYshOhlkydP5s033wwo37VrF59//jnXX+/dP7CpqYmxY8dy3333AZCUlMRtt93GbbfdRlRUFOvWrePyyy9v6YNq7Z133mHMmDEMHjwYgNtuu42ioiLuuuuunr24CBQ2d1CpCXFkDE4jK204VwwdSnyUN7TRAy/FknwJ+6q9G04O7NcPa3p6y+sSEhKYOHEi2dnZZGdnBySn9kbk7Nmzh7lz5zJx4kQmTJjAf/7nf9K89ciKFSsYPHgwWVlZZGRk8LvffbGlz9tvv41Sin/84x89/SMRQvSyuXPn4nA42vzNb926lYceeognn3ySiooKKioqOHbsGEePHuXQoUNs2rSJM2fOAN7EtXfv3jZ9Uv5GjhzJ5s2bsdvtaK1Zv349l19+eY9fWyQKmwQVHaWYMyqNhRlZ3DnlSr45fRbLr5vH8zctILWfIkpBrFJwEftXdTQiZ/78+Tz66KMcOHCA4uJiioqK+OUvv9hT8I477mDnzp1s2LCBpUuXthnFM3v2bBl1I4QJKaV4++23+ctf/sK4ceOYPHkyTz75JBs2bODWW29tc+6tt97K6tWrKSsr49prr2XKlClMmzaN7OxsFi5c2O7/kZOTw1e+8hWsVitTpkzB4/HwzW9+s6cvLSKFxYaF2dnZetu2bYB3NF9NYxONThcJsTGkJsRx95q3eW337pbzbxg3jj934nb4k08+4cknn+Svf/1rm/LCwkI+++wzXnnllZaysrIy5syZw5EjR1ixYgXbtm3jxRdfBGDmzJm88MILZGRkMGnSJD799FPmz58vd1Gdo7ryotZ1QpiO1AnhL2idCJs7qGbRUYqhifGMHpjI0MR4oqMUB/zW4PrfQ4fYUll5we/V3oicPXv2BJSPGzeO+vr6gAEX5eXllJeXM378eNasWcONN97IxIkTGTRoENu3b+/CFQohwoXH4+HgwYNGhyHaETaDJDoyMTWVrVVVLccxUVFsr6oix2Lp0vfTWqNU8A9xzeVvvPEGGzduJD4+nt/85jcMGjSIVatW8e1vfxuARYsWsWrVKqxWa5diEEJ0XXNLi93por+vpSU66sI3ZlprDhw4QF1dHVproqKiuOyyy3ohYtEVEZGgHpgxgw8+/5zTvuVAmtxu4qKjL/i69kbkTJ48OaDZr7y8nKSkJAYMGAB4+6Cam/gATp06xSeffEJJSQlKKdxuN0opfvzjH7eb7IQQoVfncLLh0HH2Vh/naN1ZhienkDE4jTmj0kiOj21zrtaagwcPcurUqZayCRMmtDuZVoSXsGviCybHYsE2bRqD+vUjNSGBIYmJNLndF3xdeyNyJkyYwMaNG/n4448B71D1Bx98kEceeaTd7/Xmm2/yta99jUOHDlFRUcGRI0cYM2YMGzdu7P4FCiE6pfWKM6t2/x+F27fw+PoPeOyTD3jnwCHcHs3hw4fZunUrW7duZdu2bSQnJzN9+vSWx8CBA42+DNFJEZGgACalpnLe5cLudFJjt3fqDqq9ETnDhg3jnXfe4Uc/+hGTJk1iypQpTJ8+nfvvv7/d77Vq1aqAUTwLFy7k9ddf7/a1CSE6x3/FmUaXC5fW7N61k1fWvcvL771PXFxcm4SUmppqdNiiiyKiiQ+8zXojUlKIjY7G6XZ36g4KYNiwYfzhD38I+tyGDRuCli9ZsqRlna2Ozn3wwQc7FYMQIjTaXXFm6FDsgwcTO2I0aWlpxgQnQi7kd1BKqbFKqUKlVGDnTzdY09NJiosjPjqapLi4NpN1hRB9Q+sVZ/ylD0gmWrqDTaVTCUop9d9KqZNKqRK/8huVUvuVUqVKqUcBtNblWmtbqAPNsVj4VnY2EwYN4lvZ2V0ewSeEiFz+K87E+lacGTcolYmXDsHubDQ4QhFKnW3iWwG8CLTMbFVKRQMvAdcDlcBWpdS7Wuu9oQ4SYEtlJT8pKuJ0YyOf+uYt2GSItxB9SvOKM5DF5YPTOHCqGpcbRg0cxNaj5SyY8yWjQxQh1KkEpbX+q1JqtF/xDKBUa10OoJRaDSwAOpWglFLfBL4J3rWpLuStffs4WFuL0+NBa82yTz4hc8gQuZMykYutE8L8gtWJ5PhYbh5vIWf4EBqdLj6rOMhHZfu5PeMyeT8wme70QQ0HjrQ6rgSGK6UuVUr9GpimlPp+ey/WWv9Wa52ttc5uXtW3Q1rj8XjwaI0GGp1OtreavCsi30XXCWF67dWJ5hVnTtSf4b82buB/Dx/mJ0VFnVphRkSO7iSoYN2RWmt9Smt9j9Z6nNZ6eTe+fxsLMzIYEBfXcnyuqYlK2QdKiD7trX37OHz2LGcdDg6fPctbvn2YhDl0J0FVAiNaHVuAY90Lp305FgvWYcOIVoooQAMfl5X11H8nhIgEvsWuld+xMIfuJKitwASl1BilVBywCHg3NGEFN2P4cNxa4/Edb6uqolAWbBWiz1qYkcHIlBSS4+MZmZLCwowMo0MSIdTZYeargL8Bk5RSlUopm9baBdwPfAjsA/6gtd7Tc6GCJTm5ZVgpgAf4rSy/L7ooOjqarKwsMjMzycvLo7a2FoCKigoSEhJaNqy85557vP2fHg8PPvggmZmZLauPNK+EvWzZMkaMGEFSUlLQ/+vNN99EKYVsFxFaORYLK/Lz+f7s2azIz5dBEibT2VF8d7ZTvg5YF9KIOmBNTydWKZytY5CFWkUXtd6O++677+all15i2bJlgHf7lZ07d+JyuZg7dy5r1qzB4XBw7Ngxdu3aRVRUFJWVlST6Jozm5eVx//33M2HChID/59y5c7zwwgvk5OT02rX1JTkWiyQmk4qYtfjAWxG/PWtWS9AKuH7sWCNDEiYxa9Ysjh49GlAeExPDVVddRWlpKVVVVaSnpxPlu4u3WCxccsklgHdTy/R2Vjd5/PHHeeSRR+jXr1/PXYAQJhRRCQpg/qRJDElMJFopYqOi+NO+fTK0VHSL2+1m/fr1zJ8/P+A5u93O+vXrmTJlCl/96ldZu3YtWVlZfPe732XHjh0X/N47duzgyJEj3HLLLT0RusA7if/GV1/lsl/8gmXr1xsdjgihiEtQ26uqiI2OJjEujoTYWBpkPpToosbGRrKysrj00ks5ffo0119/fctzZWVlZGVlcfXVV3PzzTdz0003YbFY2L9/P8uXLycqKorc3FzWd/CG6PF4+M53vsNzzz3XG5fTJ22prCR/1So+LCtj/+nTLN+4UZKUiURcgrKmp5MYG4vD5cLe1ESUr0yIi9XcB3Xo0CGampp46aWXWp5r7oPasWMHTz75ZEt5fHw8N910E88++yxLly5lzZo17X7/c+fOUVJSwpw5cxg9ejSbN29m/vz5MlAihLZXVXHW4QC8Tf4a+LC01NCYROhEXILKsVj43lVXkRgbS7/Y2Au/QIgLSElJ4YUXXuAnP/kJTqez3fO2b9/OsWPeqX4ej4ddu3YxatSoDr9vTU0NFRUVVFRUMHPmTN59912ys7NDfg19lTU9nZT4eMCbnBRww/jxhsYkQifiEhTA/lOnONfUhN3p5Fh9vcweF902bdo0pk6dyurVq9s95+TJk+Tl5ZGZmckVV1xBTExMyyaXjzzyCBaLBbvdjsViaXPXJXpOjsXCmjvv5IZx45g0aBDfnz2bgtxco8MSIaJ0GMy8zs7O1hfT7HHj73/Ph+XlLceTU1Mpue++nghNdF+X5gFcbJ0QEUXqhPAXtE5E5B2U2y+p7qmpkRUlRLs8Hg979uxh69atQYeSCyHCU8Rs+d7aosxMPvbN4G/206Ii2R/KpNweTU1jE3ani/6xMaQmxBEd1f6HcK01paWlnDlzBqUUUVFRXHbZZS2TaoUQkSEiE5TNauXxTz6hqqGhpezA6dNsqayUGeUmU+dwsuHQcfZWH+do3VmGJ6eQMTiNOaPSSI7/YpDM4cOHOX78OABKKcaPHx90VQdhTlsqK9leVYU1PV3eA0wkIhMUwNyxY3lt9+6WY5fW/OLvf5fKaSJuj2bDoeO8tXcnZaerOetwkBIfz77q49SesjAxytlyJzVixAhmzJhhcMTCCFsqK1myZg0NTieJsbGyJp+JRGyCemDGDP6wZw9Oj6el7EBNjYERiVCraWxib/Vxyk5XU3LyJODdIbPhyBHSMjU33HgDQxPjjQ1SGK55Tyi31jjdbp7YsIE/33WX0WGJEIjIQRLg2x/Kb4JuTWOjQdGInmB3ujhad7ZlImZ8jPfzlGvIEBoT+9PodBkZnggXWuP0eHC43Xjw7hMng6bMIWITFMCXRo5scwEHa2tlmRMT6R8bw/DkFFLi43F5PNQ5HLg8HlLi4xk2IIWE2IhtABAhtDAjg9hWuxp4gDdKSowLSIRMRCeohRkZbfaHUsgyJ2aSmhBHxuA0BicOaFM+OHEAGYPTSE2IMygyEU5yLBauGT0a+OINbUg7+3KJyBLRH0FzLBYWTZnCK8XFLWWyzIl5REcp5oxK42hdBgPiEjnRcI5BCYlMHz6COaPSOhxqLvqWp+bMofj48ZaBNA/IgBlTiOg7KIAV+fncNH48A+PjuWn8eFnmxGSS42OZNjSF0431eDya8tPV/M/2TeyrPmF0aCLMDOzXj9T+/Rko+26ZRsQnqMLt29lQUUG908mfS0tZ0sHq0iIyzRxhISN1IPtPVREbrYmPiZEtVkQb26uqSIyLY2paGolxcVI/TCLiE9TH5eU4PR6cHg8eYGVxMTe/9prRYYkQW5iRwZDERM67vCP3ZIsV0Vpzfag6d67NsYhsEd0HBXDd2LGs3rOnTdk6353Uivx8Y4ISIZdjsfDSvHmyWoAISuqHOUX8HZTNamV0SkpA+Wu7dslW8EL0ITkWC9+aPl2Sk4lEfIICeOxLXwooc2kt+0SZyJbKSu5bt47CHTu4b906+fAhAmyprORXW7dK3TARUyQom9XK3VOntilTwLG6OmMCEiHX3OmdPmBAm2MhQD7AmJUpEhR4h5vfMG6c0WGIHmJNT6ehqYni48dpaGqSTnDRRvMHln4xMZxsaJDWE5MwTYICuLLVm5YG1ldUyCcpIfqA5g8wO6qqOGW3s3b/fvnbNwFTJajyM2eIVqpl7+CGpiZpCjIJmeciOpJjsZA3cSLRvjX5ahsbpY6YgKkS1HVjxxKlFM0bwjvcbuKiow2NSYSGzHMRF3LSbuec04nd5eK43c6Wo0eNDkl0U8TPg2rNZrXyaUUFb+/bR2x0NCnx8TS53UaHJUJA5rmICyk54V3+SuFt4m8+FpHLVHdQAF8ePRqnx8NZh4PDdXXyKcpEZJ6L6Ejm0KEALS0ozccicpkuQX1aUdFml93fFxfL5mVC9AE5w4eTlphISnw8aYmJ5AwfbnRIoptMl6BO1Ne3OZbNy4ToG6zp6QxPTmb8oEEMT06WfkoTMFUfFMCizEw+PniwTVlUlOnysBDCj/RTmo/pEpTNauWPe/fyYVkZ4O0wTU1IMDYoIUSvyLFYJDGZiClvLW7PyCDOd9ekgQ9KS2XSnhBCRBhTJqgmt5v4VvOfTp8/zy/+/ncDIxJC9BZZNNY8TJmgrOnpuFuN5AN4c88eqbARrnD7du58800ZlSnataWykiVr1rB840aWrFkjf/MRzpQJKsdi4ZrRo9uUOTwe5r7yilTYCFW4fTsPfvABb//jH3zrvfdYtn690SGJMPTWvn0crK3lZEMDB2trZdHYCGfKBAXw1Jw5AWV2p1MqbIT6uLwct9a4PR5cWvOrbdvkw4YIcKyuDofb3fKQLXcim2kTVI7FwlVBRvOsO3DAgGhEd103diwejwe31iilSI6Pl8VARVCq1dc1drthcYjuM22CAthkswUMMd9TU8PVhYUGRSS6yma18vDVVzOwXz9GJCeT2r+/TMQUAYYNGEBMVBQKb6LaeeKE3GlHMFMnKICstLSAsqLKSoY/95xU3AhTkJvLB4sX8x9XX81L8+bJfBcRYGFGBoP69Ws5drpccqcdwUyfoBZlZgYtP1Zfz7UrVkiSijCyYKzoSI7FwpXDhqHxzoE87XDwXFGR0WGJLjJ9grJZrdw9dWrQ5xxuN7Z33+3liIQQPan89Ok2x2W1tSxZs8aYYES3mD5BAazIz283Se2prpZ5NUKYyKhLLgko+6S83IBIRHf1iQQF3iT1cl4e8UEWjv1/a9dKkooQMllXXMhTc+a0WUkGYERKikHRiO7oMwkKvM19n/3rvwZctBv4hiSpsFe4fTv3rVvHH/bs4d/WrpVmGxFU86rmca0+jG6prJS/7wjUpxIUeCvv9ePGBX3uR5991svRiIvxRkkJDrcbD94O8FdkM0rRjia3u818KDfy9x2J+lyCguBNAAAVdXXyhhfGhiQltTnWyGaUIjhrejoerduUVdbXy6jdCNMnE1SOxcJnS5aQHBcX8NzjssZb2Hpgxgz6x3yxhZkiMGkJAd6/8X++4oo2ZdrjkaXOIkyfTFDgrcAf/cu/EKtUm/Iqu52bX3vNoKhER3IsFl646SaS4+OJj45maGIiD8yYYXRYIkytyM9n8uDBLcduvP2YchcVOfpsggLvG97DV18dUL6utFSWQwpTmUOGMCwpiSGJiQxstWKAEMFYBgxoc3z6/Hlufv11SVIRok8nKPAun5OWmBhQXlRZKUkqDG2vqiIxLo6paWkkxsXJMjaiQ26/fiiAM42NUm8iRJ9PUAA/mjs3aHlRZaU094UZa3o6DU1NbDp0iAM1NVTKdgqiA4syMwPe5Dwg9SZCSILCOz9q3vjxQZ9bV1oqSSrM1DY2csbh4JzTyY83bZKRl6JdNquV3+bl0d83alcBsVFRlJ85Y2xgolMkQfm8v3hxu8shrSstlTfBMLG9qoo6h6Pl2KU1v922zcCIRLizWa18e9YswDs1wenxsOfkSWODEp0iCaqVjtbs+6msiBwWgu0Bpf1GYgrhz+nxtJm4u7u6WlYiiQCSoPysyM8P2txXVlsrI3/CQI7Fwu2ttlBRwPVjxxoXkIgMWuM/XOKDzz83JBTReZKgggjW3OeSSX5hI2f4cEalpDAsKYmRKSlYkpONDkmEuYUZGQELRUcHWThahBf5DbUjYJKf1qw7cMDAiEQza3o6qf37kz5ggGz9Ljolx2LhpZtvJqZVc3BVfT3LZOWYsCYJqgOtl9UB2FNTIyP6wkDzatW2adNk63fRaTarlWF+E3df3LLFoGhEZ0iC6sDE1NSAsnWlpcT/8IfyyctgsvW76Ip0vwRV53TKhPwwJgmqAw/MmBH0B9SkNU9v3CijgISIMF8aOTKgrKiyUj5whilJUB3IsVj4l3aGnQOsLC4m7oc/lERlkC2Vlfxq61YZXSk6LdhgCYDlGzfKXMcwJAnqAtobdt7MqTUri4sZuHy5vFH2oi2VlSxZs4blvjtZ+dmLzmgeLOFPA//+4YdSj8KMJKhOeH/xYl7Oy+PS+Ph2zznb1MTMwkLGPf+8VPJe8Na+fRw+e5azDgeHz56VKQCi02xWa9AJ+Q63WxaRDTOSoDrJZrVS8+ijHd5NAZTX1jKrsFCaC3qa1jjcbuocDhpdLo7J4p/iIqzIz+cqvwE2Lo+HuCA7bQvjSIK6SM2TeP03OmxNA/e8957cSfWgk3Z7y1YKGvhzWZn8vMVF2WSzMcVvruODf/6z1KMwIgmqC1bk59P0gx+w2Wbj0oSEoOe4tObWN97o5cj6jpITJ9oc1zc1SfOMuGi158+3ObY7nTyxYYMxwYgAkqC6IcdioeaRR1g6ezbBGgaq6uu54pe/7PW4+oIb/Jpa3VpL84y4aHODrOP4YVmZjMwNE5KgQqAgNxfXE08wLCkp4Lnd1dUkFRRIn1SIFeTmcsO4ccRHRzO4f3/GXXIJTW630WGJCLMiPz9oK8jK4mKZGxUGJEGF0J/uuIPoIH1TDS4X31i7Vu6mQuypOXMYM3Ag/WJiUATfikOIC3nmuuuClv+3fKg0nCSoEMqxWPjNLbfQ3vCJ3dXVxP/wh3I3FUKNTif1DgeNTqfRoYgIZbNaA0b0ARy32+Vv1WCSoELMZrXyu7y8dn+wTVrzjbVrGbh8uVT+bnpr3z6q7XbcQLXdLnOhRJdtstmYHGTtzf/46CMDohHNJEH1AJvVSlEHI/zAO7H339aulSTVHVrj1poml8s75Fz7b0knROcVLlhAf7+BNqccDsY//7xBEQlJUD2k9Qi/9n7IGnhcOmK7bFJqKgrvdt5uj4dYGcUnuiHHYuHWjIyA8rLaWuk/NogkqB5WkJuL+4kn2kwIbK3Kbsfy3HO9HJU5NLndDElMJDY6muioKF7dtUsmWYpueWDGjIC7KPD2H8vQ894nCaqX7Lr3Xl7Oywu6AsXR+noGFBTIm+tFsqano/FW4n4xMSTExspkXdEtORYLnyxZEvTv9E979xoQUd8mCaoX2axWfnXLLUGfq3e5mFlYSGJBgcy/6KQci4Unrr2W1P79SUtKIikuToaai27LsViC/p2eczqlz7iXSYLqZTarlaWzZ7f7vN3l4umNG2VybyfZrFa+PGYM9qYmMocMkR12RUi093f6RkmJAdH0XZKgDFCQm8tmm43+sbHtntM8uVfavTu2bP16Xiku5si5c6wsLpbtu0XIFOTmMtmv77ikutqgaPomSVAGybFYaFi6lHEDB3Z4nrzpduzD0lJaDy6X7btFKKX7LV8m62v2LklQBit96CGWzp4ddBvqZkWVldz82mu9GFXk8F80FuDXW7caEIkwo0WZmQFlu6urJUn1EklQYaAgN5fzjz/Oy3l5QYe4AqwrLZXmviAKcnMDJkSfdjik/06EhM1qDbrCxO7qarlT7wWSoMKIzWql4bHH2t21d2VxscxqDyLYYp8/LSoyIBJhRoULFgRdBPqZjRtlakgPkwQVht5fvLjdkX5ltbUk/OhHcofQis1qZYxfX97eU6fkjlOERPMi0P7cwNxXXpEk1YMkQYWpgtzcdu+kzrvdsuCsn3GXXBJQtrK4WH4+IiTaG3ZudzplkeIeJAkqjL2/eHHQbQCanW1qkqHoPsE6s0HWOhShU5CbG3TJsmN1dQZE0zdIggpzm2w2Xs7L63CU38ri4j6/AoXNag16x1llt/fpn4sIrV333huwc3bRkSMGRWN+kqAigM1q5fzjj7e74Cx8sQJFX54z9f7ixUFHXMnOqCKUvjx6dJvjg2fPSitGD5EEFUF23XsvS2fPDrqQZbOiyso+naQKFywg3m+ovuyMKkJp2IABAWWykGzPkAQVYQpyc2n6wQ867Jsqqqwk6qmn+uTk3hyLhZfmzQsol74oESoLMzJI9lum7JzTKU3JPUASVITaZLOxdPbsdif2aryTe/vi3ZTNamVoYmKbsiq5ixIhkmOx8NHXvkasX7+wNCWHniSoCFaQm0vDY49x99Sp7Z7TV5v8bNOmBZT9dts2AyIRZpRjsTB+0KA2Zcft9j7ZatGTJEGZwIr8/Asmqb62a29Bbi5pfndR55xOg6IRZvTdWbMCymRJstCSBGUSK/LzeTkvj/T+/YM+f7S+noHLl/epWe9f97uLOnP+fJ+6ftGzbFYrA+LiAspX795tQDTmJAnKRGxWK8cefrjdZZLONjUxs7CQ9J/8pE/0x1iSkxnUrx9RgAIcTqdsCS9C6rbLLw8oc3g8fbJZvSdIgjKhgtxcls6eTXuD0Y83NPSJFSis6enERUXhwTtopNbhoFJm/YsQWpGfH3SCuOxLFhqSoEyqIDeXv9lsQVdhbmb2tepyLBYyhw4lJiqK/rGxxEVHU37mjNFhCZNpb4K4rHbefZKgTCzHYmHT179Ocnx8u+f8x0cf9WJEvW9RZiYK76KeDrebPSdPGh2SMKHCBQsCytzAExs29HosZiIJyuRyLBbOPvood0+dSrAZU6ccDtM3RTg9npavZTdU0RNyLJagI2l3Sp9nt0iC6iNW5OfjeuKJoCtQmHmZlo/LywPKZDdU0RNW5OcH/H2dOn/e1M3oPU0SVB+zyWYLWI35WEODQdH0vOvGjg06WOTnmzdL/4AIuU02G5e36o/SWgf9kCQ6RxJUH7R4ypQ2x3UOh2mHxdqsVn6XlxfQvOlwu2XIuegRrXd3dmst/Z7dIAmqD1qYkRFwV1FUWWnaOwqb1cpv8vLaXLNba7YcPWpYTMK8TtTXtzneXV0tzXxdJAmqD8qxWJgVpC9q7sqVpk5SI1NS2pRtOHjQoGiEmd0QZF6UrKbfNZKg+qhNNhv9Y2LalNldLua+8oppk9Qcv43mKs+dk0+2IuQKcnMDlkCS1fS7RhJUH/btmTMDyuxOJ2/t22dAND0vZ/jwls0Mo4AopaQDW/SIB2bMCCj7aVGRAZFENklQfVhBbi7jWnXoNjtm0uWArOnppCUmovAufRStFNeNHWt0WMKECnJzSfdbTX/vqVNyF3WRJEH1caUPPURqQkKbstdKSky5Tl+OxcIbt9/OP0+ZQu6YMbw4bx42q9XosIRJTR4yJKDsjZISAyKJXDEXPkWYXVZaGh/7DRhYWVzMyuJiJqemUrhgATkdbDEfSXIsFtNciwhvizIzA/6ujvmN8BMdkzsowaLMzHaf21NTw8zCQlPeUQnRk2xWKznDh7cp2ytDzi+KJCiBzWplyuDBHZ6zsrjYtJN5hegpXxo5ss2xxvwLNIeSJCgBwK577w26Tl9rsseNEBdnYUYG8VFt32ZPORzc/NprBkUUWSRBiRabbDY222wBzRKtfVha2osRCRHZciwWvnvVVQHlH5WVmXa+YShJghJt5FgsbP7GN9DtrHw+1G+hWSFExwpycxnrN50jLiZG1oLsBElQol2bbDZuGDeuTdnGw4flk58QF+n1hQtbNg6NVoqRyclY09MNjir8SYISHbpiyJA2laSuqYm5K1fKSCQhLkKOxcJHd93Fw1ddxb/PnMmK/HyZ7tAJMg9KdGhhRgY/37wZj9YtZXaXi2+sXcvzmzez6957DYxOiMghc/AuntxBiQ7lWCyMv/TSoM/trq5m/PPP93JEQoi+QhKUuKDvzpoVdFdagLLaWq745S97NR4hRN8gCUpcUPOutMmxsUGf311dLfM6hBAhJwlKdIrNauXs0qUsnT076PMflJbK6D4hREhJghIXpSA3l5fz8gLKNXDnW2/1fkBCCNOSBCUums1qDXondbC2VpZCEkKEjCQo0SUFubncPXVqQPmf9u41IBohhBlJghJdtiI/P2AJl1GXXGJMMEII05EEJbrl9YULSUtKon9MDGlJSTw1Z47RIQkhTEJWkhDdkmOxsOaOO9heVYU1PV1mygshQkYSlOg2WcJFCNETpIlPCCFEWJIEJYQQIixJghJCCBGWlG61jYJhQShVDRzyK04FagwI52JFQpxGxlijtb7xYl8kdaLHSZ3oXZEQZ9jVibBIUMEopbZprbONjuNCIiHOSIixMyLlOiIhzkiIsTMi5ToiIc5wjFGa+IQQQoQlSVBCCCHCUjgnqN8aHUAnRUKckRBjZ0TKdURCnJEQY2dEynVEQpxhF2PY9kEJIYTo28L5DkoIIUQfJglKCCFEWAq7BKWUulEptV8pVaqUetTgWEYopT5VSu1TSu1RSj3kKx+klPqLUupz37+XtHrN932x71dK3dCLsUYrpXYopd4L1xi7SupEl2M1bZ2A8KkXUid6kNY6bB5ANFAGjAXigGIgw8B40gGr7+sBwAEgA/gx8Kiv/FHgGd/XGb6Y44ExvmuJ7qVY/x14HXjPdxx2MUqdkDphxnohdaLnHuF2BzUDKNVal2utm4DVwAKjgtFaV2mtt/u+PgfsA4b7YlrpO20lkO/7egGwWmvt0FofBErxXlOPUkpZgJuBl1sVh1WM3SB1ogtMXicgjOqF1ImeE24JajhwpNVxpa/McEqp0cA0YAswVGtdBd7KCQzxnWZU/D8HHgE8rcrCLcauCtt4pU4YKixjljoRWuGWoFSQMsPHwSulkoC3gG9rres6OjVIWY/Gr5S6BTiptf6/zr4kSJnhP+MOhGW8UicMF3YxS50IvXDbsLASGNHq2AIcMygWAJRSsXgr3Wta6z/5ik8opdK11lVKqXTgpK/ciPivBuYrpeYB/YBkpdSrYRZjd4RdvFInwkJYxSx1oof0dqfXBTrwYoByvJ1yzR2fkw2MRwGvAD/3K3+Wth2LP/Z9PZm2HYvl9GLHIjCHLzo/wzJGqRNSJ8xWL6RO9GCsRle0ID+8eXhHwZQBywyOZTbe29pdwE7fYx5wKbAe+Nz376BWr1nmi30/cFMvx9u64oVljFInpE6YrV5Inei5hyx1JIQQIiyF2yAJIYQQApAEJYQQIkxJghJCCBGWJEEJIYQIS5KghBBChCVJUEIIIcKSJCghhBBh6f8DoFxRXuXm8PwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x216 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "f, axes = plt.subplots(1, 3,figsize=(6,3))\n",
    "\n",
    "axes[0].get_shared_y_axes().join(*axes)\n",
    "axes[0].get_shared_x_axes().join(*axes)\n",
    "\n",
    "rank_plot_pvalue(df_6h,  axes[0],genes=1,title='6h')\n",
    "rank_plot_pvalue(df_72h, axes[1],genes=1,title='72h')\n",
    "rank_plot_pvalue(df_120h,axes[2],genes=1,title='120h')\n",
    "\n",
    "for i in [1,2]:\n",
    "    axes[i].set_ylabel(None)\n",
    "    axes[i].tick_params(axis='y',label1On=False)\n",
    "\n",
    "f.tight_layout()\n",
    "# f.savefig('')\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5768da91-ec28-4032-8b3c-10157c5689c3",
   "metadata": {},
   "source": [
    "## `results`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 540,
   "id": "292ea253-0253-4edf-8bb4-388086acbca6",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_6h   = pd.read_csv('iso/hl60_6h_delta_iso/jscs/sigGenes.results.txt.gz',sep='\\t')\n",
    "# df_72h  = pd.read_csv('iso/hl60_72h_delta_iso/jscs/sigGenes.genewiseResults.txt.gz',sep='\\t')\n",
    "# df_120h = pd.read_csv('iso/hl60_120h_delta_iso/jscs/sigGenes.genewiseResults.txt.gz',sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 552,
   "id": "515793bc-6fdf-432f-ae59-bf8c6c286120",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'exonic_part', 'splice_site'}"
      ]
     },
     "execution_count": 552,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(df_6h.featureType)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 555,
   "id": "e22a9716-6c4a-429b-8845-5b82eaa98ff7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['featureID', 'geneID', 'countbinID', 'testable', 'status', 'allZero',\n",
       "       'baseMean', 'baseVar', 'dispBeforeSharing', 'dispFitted', 'dispersion',\n",
       "       'pvalue', 'padjust', 'chr', 'start', 'end', 'strand', 'transcripts',\n",
       "       'featureType', 'padjust_noFilter', 'log2FC(Decitabine/DMSO)',\n",
       "       'log2FCvst(Decitabine/DMSO)', 'expr_DMSO', 'expr_Decitabine',\n",
       "       'geneWisePadj'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 555,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_6h.columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 550,
   "id": "801d9210-6d71-435c-8c0f-ff782f263888",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>featureID</th>\n",
       "      <th>geneID</th>\n",
       "      <th>countbinID</th>\n",
       "      <th>testable</th>\n",
       "      <th>status</th>\n",
       "      <th>allZero</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>baseVar</th>\n",
       "      <th>dispBeforeSharing</th>\n",
       "      <th>dispFitted</th>\n",
       "      <th>...</th>\n",
       "      <th>end</th>\n",
       "      <th>strand</th>\n",
       "      <th>transcripts</th>\n",
       "      <th>featureType</th>\n",
       "      <th>padjust_noFilter</th>\n",
       "      <th>log2FC(Decitabine/DMSO)</th>\n",
       "      <th>log2FCvst(Decitabine/DMSO)</th>\n",
       "      <th>expr_DMSO</th>\n",
       "      <th>expr_Decitabine</th>\n",
       "      <th>geneWisePadj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000116288.13:E001</td>\n",
       "      <td>ENSG00000116288.13</td>\n",
       "      <td>E001</td>\n",
       "      <td>False</td>\n",
       "      <td>LOW_COUNTS_INDEP_FILTER</td>\n",
       "      <td>False</td>\n",
       "      <td>1.576026</td>\n",
       "      <td>3.179682</td>\n",
       "      <td>0.131963</td>\n",
       "      <td>0.206330</td>\n",
       "      <td>...</td>\n",
       "      <td>7954323</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000493373.5</td>\n",
       "      <td>exonic_part</td>\n",
       "      <td>0.961383</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.652160e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000116288.13:E002</td>\n",
       "      <td>ENSG00000116288.13</td>\n",
       "      <td>E002</td>\n",
       "      <td>True</td>\n",
       "      <td>OK</td>\n",
       "      <td>False</td>\n",
       "      <td>471.911424</td>\n",
       "      <td>3594.830714</td>\n",
       "      <td>0.000309</td>\n",
       "      <td>0.006006</td>\n",
       "      <td>...</td>\n",
       "      <td>7961691</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000460192.5</td>\n",
       "      <td>exonic_part</td>\n",
       "      <td>0.177438</td>\n",
       "      <td>-0.321035</td>\n",
       "      <td>-0.160502</td>\n",
       "      <td>521.331754</td>\n",
       "      <td>422.658770</td>\n",
       "      <td>3.652160e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000116288.13:E003</td>\n",
       "      <td>ENSG00000116288.13</td>\n",
       "      <td>E003</td>\n",
       "      <td>True</td>\n",
       "      <td>OK</td>\n",
       "      <td>False</td>\n",
       "      <td>506.409314</td>\n",
       "      <td>3352.794505</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.005960</td>\n",
       "      <td>...</td>\n",
       "      <td>7961700</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000460192.5;ENST00000493678.5</td>\n",
       "      <td>exonic_part</td>\n",
       "      <td>0.223329</td>\n",
       "      <td>-0.293744</td>\n",
       "      <td>-0.146856</td>\n",
       "      <td>554.488998</td>\n",
       "      <td>457.948682</td>\n",
       "      <td>3.652160e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000116288.13:E004</td>\n",
       "      <td>ENSG00000116288.13</td>\n",
       "      <td>E004</td>\n",
       "      <td>True</td>\n",
       "      <td>OK</td>\n",
       "      <td>False</td>\n",
       "      <td>600.378737</td>\n",
       "      <td>4669.088097</td>\n",
       "      <td>0.002974</td>\n",
       "      <td>0.005862</td>\n",
       "      <td>...</td>\n",
       "      <td>7961710</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000460192.5;ENST00000493678.5;ENST000003...</td>\n",
       "      <td>exonic_part</td>\n",
       "      <td>0.650891</td>\n",
       "      <td>-0.273228</td>\n",
       "      <td>-0.136597</td>\n",
       "      <td>651.198600</td>\n",
       "      <td>546.320694</td>\n",
       "      <td>3.652160e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000116288.13:E005</td>\n",
       "      <td>ENSG00000116288.13</td>\n",
       "      <td>E005</td>\n",
       "      <td>True</td>\n",
       "      <td>OK</td>\n",
       "      <td>False</td>\n",
       "      <td>1837.592820</td>\n",
       "      <td>29454.363452</td>\n",
       "      <td>0.006440</td>\n",
       "      <td>0.005507</td>\n",
       "      <td>...</td>\n",
       "      <td>7961735</td>\n",
       "      <td>+</td>\n",
       "      <td>ENST00000460192.5;ENST00000493678.5;ENST000003...</td>\n",
       "      <td>exonic_part</td>\n",
       "      <td>0.967760</td>\n",
       "      <td>-0.123193</td>\n",
       "      <td>-0.061567</td>\n",
       "      <td>1898.335731</td>\n",
       "      <td>1776.812347</td>\n",
       "      <td>3.652160e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28169</th>\n",
       "      <td>ENSG00000184216.14:J058</td>\n",
       "      <td>ENSG00000184216.14</td>\n",
       "      <td>J058</td>\n",
       "      <td>False</td>\n",
       "      <td>LOW_COUNTS_INDEP_FILTER</td>\n",
       "      <td>False</td>\n",
       "      <td>23.324141</td>\n",
       "      <td>217.160726</td>\n",
       "      <td>0.007160</td>\n",
       "      <td>0.017860</td>\n",
       "      <td>...</td>\n",
       "      <td>154018598</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000429936.6;ENST00000393687.6;ENST000004...</td>\n",
       "      <td>splice_site</td>\n",
       "      <td>0.010901</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.710530e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28170</th>\n",
       "      <td>ENSG00000184216.14:J059</td>\n",
       "      <td>ENSG00000184216.14</td>\n",
       "      <td>J059</td>\n",
       "      <td>False</td>\n",
       "      <td>LOW_COUNTS_INDEP_FILTER</td>\n",
       "      <td>False</td>\n",
       "      <td>16.178014</td>\n",
       "      <td>37.415212</td>\n",
       "      <td>0.011538</td>\n",
       "      <td>0.022815</td>\n",
       "      <td>...</td>\n",
       "      <td>154018974</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000429936.6;ENST00000393687.6;ENST000003...</td>\n",
       "      <td>splice_site</td>\n",
       "      <td>0.710060</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.710530e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28171</th>\n",
       "      <td>ENSG00000184216.14:J060</td>\n",
       "      <td>ENSG00000184216.14</td>\n",
       "      <td>J060</td>\n",
       "      <td>True</td>\n",
       "      <td>OK</td>\n",
       "      <td>False</td>\n",
       "      <td>38.121679</td>\n",
       "      <td>99.792823</td>\n",
       "      <td>0.019023</td>\n",
       "      <td>0.013506</td>\n",
       "      <td>...</td>\n",
       "      <td>154019196</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000429936.6;ENST00000393687.6;ENST000003...</td>\n",
       "      <td>splice_site</td>\n",
       "      <td>0.966168</td>\n",
       "      <td>0.296873</td>\n",
       "      <td>0.148433</td>\n",
       "      <td>32.593999</td>\n",
       "      <td>42.151988</td>\n",
       "      <td>3.710530e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28172</th>\n",
       "      <td>ENSG00000184216.14:J061</td>\n",
       "      <td>ENSG00000184216.14</td>\n",
       "      <td>J061</td>\n",
       "      <td>False</td>\n",
       "      <td>LOW_COUNTS_INDEP_FILTER</td>\n",
       "      <td>False</td>\n",
       "      <td>9.919710</td>\n",
       "      <td>8.519666</td>\n",
       "      <td>0.000514</td>\n",
       "      <td>0.033018</td>\n",
       "      <td>...</td>\n",
       "      <td>154019430</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000429936.6;ENST00000393687.6;ENST000003...</td>\n",
       "      <td>splice_site</td>\n",
       "      <td>0.959519</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.710530e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28173</th>\n",
       "      <td>ENSG00000184216.14:J062</td>\n",
       "      <td>ENSG00000184216.14</td>\n",
       "      <td>J062</td>\n",
       "      <td>False</td>\n",
       "      <td>LOW_COUNTS_INDEP_FILTER</td>\n",
       "      <td>False</td>\n",
       "      <td>7.432055</td>\n",
       "      <td>6.336268</td>\n",
       "      <td>0.000535</td>\n",
       "      <td>0.041847</td>\n",
       "      <td>...</td>\n",
       "      <td>154019676</td>\n",
       "      <td>-</td>\n",
       "      <td>ENST00000393687.6;ENST00000369980.8;ENST000004...</td>\n",
       "      <td>splice_site</td>\n",
       "      <td>0.966168</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.710530e-08</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>25310 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     featureID              geneID countbinID  testable  \\\n",
       "0      ENSG00000116288.13:E001  ENSG00000116288.13       E001     False   \n",
       "1      ENSG00000116288.13:E002  ENSG00000116288.13       E002      True   \n",
       "2      ENSG00000116288.13:E003  ENSG00000116288.13       E003      True   \n",
       "3      ENSG00000116288.13:E004  ENSG00000116288.13       E004      True   \n",
       "4      ENSG00000116288.13:E005  ENSG00000116288.13       E005      True   \n",
       "...                        ...                 ...        ...       ...   \n",
       "28169  ENSG00000184216.14:J058  ENSG00000184216.14       J058     False   \n",
       "28170  ENSG00000184216.14:J059  ENSG00000184216.14       J059     False   \n",
       "28171  ENSG00000184216.14:J060  ENSG00000184216.14       J060      True   \n",
       "28172  ENSG00000184216.14:J061  ENSG00000184216.14       J061     False   \n",
       "28173  ENSG00000184216.14:J062  ENSG00000184216.14       J062     False   \n",
       "\n",
       "                        status  allZero     baseMean       baseVar  \\\n",
       "0      LOW_COUNTS_INDEP_FILTER    False     1.576026      3.179682   \n",
       "1                           OK    False   471.911424   3594.830714   \n",
       "2                           OK    False   506.409314   3352.794505   \n",
       "3                           OK    False   600.378737   4669.088097   \n",
       "4                           OK    False  1837.592820  29454.363452   \n",
       "...                        ...      ...          ...           ...   \n",
       "28169  LOW_COUNTS_INDEP_FILTER    False    23.324141    217.160726   \n",
       "28170  LOW_COUNTS_INDEP_FILTER    False    16.178014     37.415212   \n",
       "28171                       OK    False    38.121679     99.792823   \n",
       "28172  LOW_COUNTS_INDEP_FILTER    False     9.919710      8.519666   \n",
       "28173  LOW_COUNTS_INDEP_FILTER    False     7.432055      6.336268   \n",
       "\n",
       "       dispBeforeSharing  dispFitted  ...        end  strand  \\\n",
       "0               0.131963    0.206330  ...    7954323       +   \n",
       "1               0.000309    0.006006  ...    7961691       +   \n",
       "2               0.000095    0.005960  ...    7961700       +   \n",
       "3               0.002974    0.005862  ...    7961710       +   \n",
       "4               0.006440    0.005507  ...    7961735       +   \n",
       "...                  ...         ...  ...        ...     ...   \n",
       "28169           0.007160    0.017860  ...  154018598       -   \n",
       "28170           0.011538    0.022815  ...  154018974       -   \n",
       "28171           0.019023    0.013506  ...  154019196       -   \n",
       "28172           0.000514    0.033018  ...  154019430       -   \n",
       "28173           0.000535    0.041847  ...  154019676       -   \n",
       "\n",
       "                                             transcripts  featureType  \\\n",
       "0                                      ENST00000493373.5  exonic_part   \n",
       "1                                      ENST00000460192.5  exonic_part   \n",
       "2                    ENST00000460192.5;ENST00000493678.5  exonic_part   \n",
       "3      ENST00000460192.5;ENST00000493678.5;ENST000003...  exonic_part   \n",
       "4      ENST00000460192.5;ENST00000493678.5;ENST000003...  exonic_part   \n",
       "...                                                  ...          ...   \n",
       "28169  ENST00000429936.6;ENST00000393687.6;ENST000004...  splice_site   \n",
       "28170  ENST00000429936.6;ENST00000393687.6;ENST000003...  splice_site   \n",
       "28171  ENST00000429936.6;ENST00000393687.6;ENST000003...  splice_site   \n",
       "28172  ENST00000429936.6;ENST00000393687.6;ENST000003...  splice_site   \n",
       "28173  ENST00000393687.6;ENST00000369980.8;ENST000004...  splice_site   \n",
       "\n",
       "       padjust_noFilter  log2FC(Decitabine/DMSO) log2FCvst(Decitabine/DMSO)  \\\n",
       "0              0.961383                      NaN                        NaN   \n",
       "1              0.177438                -0.321035                  -0.160502   \n",
       "2              0.223329                -0.293744                  -0.146856   \n",
       "3              0.650891                -0.273228                  -0.136597   \n",
       "4              0.967760                -0.123193                  -0.061567   \n",
       "...                 ...                      ...                        ...   \n",
       "28169          0.010901                      NaN                        NaN   \n",
       "28170          0.710060                      NaN                        NaN   \n",
       "28171          0.966168                 0.296873                   0.148433   \n",
       "28172          0.959519                      NaN                        NaN   \n",
       "28173          0.966168                      NaN                        NaN   \n",
       "\n",
       "         expr_DMSO expr_Decitabine  geneWisePadj  \n",
       "0              NaN             NaN  3.652160e-09  \n",
       "1       521.331754      422.658770  3.652160e-09  \n",
       "2       554.488998      457.948682  3.652160e-09  \n",
       "3       651.198600      546.320694  3.652160e-09  \n",
       "4      1898.335731     1776.812347  3.652160e-09  \n",
       "...            ...             ...           ...  \n",
       "28169          NaN             NaN  3.710530e-08  \n",
       "28170          NaN             NaN  3.710530e-08  \n",
       "28171    32.593999       42.151988  3.710530e-08  \n",
       "28172          NaN             NaN  3.710530e-08  \n",
       "28173          NaN             NaN  3.710530e-08  \n",
       "\n",
       "[25310 rows x 25 columns]"
      ]
     },
     "execution_count": 550,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_6h[~df_6h.allZero]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8741091a-e44b-492b-8b0a-b3b063ac47d6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6aaaa0aa-4a18-48e6-8ae7-ca84152a89c2",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da5ca9a9-827e-4608-b41a-e9d97c23484a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genewiseResults 2 binary "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "c294365a-77bc-4ce4-b994-0a0650df390b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sun Feb  6 20:54:41 PST 2022\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "fd495408-a87a-43df-b755-9fbd131e00a1",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:down]",
   "language": "python",
   "name": "conda-env-down-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
