{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> We have already established that using annotated HeLa m6A sites, we can observe changes in genes with m6A sites (HL-60) cells. In order to confirm this m6A sites, we performed MeRIP-seq in treated and untreated cells, and did observe a general increase in m6A levels upon treatments for a large number of annotated sites. Here, our goal is to indpendently analyze the MeRIP data without relying on HeLa annotations and use it to define a **treatment-induced hyper-methylation sites**. We will then assess the location and behaviour of these targets across the other datasets generated in this study."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Test enrichment of treatment-induced hyper/hypo-methylation sites\n",
    "\n",
    "### Goal\n",
    "Here, I aim to identify the genes that are hyper or hypo methylated as genesets, and see if they have enriched accross all datasets; the input table is list of genes with control vs. treated fold change of RNA expression, RNA stability and translational efficiency. \n",
    "### Steps \n",
    "1. Prepare inputs  \n",
    "    - Filtering genes with $\\Delta$methylation >= 1 as hyper-methylation sites (P-Value < 0.01)\n",
    "    - Filtering genes with $\\Delta$methylation <= -1 as hypo-methylation sites (P-Value < 0.01)\n",
    "2. Run `run_mi_gene_list.pl` command \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd \n",
    "import numpy as np\n",
    "    \n",
    "def write_gene_file(df,file_name):\n",
    "    \n",
    "    df = pd.DataFrame({'ensembl':[ens[:-3] for ens in df.ensembl.tolist()]}).drop_duplicates('ensembl')\n",
    "    df.to_csv(file_name,sep='\\t',index=None,header=None)\n",
    "    \n",
    "\n",
    "hyper, hypo = two_sided_mtyl()\n",
    "write_gene_file(hyper,'mtyl-enrichment/hyper_mtyl.txt')\n",
    "write_gene_file(hypo,'mtyl-enrichment/hypo_mtyl.txt')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Using a [TEISER](https://github.com/goodarzilab/TEISER) script to do enrichment test \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../Ribo-seq/hl60_delta_te.txt hyper_mtyl\n",
      "done!\n",
      "../Ribo-seq/hl60_delta_te.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-exp/6h_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-exp/6h_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-exp/72h_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-exp/72h_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-exp/120h_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-exp/120h_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-stbl/120h_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-stbl/120h_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-stbl/6h_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/hl60-stbl/6h_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/kg1_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/kg1_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/molm14_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/molm14_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/ociaml2_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/ociaml2_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/ociaml3_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/ociaml3_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/thp1_delta_exp.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-exp/thp1_delta_exp.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/kg1_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/kg1_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/molm14_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/molm14_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/ociaml2_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/ociaml2_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/ociaml3_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/ociaml3_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/thp1_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/other-stbl/thp1_delta_stbl.txt hypo_mtyl\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "declare -a Genesets=('hyper_mtyl' 'hypo_mtyl')\n",
    "declare -a Experiments=(\n",
    "# Ribo-seq\n",
    "'../Ribo-seq/hl60_delta_te.txt'\n",
    "\n",
    "## HL-60 RNA-seq \n",
    "# RNA experssion \n",
    "'../RNA-seq/hl60-exp/6h_delta_exp.txt' \n",
    "'../RNA-seq/hl60-exp/72h_delta_exp.txt' \n",
    "'../RNA-seq/hl60-exp/120h_delta_exp.txt' \n",
    "# RNA stability  \n",
    "'../RNA-seq/hl60-stbl/120h_delta_stbl.txt'  \n",
    "'../RNA-seq/hl60-stbl/6h_delta_stbl.txt'\n",
    "\n",
    "## 5 other AML cell lines RNA-seq\n",
    "# RNA experssion \n",
    "'../RNA-seq/other-exp/kg1_delta_exp.txt' \n",
    "'../RNA-seq/other-exp/molm14_delta_exp.txt'\n",
    "'../RNA-seq/other-exp/ociaml2_delta_exp.txt' \n",
    "'../RNA-seq/other-exp/ociaml3_delta_exp.txt'\n",
    "'../RNA-seq/other-exp/thp1_delta_exp.txt'\n",
    "# RNA stability  \n",
    "'../RNA-seq/other-stbl/kg1_delta_stbl.txt' \n",
    "'../RNA-seq/other-stbl/molm14_delta_stbl.txt' \n",
    "'../RNA-seq/other-stbl/ociaml2_delta_stbl.txt' \n",
    "'../RNA-seq/other-stbl/ociaml3_delta_stbl.txt'\n",
    "'../RNA-seq/other-stbl/thp1_delta_stbl.txt'\n",
    ")\n",
    "\n",
    "for exp in \"${Experiments[@]}\"; do\n",
    "    for geneset in \"${Genesets[@]}\"; do\n",
    "    \n",
    "        echo $exp $geneset\n",
    "        base=`basename $exp`\n",
    "        base=${base/.txt/}\n",
    "        \n",
    "        # get intersect \n",
    "        awk 'NR==FNR{A[$1];next}$1 in A' $exp mtyl-enrichment/${geneset}.txt > mtyl-enrichment/${geneset}_${base}.txt\n",
    "        \n",
    "        perl $TEISERDIR/run_mi_gene_list.pl \\\n",
    "            --expfile=$exp \\\n",
    "            --genefile=mtyl-enrichment/${geneset}_${base}.txt \\\n",
    "            --exptype=continuous \\\n",
    "            --ebins=7 \\\n",
    "            --species=human \\\n",
    "            --doremovedups=0 \\\n",
    "            --doremoveextra=0 &> mtyl-enrichment/${geneset}_${base}.log\n",
    "#         # remove results from previous run \n",
    "#         rm -fr mtyl-enrichment/${geneset}_${base}_GENESET\n",
    "        \n",
    "        rm mtyl-enrichment/${geneset}_${base}.txt\n",
    "        mv ${exp}_GENESET mtyl-enrichment/${geneset}_${base}_GENESET\n",
    "        \n",
    "        echo 'done!'\n",
    "        \n",
    "    done \n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Redraw heatmaps using `--min=-3 --max=3` thresholds for those plots which have smaller range of signals:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6h_delta_stbl hyper_mtyl\n",
      "done!\n",
      "6h_delta_stbl hypo_mtyl\n",
      "done!\n",
      "120h_delta_stbl hyper_mtyl\n",
      "done!\n",
      "120h_delta_stbl hypo_mtyl\n",
      "done!\n",
      "kg1_delta_stbl hyper_mtyl\n",
      "done!\n",
      "kg1_delta_stbl hypo_mtyl\n",
      "done!\n",
      "ociaml2_delta_stbl hyper_mtyl\n",
      "done!\n",
      "ociaml2_delta_stbl hypo_mtyl\n",
      "done!\n",
      "molm14_delta_stbl hyper_mtyl\n",
      "done!\n",
      "molm14_delta_stbl hypo_mtyl\n",
      "done!\n",
      "ociaml3_delta_stbl hyper_mtyl\n",
      "done!\n",
      "ociaml3_delta_stbl hypo_mtyl\n",
      "done!\n",
      "thp1_delta_stbl hyper_mtyl\n",
      "done!\n",
      "thp1_delta_stbl hypo_mtyl\n",
      "done!\n",
      "hl60_delta_te hyper_mtyl\n",
      "done!\n",
      "hl60_delta_te hypo_mtyl\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "declare -a Genesets=('hyper_mtyl' 'hypo_mtyl')\n",
    "declare -a Experiments=(\n",
    "'6h_delta_stbl' '120h_delta_stbl' \n",
    "'kg1_delta_stbl' 'ociaml2_delta_stbl' 'molm14_delta_stbl' \n",
    "'ociaml3_delta_stbl' 'thp1_delta_stbl'\n",
    "'hl60_delta_te'\n",
    ")\n",
    "for exp in \"${Experiments[@]}\"; do\n",
    "    for geneset in \"${Genesets[@]}\"; do\n",
    "        echo $exp $geneset    \n",
    "        perl $TEISERDIR/Scripts/teiser_draw_matrix.pl \\\n",
    "        --pvmatrixfile=mtyl-enrichment/${geneset}_${exp}_GENESET/${exp}.txt.matrix \\\n",
    "        --summaryfile=mtyl-enrichment/${geneset}_${exp}_GENESET/${exp}.txt.summary \\\n",
    "        --expfile=mtyl-enrichment/${geneset}_${exp}_GENESET/${exp}.txt \\\n",
    "        --quantized=0 \\\n",
    "        --colmap=$TEISERDIR/Scripts/HEATMAPS/cmap_1.txt \\\n",
    "        --order=0 \\\n",
    "        --min=-3 --max=3 \\\n",
    "        --cluster=5 &>> mtyl-enrichment/${geneset}_${exp}.log\n",
    "        \n",
    "        echo 'done!'\n",
    "        \n",
    "    done \n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Make `png` figures:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "mkdir -p mtyl-enrichment/plots/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mtyl_enrichment/hyper_mtyl_120h_delta_exp_GENESET/120h_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_120h_delta_exp_GENESET/120h_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_120h_delta_stbl_GENESET/120h_delta_stbl.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_120h_delta_stbl_GENESET/120h_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_6h_delta_exp_GENESET/6h_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_6h_delta_exp_GENESET/6h_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_6h_delta_stbl_GENESET/6h_delta_stbl.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_6h_delta_stbl_GENESET/6h_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_72h_delta_exp_GENESET/72h_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_72h_delta_exp_GENESET/72h_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_hl60_delta_te_GENESET/hl60_delta_te.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_hl60_delta_te_GENESET/hl60_delta_te.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_kg1_delta_exp_GENESET/kg1_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_kg1_delta_exp_GENESET/kg1_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_kg1_delta_stbl_GENESET/kg1_delta_stbl.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_kg1_delta_stbl_GENESET/kg1_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_molm14_delta_exp_GENESET/molm14_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_molm14_delta_exp_GENESET/molm14_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_molm14_delta_stbl_GENESET/molm14_delta_stbl.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_molm14_delta_stbl_GENESET/molm14_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_ociaml2_delta_exp_GENESET/ociaml2_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_ociaml2_delta_exp_GENESET/ociaml2_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_ociaml2_delta_stbl_GENESET/ociaml2_delta_stbl.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_ociaml2_delta_stbl_GENESET/ociaml2_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_ociaml3_delta_exp_GENESET/ociaml3_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_ociaml3_delta_exp_GENESET/ociaml3_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_ociaml3_delta_stbl_GENESET/ociaml3_delta_stbl.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_ociaml3_delta_stbl_GENESET/ociaml3_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_thp1_delta_exp_GENESET/thp1_delta_exp.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_thp1_delta_exp_GENESET/thp1_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hyper_mtyl_thp1_delta_stbl_GENESET/thp1_delta_stbl.txt.summary.pdf > mtyl_enrichment/hyper_mtyl_thp1_delta_stbl_GENESET/thp1_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_120h_delta_exp_GENESET/120h_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_120h_delta_exp_GENESET/120h_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_120h_delta_stbl_GENESET/120h_delta_stbl.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_120h_delta_stbl_GENESET/120h_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_6h_delta_exp_GENESET/6h_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_6h_delta_exp_GENESET/6h_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_6h_delta_stbl_GENESET/6h_delta_stbl.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_6h_delta_stbl_GENESET/6h_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_72h_delta_exp_GENESET/72h_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_72h_delta_exp_GENESET/72h_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_hl60_delta_te_GENESET/hl60_delta_te.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_hl60_delta_te_GENESET/hl60_delta_te.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_kg1_delta_exp_GENESET/kg1_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_kg1_delta_exp_GENESET/kg1_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_kg1_delta_stbl_GENESET/kg1_delta_stbl.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_kg1_delta_stbl_GENESET/kg1_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_molm14_delta_exp_GENESET/molm14_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_molm14_delta_exp_GENESET/molm14_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_molm14_delta_stbl_GENESET/molm14_delta_stbl.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_molm14_delta_stbl_GENESET/molm14_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_ociaml2_delta_exp_GENESET/ociaml2_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_ociaml2_delta_exp_GENESET/ociaml2_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_ociaml2_delta_stbl_GENESET/ociaml2_delta_stbl.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_ociaml2_delta_stbl_GENESET/ociaml2_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_ociaml3_delta_exp_GENESET/ociaml3_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_ociaml3_delta_exp_GENESET/ociaml3_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_ociaml3_delta_stbl_GENESET/ociaml3_delta_stbl.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_ociaml3_delta_stbl_GENESET/ociaml3_delta_stbl.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_thp1_delta_exp_GENESET/thp1_delta_exp.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_thp1_delta_exp_GENESET/thp1_delta_exp.txt.summary.png\n",
      "done!\n",
      "mtyl_enrichment/hypo_mtyl_thp1_delta_stbl_GENESET/thp1_delta_stbl.txt.summary.pdf > mtyl_enrichment/hypo_mtyl_thp1_delta_stbl_GENESET/thp1_delta_stbl.txt.summary.png\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "for pdf in mtyl-enrichment/*_GENESET/*.txt.summary.pdf; do \n",
    "    png=${pdf/.pdf/.png}\n",
    "    di=`dirname $pdf`\n",
    "    out=`basename $di`\n",
    "    outpng=${out/_GENESET/.png}\n",
    "    outpdf=${out/_GENESET/.pdf}\n",
    "    \n",
    "    bash /rumi/shams/abe/GitHub/Abe/my_scripts/pdf2png.sh $pdf\n",
    "    cp $pdf mtyl-enrichment/$outpdf\n",
    "    mv $png mtyl-enrichment/$outpng\n",
    "done "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "5. Write README.md draft\n",
    "    - Write HTML codes which link all plots into a `README.md` format to prepare GitHub friendly report"
   ]
  },
  {
   "cell_type": "raw",
   "metadata": {},
   "source": [
    "%%bash \n",
    "readme='mtyl-enrichment.md'\n",
    "touch $readme\n",
    "for f in mtyl-enrichment/*.png; do \n",
    "    b=`basename $f`\n",
    "    t=${b/.png/}\n",
    "    echo '#### '$t >> $readme\n",
    "    echo -e \"<img src=\\\"\"$f\"\\\" title=\\\"\"$t\"\\\" style=\\\"width:1000px\\\">\\n\" >> $readme\n",
    "done "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:deseq]",
   "language": "python",
   "name": "conda-env-deseq-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
