{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> We have already established that using annotated HeLa m6A sites, we can observe changes in genes with m6A sites (HL-60) cells. In order to confirm this m6A sites, we performed MeRIP-seq in treated and untreated cells, and did observe a general increase in m6A levels upon treatments for a large number of annotated sites. Here, our goal is to indpendently analyze the MeRIP data without relying on HeLa annotations and use it to define a **treatment-induced hyper-methylation sites**. We will then assess the location and behaviour of these targets across the other datasets generated in this study."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "jp-MarkdownHeadingCollapsed": true,
    "tags": []
   },
   "source": [
    "## Test enrichment of treatment-induced hyper/hypo-methylation sites\n",
    "\n",
    "### Goal\n",
    "Here, I aim to identify the genes that are hyper or hypo methylated as genesets, and see if they have enriched accross all datasets; the input table is list of genes with control vs. treated fold change of RNA expression, RNA stability and translational efficiency. \n",
    "### Steps \n",
    "1. Prepare inputs  \n",
    "    - Filtering genes with $\\Delta$methylation >= 2 as hyper-methylation sites (P-Value < 0.01)\n",
    "    - Filtering genes with $\\Delta$methylation <= -2 as hypo-methylation sites (P-Value < 0.01)\n",
    "2. Run `run_mi_gene_list.pl` command \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os \n",
    "from glob import glob \n",
    "import sys \n",
    "import pandas as pd \n",
    "import numpy as np\n",
    "sys.path.append('../')\n",
    "\n",
    "from util import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "for exp in [\n",
    "    '../RNA-seq/exp/hl60_6h_delta_exp.txt',\n",
    "    '../RNA-seq/exp/hl60_72h_delta_exp.txt',\n",
    "    '../RNA-seq/exp/hl60_72h_only_delta_exp.txt',\n",
    "    '../RNA-seq/exp/hl60_120h_delta_exp.txt',\n",
    "    '../RNA-seq/exp/kg1_delta_exp.txt',\n",
    "    '../RNA-seq/exp/molm14_delta_exp.txt',\n",
    "    '../RNA-seq/exp/ociaml2_delta_exp.txt',\n",
    "    '../RNA-seq/exp/ociaml3_delta_exp.txt',\n",
    "    '../RNA-seq/exp/thp1_delta_exp.txt',\n",
    "    '../DAC-rg3039/RNA-seq/comb_vs_decitabine_delta_exp.txt',\n",
    "    '../DAC-rg3039/RNA-seq/comb_vs_dmso_delta_exp.txt'\n",
    "\n",
    "]: \n",
    "    pd.read_csv(exp,sep='\\t').drop_duplicates('gene_name').to_csv(exp.replace('.txt','.c.txt'),sep='\\t',index=None)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 4.17 s, sys: 3.62 ms, total: 4.18 s\n",
      "Wall time: 4.24 s\n"
     ]
    }
   ],
   "source": [
    "%%time \n",
    "data = pd.read_csv('../meRIP-seq/hl60_delta_mtyl_table.txt',sep='\\t').loc[:,['ensembl','name','logFC','p_value']]\n",
    "data = data.iloc[[int(data[(data.ensembl == gene)].logFC.abs().idxmax()) for gene in set(data.ensembl)],:].set_index('ensembl')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "!mkdir -p mtyl-enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>name</th>\n",
       "      <th>logFC</th>\n",
       "      <th>p_value</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ensembl</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000235513.2</th>\n",
       "      <td>L3MBTL2-AS1</td>\n",
       "      <td>20.814131</td>\n",
       "      <td>2.732964e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000254996.5</th>\n",
       "      <td>ANKHD1-EIF4EBP3</td>\n",
       "      <td>6.746412</td>\n",
       "      <td>4.497806e-08</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000116731.22</th>\n",
       "      <td>PRDM2</td>\n",
       "      <td>6.637258</td>\n",
       "      <td>2.348933e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000005339.15</th>\n",
       "      <td>CREBBP</td>\n",
       "      <td>6.599870</td>\n",
       "      <td>1.802126e-10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000167258.15</th>\n",
       "      <td>CDK12</td>\n",
       "      <td>6.539586</td>\n",
       "      <td>8.036084e-09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000272333.7</th>\n",
       "      <td>KMT2B</td>\n",
       "      <td>2.031875</td>\n",
       "      <td>9.915920e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000004534.15</th>\n",
       "      <td>RBM6</td>\n",
       "      <td>2.029048</td>\n",
       "      <td>3.924308e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000049245.13</th>\n",
       "      <td>VAMP3</td>\n",
       "      <td>2.014903</td>\n",
       "      <td>5.639545e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000166377.20</th>\n",
       "      <td>ATP9B</td>\n",
       "      <td>2.011122</td>\n",
       "      <td>3.052605e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000133561.15</th>\n",
       "      <td>GIMAP6</td>\n",
       "      <td>2.001791</td>\n",
       "      <td>3.691276e-03</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1179 rows Ã— 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                               name      logFC       p_value\n",
       "ensembl                                                     \n",
       "ENSG00000235513.2       L3MBTL2-AS1  20.814131  2.732964e-04\n",
       "ENSG00000254996.5   ANKHD1-EIF4EBP3   6.746412  4.497806e-08\n",
       "ENSG00000116731.22            PRDM2   6.637258  2.348933e-09\n",
       "ENSG00000005339.15           CREBBP   6.599870  1.802126e-10\n",
       "ENSG00000167258.15            CDK12   6.539586  8.036084e-09\n",
       "...                             ...        ...           ...\n",
       "ENSG00000272333.7             KMT2B   2.031875  9.915920e-05\n",
       "ENSG00000004534.15             RBM6   2.029048  3.924308e-03\n",
       "ENSG00000049245.13            VAMP3   2.014903  5.639545e-05\n",
       "ENSG00000166377.20            ATP9B   2.011122  3.052605e-03\n",
       "ENSG00000133561.15           GIMAP6   2.001791  3.691276e-03\n",
       "\n",
       "[1179 rows x 3 columns]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "hyper, hypo = two_sided_mtyl(data,fcthr=2)\n",
    "\n",
    "hyper.reindex(\n",
    "    hyper.logFC.abs().sort_values(ascending=False).index\n",
    ")\n",
    "# .to_csv('mtyl-enrichment/hyper_mtyl.txt',sep='\\t',index=None,header=None)\n",
    "\n",
    "# hypo.reindex(\n",
    "#     hypo.logFC.abs().sort_values(ascending=False).index\n",
    "# ).to_csv ('mtyl-enrichment/hypo_mtyl.txt', sep='\\t',index=None,header=None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Using a [TEISER](https://github.com/goodarzilab/TEISER) script to do enrichment test \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "../Ribo-seq/hl60_delta_te.txt hyper_mtyl\n",
      "done!\n",
      "../Ribo-seq/hl60_delta_te.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_6h_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_6h_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_72h_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_72h_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_72h_only_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_72h_only_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_120h_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/hl60_120h_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/hl60_120h_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/hl60_120h_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/hl60_72h_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/hl60_72h_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/hl60_6h_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/hl60_6h_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/kg1_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/kg1_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/molm14_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/molm14_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/ociaml2_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/ociaml2_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/ociaml3_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/ociaml3_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/thp1_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/exp/thp1_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/kg1_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/kg1_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/molm14_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/molm14_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/ociaml2_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/ociaml2_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/ociaml3_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/ociaml3_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/thp1_delta_stbl.txt hyper_mtyl\n",
      "done!\n",
      "../RNA-seq/stbl/thp1_delta_stbl.txt hypo_mtyl\n",
      "done!\n",
      "../DAC-rg3039/RNA-seq/comb_vs_decitabine_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../DAC-rg3039/RNA-seq/comb_vs_decitabine_delta_exp.c.txt hypo_mtyl\n",
      "done!\n",
      "../DAC-rg3039/RNA-seq/comb_vs_dmso_delta_exp.c.txt hyper_mtyl\n",
      "done!\n",
      "../DAC-rg3039/RNA-seq/comb_vs_dmso_delta_exp.c.txt hypo_mtyl\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "# export PAGEDIR='/data_gilbert/home/aarab/iPAGE'\n",
    "# export TEISERDIR='/data_gilbert/home/aarab/iPAGE'\n",
    "\n",
    "declare -a Genesets=('hyper_mtyl' 'hypo_mtyl')\n",
    "declare -a Experiments=(\n",
    "# Ribo-seq\n",
    "'../Ribo-seq/hl60_delta_te.txt'\n",
    "\n",
    "## HL-60 RNA-seq \n",
    "# RNA experssion \n",
    "'../RNA-seq/exp/hl60_6h_delta_exp.c.txt' \n",
    "'../RNA-seq/exp/hl60_72h_delta_exp.c.txt' \n",
    "'../RNA-seq/exp/hl60_72h_only_delta_exp.c.txt' \n",
    "'../RNA-seq/exp/hl60_120h_delta_exp.c.txt' \n",
    "# RNA stability  \n",
    "'../RNA-seq/stbl/hl60_120h_delta_stbl.txt'  \n",
    "'../RNA-seq/stbl/hl60_72h_delta_stbl.txt'\n",
    "'../RNA-seq/stbl/hl60_6h_delta_stbl.txt'\n",
    "\n",
    "## 5 other AML cell lines RNA-seq\n",
    "# RNA experssion \n",
    "'../RNA-seq/exp/kg1_delta_exp.c.txt'\n",
    "'../RNA-seq/exp/molm14_delta_exp.c.txt'\n",
    "'../RNA-seq/exp/ociaml2_delta_exp.c.txt' \n",
    "'../RNA-seq/exp/ociaml3_delta_exp.c.txt'\n",
    "'../RNA-seq/exp/thp1_delta_exp.c.txt'\n",
    "# RNA stability  \n",
    "'../RNA-seq/stbl/kg1_delta_stbl.txt' \n",
    "'../RNA-seq/stbl/molm14_delta_stbl.txt' \n",
    "'../RNA-seq/stbl/ociaml2_delta_stbl.txt' \n",
    "'../RNA-seq/stbl/ociaml3_delta_stbl.txt'\n",
    "'../RNA-seq/stbl/thp1_delta_stbl.txt'\n",
    "\n",
    "## drug combination\n",
    "# RNA experssion \n",
    "'../DAC-rg3039/RNA-seq/comb_vs_decitabine_delta_exp.c.txt'\n",
    "'../DAC-rg3039/RNA-seq/comb_vs_dmso_delta_exp.c.txt'\n",
    "\n",
    ")\n",
    "\n",
    "for exp in \"${Experiments[@]}\"; do\n",
    "    for geneset in \"${Genesets[@]}\"; do\n",
    "    \n",
    "        echo $exp $geneset\n",
    "        base=`basename $exp`\n",
    "        base=${base/.txt/}\n",
    "        \n",
    "        # # remove results from previous run \n",
    "        # rm ${exp}_GENESET\n",
    "        \n",
    "        # get intersect\n",
    "        awk 'NR==FNR{A[$1];next}$1 in A' $exp mtyl-enrichment/${geneset}.txt > mtyl-enrichment/${geneset}_${base}.txt\n",
    "        \n",
    "        perl ${TEISERDIR}run_mi_gene_list.pl \\\n",
    "            --expfile=$exp \\\n",
    "            --genefile=mtyl-enrichment/${geneset}_${base}.txt \\\n",
    "            --exptype=continuous \\\n",
    "            --ebins=11 \\\n",
    "            --species=human \\\n",
    "            --doremovedups=0 \\\n",
    "            --doremoveextra=0 &> mtyl-enrichment/${geneset}_${base}.log\n",
    "        \n",
    "        # # remove results from previous run \n",
    "        # rm -fr mtyl-enrichment/${geneset}_${base}\n",
    "        \n",
    "        rm mtyl-enrichment/${geneset}_${base}.txt mtyl-enrichment/${geneset}_${base}.log\n",
    "        mv ${exp}_GENESET mtyl-enrichment/${geneset}_${base}_GENESET\n",
    "        \n",
    "        echo 'done!'\n",
    "        \n",
    "    done\n",
    "\n",
    "done"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "3. Merge hypo and hyper results"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "comps = [(\n",
    "    comp.split('/')[1].split('_mtyl_')[0],\n",
    "    comp.split('/')[1].split('_mtyl_')[1].replace('_GENESET','')\n",
    ") for comp in glob('mtyl-enrichment/*delta*')]\n",
    "\n",
    "comps.sort(key=lambda a: len(a[1]))\n",
    "\n",
    "expfiles = {}\n",
    "summaryfiles = {}\n",
    "pvmatrixfiles = {}\n",
    "\n",
    "for c in {comp for _,comp in comps}:\n",
    "    \n",
    "    path = f'mtyl-enrichment/{c}'\n",
    "    os.mkdir(path) \n",
    "    expfiles[c] = {}\n",
    "    summaryfiles[c] = {}\n",
    "    pvmatrixfiles[c] = {}\n",
    "    \n",
    "    for m in {mtyl for mtyl,_ in comps}:\n",
    "        expfiles[c][m] = f'mtyl-enrichment/{m}_mtyl_{c}_GENESET/{c}.txt'\n",
    "        summaryfiles[c][m] = f'mtyl-enrichment/{m}_mtyl_{c}_GENESET/{c}.txt.summary'\n",
    "        pvmatrixfiles[c][m] = f'mtyl-enrichment/{m}_mtyl_{c}_GENESET/{c}.txt.matrix'\n",
    "    \n",
    "    exp_df = pd.read_csv(expfiles[c]['hyper'],sep='\\t', header=None)\n",
    "    exp_df.to_csv(f'{path}/{c}.txt',header=None,index=False,sep='\\t')\n",
    "    \n",
    "    sum_df = pd.concat([pd.read_csv(summaryfiles[c]['hyper'],sep='\\t'),pd.read_csv(summaryfiles[c]['hypo'],sep='\\t')])\n",
    "    sum_df['index'] = ['Hyper-methylated geneset','Hypo-methylated geneset']\n",
    "    sum_df.to_csv(f'{path}/{c}.txt.summary',index=False,sep='\\t')\n",
    "    \n",
    "    \n",
    "    mtx_df = pd.concat([pd.read_csv(pvmatrixfiles[c]['hyper'],sep='\\t'),pd.read_csv(pvmatrixfiles[c]['hypo'],sep='\\t')])\n",
    "    mtx_df.MOTIF = ['Hyper-methylated geneset','Hypo-methylated geneset']\n",
    "    mtx_df.to_csv(f'{path}/{c}.txt.matrix',index=False,sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "comb_vs_decitabine_delta_exp.c\n",
      "done!\n",
      "comb_vs_dmso_delta_exp.c\n",
      "done!\n",
      "hl60_120h_delta_exp.c\n",
      "done!\n",
      "hl60_120h_delta_stbl\n",
      "done!\n",
      "hl60_6h_delta_exp.c\n",
      "done!\n",
      "hl60_6h_delta_stbl\n",
      "done!\n",
      "hl60_72h_delta_exp.c\n",
      "done!\n",
      "hl60_72h_delta_stbl\n",
      "done!\n",
      "hl60_72h_only_delta_exp.c\n",
      "done!\n",
      "hl60_delta_te\n",
      "done!\n",
      "kg1_delta_exp.c\n",
      "done!\n",
      "kg1_delta_stbl\n",
      "done!\n",
      "molm14_delta_exp.c\n",
      "done!\n",
      "molm14_delta_stbl\n",
      "done!\n",
      "ociaml2_delta_exp.c\n",
      "done!\n",
      "ociaml2_delta_stbl\n",
      "done!\n",
      "ociaml3_delta_exp.c\n",
      "done!\n",
      "ociaml3_delta_stbl\n",
      "done!\n",
      "thp1_delta_exp.c\n",
      "done!\n",
      "thp1_delta_stbl\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd mtyl-enrichment\n",
    "for exp in `ls | grep -v '_GENESET' | grep 'delta'`; do \n",
    "\n",
    "    echo $exp\n",
    "\n",
    "    cd $exp\n",
    "\n",
    "    perl ${TEISERDIR}Scripts/teiser_draw_matrix.pl \\\n",
    "        --pvmatrixfile=${exp}.txt.matrix \\\n",
    "        --summaryfile=${exp}.txt.summary \\\n",
    "        --expfile=${exp}.txt \\\n",
    "        --quantized=0 \\\n",
    "        --order=0 \\\n",
    "        --min=-10 --max=10 --cluster=5 \\\n",
    "        --colmap=${TEISERDIR}Scripts/HEATMAPS/cmap_1.txt &> ${exp}.log\n",
    "        \n",
    "        rm ${exp}.log\n",
    "    cd ..\n",
    "    echo \"done!\"\n",
    "\n",
    "done \n",
    "cd ../"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Redraw heatmaps using `--min=-3 --max=3` thresholds for those plots which have smaller range of signals:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hl60_6h_delta_stbl\n",
      "done!\n",
      "hl60_72h_delta_stbl\n",
      "done!\n",
      "hl60_120h_delta_stbl\n",
      "done!\n",
      "kg1_delta_stbl\n",
      "done!\n",
      "ociaml2_delta_stbl\n",
      "done!\n",
      "molm14_delta_stbl\n",
      "done!\n",
      "ociaml3_delta_stbl\n",
      "done!\n",
      "thp1_delta_stbl\n",
      "done!\n",
      "hl60_delta_te\n",
      "done!\n"
     ]
    }
   ],
   "source": [
    "%%bash \n",
    "cd mtyl-enrichment\n",
    "\n",
    "declare -a Experiments=(\n",
    "'hl60_6h_delta_stbl' 'hl60_72h_delta_stbl' 'hl60_120h_delta_stbl' \n",
    "'kg1_delta_stbl' 'ociaml2_delta_stbl' 'molm14_delta_stbl' \n",
    "'ociaml3_delta_stbl' 'thp1_delta_stbl'\n",
    "'hl60_delta_te'\n",
    ")\n",
    "for exp in \"${Experiments[@]}\"; do\n",
    "\n",
    "    echo $exp\n",
    "\n",
    "    cd $exp\n",
    "\n",
    "    perl /flash/bin/TEISERv1.1/Scripts/teiser_draw_matrix.pl \\\n",
    "        --pvmatrixfile=${exp}.txt.matrix \\\n",
    "        --summaryfile=${exp}.txt.summary \\\n",
    "        --expfile=${exp}.txt \\\n",
    "        --quantized=0 \\\n",
    "        --order=0 \\\n",
    "        --min=-3 --max=3 --cluster=5 \\\n",
    "        --colmap=${TEISERDIR}Scripts/HEATMAPS/cmap_1.txt &> ${exp}.log\n",
    "        \n",
    "        rm ${exp}.log\n",
    "    cd ..\n",
    "    echo \"done!\"\n",
    "\n",
    "done \n",
    "cd ../"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# !mkdir mtyl-enrichment/log\n",
    "# !mv mtyl-enrichment/*_GENESET mtyl-enrichment/log\n",
    "# !mv mtyl-enrichment/*.log mtyl-enrichment/log\n",
    "!rm -r mtyl-enrichment/*_GENESET"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "4. Make `png` figures:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "for pdf in mtyl-enrichment/*/*.txt.summary.pdf; do \n",
    "    png=${pdf/.pdf/.png}\n",
    "    di=`dirname $pdf`\n",
    "    out=`basename $di`\n",
    "    \n",
    "    bash /rumi/shams/abe/GitHub/Abe/my_scripts/pdf2png.sh $pdf \n",
    "\n",
    "    mv $pdf mtyl-enrichment/${out}.pdf\n",
    "    mv $png mtyl-enrichment/${out}.png\n",
    "    \n",
    "done "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "rm mtyl-enrichment/*.png"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%bash \n",
    "for f in mtyl-enrichment/*.c.pdf; do o=${f/.c./.}; mv $f $o; done "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "rm mtyl-enrichment/hyper_mtyl.txt mtyl-enrichment/hypo_mtyl.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Wed Dec 22 18:10:46 PST 2021\n"
     ]
    }
   ],
   "source": [
    "!date"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:deseq]",
   "language": "python",
   "name": "conda-env-deseq-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
